{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Concatonate All Deseq2 Data with AHR, DRE, RefGenome Metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 1. Import Required Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 2. Import Deseq2 Data and All Other Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>pDRE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Spire1</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Wwox</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Apbb2</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sdk1</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Apoc2</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23597</th>\n",
       "      <td>lnc14044</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23598</th>\n",
       "      <td>lnc30181</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23599</th>\n",
       "      <td>lnc30180</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23600</th>\n",
       "      <td>Akain1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23601</th>\n",
       "      <td>Erdr1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23601 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene  pDRE\n",
       "1        Spire1    75\n",
       "2          Wwox    47\n",
       "3         Apbb2    43\n",
       "4          Sdk1    38\n",
       "5         Apoc2    37\n",
       "...         ...   ...\n",
       "23597  lnc14044     1\n",
       "23598  lnc30181     1\n",
       "23599  lnc30180     1\n",
       "23600    Akain1     1\n",
       "23601     Erdr1     1\n",
       "\n",
       "[23601 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load DRE count data and clean it\n",
    "pDREs_Master = pd.read_csv('../../99_mm39_DREs/02_CleanUp_DRE_BED_File/DRE_Counts_by_Gene.txt', \n",
    "                            delimiter = '\\t')\n",
    "\n",
    "# Drop rows with missing values\n",
    "pDREs_Master = pDREs_Master.dropna()\n",
    "\n",
    "# Rename columns for consistency\n",
    "pDREs_Master = pDREs_Master.rename(columns={'gene_name': 'Gene', 'count': 'pDRE'})\n",
    "pDREs_Master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>AHR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc10851</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc7308</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hjurp</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc3293</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21275</th>\n",
       "      <td>Ralbp1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21276</th>\n",
       "      <td>lnc30141</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21277</th>\n",
       "      <td>lnc30142</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21278</th>\n",
       "      <td>lnc30143</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21279</th>\n",
       "      <td>Ddx3y</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21280 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene   AHR\n",
       "0           NaN  2652\n",
       "1      lnc10851   153\n",
       "2       lnc7308    51\n",
       "3         Hjurp    44\n",
       "4       lnc3293    43\n",
       "...         ...   ...\n",
       "21275    Ralbp1     1\n",
       "21276  lnc30141     1\n",
       "21277  lnc30142     1\n",
       "21278  lnc30143     1\n",
       "21279     Ddx3y     1\n",
       "\n",
       "[21280 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load AHR binding count data and rename columns\n",
    "AHR_Master = pd.read_csv('../../99_mm39_AHR_Enrichment/02_CleanUp_BED_File/AHR_Binding_Counts.txt', \n",
    "                            delimiter = '\\t')\n",
    "\n",
    "AHR_Master = AHR_Master.rename(columns={'gene_name': 'Gene', 'count': 'AHR'})\n",
    "\n",
    "AHR_Master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/local/52526326/ipykernel_141480/2770042493.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  Gene_Type = pd.read_csv('../../00_Karri_et_al_GTF/MGI_and_Karri_Annotations_Restructured_mm39.txt',\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['Chr', 'Start', 'End', 'blank', 'Score', 'Strand', 'Genebody', 'Exon',\n",
       "       'Frame', 'mgi_chr', 'mgi_start', 'mgi_end', 'MGI_Symbol', 'mgi_score',\n",
       "       'mgi_strand', 'gene_id', 'transcript_id', 'Gene', 'Type',\n",
       "       'NR_Annotation', 'NR_annotation2'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load and rename columns in gene annotation file for consistency\n",
    "Gene_Type = pd.read_csv('../../00_Karri_et_al_GTF/MGI_and_Karri_Annotations_Restructured_mm39.txt', \n",
    "                           delimiter = '\\t')\n",
    "\n",
    "\n",
    "Gene_Type = Gene_Type.rename(columns={'chrom': 'Chr', \n",
    "                                        'source': 'Genebody',\n",
    "                                        'feature': 'Exon',\n",
    "                                        'start': 'Start',\n",
    "                                        'end': 'End',\n",
    "                                        'score': 'Score',\n",
    "                                        'strand': 'Strand',\n",
    "                                        'frame': 'Frame',\n",
    "                                        'gene_name': 'Gene',\n",
    "                                        'gene_biotype': 'Type',\n",
    "                                        'NR_annotation': 'NR_Annotation',\n",
    "                                        'mgi_symbol': 'MGI_Symbol'\n",
    "                                       })\n",
    "\n",
    "\n",
    "Gene_Type.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load DESeq2 output files for all cell types across multiple dose comparisons (e.g., 0.01 vs 0, 0.03 vs 0, etc.)\n",
    "# Each file is read into a separate DataFrame and index is reset for consistency\n",
    "directory = './Results/03b_Deseq2_Output/'\n",
    "\n",
    "Deseq2_B_Cell_0_01_v_0 = pd.read_csv(directory + 'B_Cells_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_B_Cell_0_03_v_0 = pd.read_csv(directory + 'B_Cells_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_B_Cell_0_1_v_0 = pd.read_csv(directory + 'B_Cells_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_B_Cell_0_3_v_0 = pd.read_csv(directory + 'B_Cells_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_B_Cell_1_v_0 = pd.read_csv(directory + 'B_Cells_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_B_Cell_3_v_0 = pd.read_csv(directory + 'B_Cells_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_B_Cell_10_v_0 = pd.read_csv(directory + 'B_Cells_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_B_Cell_30_v_0 = pd.read_csv(directory + 'B_Cells_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n",
    "Deseq2_Centrilobular_Hepatocyte_0_01_v_0 = pd.read_csv(directory + 'Centrilobular_Hepatocytes_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Centrilobular_Hepatocyte_0_03_v_0 = pd.read_csv(directory + 'Centrilobular_Hepatocytes_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Centrilobular_Hepatocyte_0_1_v_0 = pd.read_csv(directory + 'Centrilobular_Hepatocytes_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Centrilobular_Hepatocyte_0_3_v_0 = pd.read_csv(directory + 'Centrilobular_Hepatocytes_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Centrilobular_Hepatocyte_1_v_0 = pd.read_csv(directory + 'Centrilobular_Hepatocytes_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Centrilobular_Hepatocyte_3_v_0 = pd.read_csv(directory + 'Centrilobular_Hepatocytes_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Centrilobular_Hepatocyte_10_v_0 = pd.read_csv(directory + 'Centrilobular_Hepatocytes_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Centrilobular_Hepatocyte_30_v_0 = pd.read_csv(directory + 'Centrilobular_Hepatocytes_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n",
    "Deseq2_Cholangiocyte_0_01_v_0 = pd.read_csv(directory + 'Cholangiocytes_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Cholangiocyte_0_03_v_0 = pd.read_csv(directory + 'Cholangiocytes_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Cholangiocyte_0_1_v_0 = pd.read_csv(directory + 'Cholangiocytes_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Cholangiocyte_0_3_v_0 = pd.read_csv(directory + 'Cholangiocytes_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Cholangiocyte_1_v_0 = pd.read_csv(directory + 'Cholangiocytes_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Cholangiocyte_3_v_0 = pd.read_csv(directory + 'Cholangiocytes_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Cholangiocyte_10_v_0 = pd.read_csv(directory + 'Cholangiocytes_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Cholangiocyte_30_v_0 = pd.read_csv(directory + 'Cholangiocytes_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n",
    "Deseq2_HSC_0_01_v_0 = pd.read_csv(directory + 'HSCs_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_HSC_0_03_v_0 = pd.read_csv(directory + 'HSCs_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_HSC_0_1_v_0 = pd.read_csv(directory + 'HSCs_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_HSC_0_3_v_0 = pd.read_csv(directory + 'HSCs_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_HSC_1_v_0 = pd.read_csv(directory + 'HSCs_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_HSC_3_v_0 = pd.read_csv(directory + 'HSCs_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_HSC_10_v_0 = pd.read_csv(directory + 'HSCs_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_HSC_30_v_0 = pd.read_csv(directory + 'HSCs_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n",
    "Deseq2_LSEC_0_01_v_0 = pd.read_csv(directory + 'LSECs_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_LSEC_0_03_v_0 = pd.read_csv(directory + 'LSECs_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_LSEC_0_1_v_0 = pd.read_csv(directory + 'LSECs_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_LSEC_0_3_v_0 = pd.read_csv(directory + 'LSECs_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_LSEC_1_v_0 = pd.read_csv(directory + 'LSECs_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_LSEC_3_v_0 = pd.read_csv(directory + 'LSECs_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_LSEC_10_v_0 = pd.read_csv(directory + 'LSECs_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_LSEC_30_v_0 = pd.read_csv(directory + 'LSECs_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n",
    "Deseq2_Macrophage_0_01_v_0 = pd.read_csv(directory + 'Macrophages_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Macrophage_0_03_v_0 = pd.read_csv(directory + 'Macrophages_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Macrophage_0_1_v_0 = pd.read_csv(directory + 'Macrophages_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Macrophage_0_3_v_0 = pd.read_csv(directory + 'Macrophages_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Macrophage_1_v_0 = pd.read_csv(directory + 'Macrophages_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Macrophage_3_v_0 = pd.read_csv(directory + 'Macrophages_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Macrophage_10_v_0 = pd.read_csv(directory + 'Macrophages_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Macrophage_30_v_0 = pd.read_csv(directory + 'Macrophages_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n",
    "Deseq2_Neutrophil_0_01_v_0 = pd.read_csv(directory + 'Neutrophils_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Neutrophil_0_03_v_0 = pd.read_csv(directory + 'Neutrophils_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Neutrophil_0_1_v_0 = pd.read_csv(directory + 'Neutrophils_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Neutrophil_0_3_v_0 = pd.read_csv(directory + 'Neutrophils_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Neutrophil_1_v_0 = pd.read_csv(directory + 'Neutrophils_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Neutrophil_3_v_0 = pd.read_csv(directory + 'Neutrophils_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Neutrophil_10_v_0 = pd.read_csv(directory + 'Neutrophils_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Neutrophil_30_v_0 = pd.read_csv(directory + 'Neutrophils_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n",
    "Deseq2_PF_0_01_v_0 = pd.read_csv(directory + 'PFs_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_PF_0_03_v_0 = pd.read_csv(directory + 'PFs_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_PF_0_1_v_0 = pd.read_csv(directory + 'PFs_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_PF_0_3_v_0 = pd.read_csv(directory + 'PFs_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_PF_1_v_0 = pd.read_csv(directory + 'PFs_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_PF_3_v_0 = pd.read_csv(directory + 'PFs_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_PF_10_v_0 = pd.read_csv(directory + 'PFs_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_PF_30_v_0 = pd.read_csv(directory + 'PFs_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n",
    "Deseq2_Periportal_Hepatocyte_0_01_v_0 = pd.read_csv(directory + 'Periportal_Hepatocytes_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Periportal_Hepatocyte_0_03_v_0 = pd.read_csv(directory + 'Periportal_Hepatocytes_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Periportal_Hepatocyte_0_1_v_0 = pd.read_csv(directory + 'Periportal_Hepatocytes_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Periportal_Hepatocyte_0_3_v_0 = pd.read_csv(directory + 'Periportal_Hepatocytes_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Periportal_Hepatocyte_1_v_0 = pd.read_csv(directory + 'Periportal_Hepatocytes_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Periportal_Hepatocyte_3_v_0 = pd.read_csv(directory + 'Periportal_Hepatocytes_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Periportal_Hepatocyte_10_v_0 = pd.read_csv(directory + 'Periportal_Hepatocytes_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_Periportal_Hepatocyte_30_v_0 = pd.read_csv(directory + 'Periportal_Hepatocytes_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n",
    "Deseq2_T_Cell_0_01_v_0 = pd.read_csv(directory + 'T_Cells_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_T_Cell_0_03_v_0 = pd.read_csv(directory + 'T_Cells_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_T_Cell_0_1_v_0 = pd.read_csv(directory + 'T_Cells_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_T_Cell_0_3_v_0 = pd.read_csv(directory + 'T_Cells_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_T_Cell_1_v_0 = pd.read_csv(directory + 'T_Cells_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_T_Cell_3_v_0 = pd.read_csv(directory + 'T_Cells_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_T_Cell_10_v_0 = pd.read_csv(directory + 'T_Cells_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_T_Cell_30_v_0 = pd.read_csv(directory + 'T_Cells_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n",
    "Deseq2_pDC_0_01_v_0 = pd.read_csv(directory + 'pDCs_0.01_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_pDC_0_03_v_0 = pd.read_csv(directory + 'pDCs_0.03_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_pDC_0_1_v_0 = pd.read_csv(directory + 'pDCs_0.1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_pDC_0_3_v_0 = pd.read_csv(directory + 'pDCs_0.3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_pDC_1_v_0 = pd.read_csv(directory + 'pDCs_1_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_pDC_3_v_0 = pd.read_csv(directory + 'pDCs_3_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_pDC_10_v_0 = pd.read_csv(directory + 'pDCs_10_v_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_pDC_30_v_0 = pd.read_csv(directory + 'pDCs_30_v_0.txt', delimiter = '\\t').reset_index()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add 'Celltype' column to each DESeq2 result DataFrame to label the corresponding cell type for downstream analysis\n",
    "Deseq2_B_Cell_0_01_v_0['Celltype'] = 'B Cells'\n",
    "Deseq2_B_Cell_0_03_v_0['Celltype'] = 'B Cells'\n",
    "Deseq2_B_Cell_0_1_v_0['Celltype'] = 'B Cells'\n",
    "Deseq2_B_Cell_0_3_v_0['Celltype'] = 'B Cells'\n",
    "Deseq2_B_Cell_1_v_0['Celltype'] = 'B Cells'\n",
    "Deseq2_B_Cell_3_v_0['Celltype'] = 'B Cells'\n",
    "Deseq2_B_Cell_10_v_0['Celltype'] = 'B Cells'\n",
    "Deseq2_B_Cell_30_v_0['Celltype'] = 'B Cells'\n",
    "\n",
    "Deseq2_Centrilobular_Hepatocyte_0_01_v_0['Celltype'] = 'Centrilobular Hepatocytes'\n",
    "Deseq2_Centrilobular_Hepatocyte_0_03_v_0['Celltype'] = 'Centrilobular Hepatocytes'\n",
    "Deseq2_Centrilobular_Hepatocyte_0_1_v_0['Celltype'] = 'Centrilobular Hepatocytes'\n",
    "Deseq2_Centrilobular_Hepatocyte_0_3_v_0['Celltype'] = 'Centrilobular Hepatocytes'\n",
    "Deseq2_Centrilobular_Hepatocyte_1_v_0['Celltype'] = 'Centrilobular Hepatocytes'\n",
    "Deseq2_Centrilobular_Hepatocyte_3_v_0['Celltype'] = 'Centrilobular Hepatocytes'\n",
    "Deseq2_Centrilobular_Hepatocyte_10_v_0['Celltype'] = 'Centrilobular Hepatocytes'\n",
    "Deseq2_Centrilobular_Hepatocyte_30_v_0['Celltype'] = 'Centrilobular Hepatocytes'\n",
    "\n",
    "Deseq2_Cholangiocyte_0_01_v_0['Celltype'] = 'Cholangiocytes'\n",
    "Deseq2_Cholangiocyte_0_03_v_0['Celltype'] = 'Cholangiocytes'\n",
    "Deseq2_Cholangiocyte_0_1_v_0['Celltype'] = 'Cholangiocytes'\n",
    "Deseq2_Cholangiocyte_0_3_v_0['Celltype'] = 'Cholangiocytes'\n",
    "Deseq2_Cholangiocyte_1_v_0['Celltype'] = 'Cholangiocytes'\n",
    "Deseq2_Cholangiocyte_3_v_0['Celltype'] = 'Cholangiocytes'\n",
    "Deseq2_Cholangiocyte_10_v_0['Celltype'] = 'Cholangiocytes'\n",
    "Deseq2_Cholangiocyte_30_v_0['Celltype'] = 'Cholangiocytes'\n",
    "\n",
    "Deseq2_HSC_0_01_v_0['Celltype'] = 'HSCs'\n",
    "Deseq2_HSC_0_03_v_0['Celltype'] = 'HSCs'\n",
    "Deseq2_HSC_0_1_v_0['Celltype'] = 'HSCs'\n",
    "Deseq2_HSC_0_3_v_0['Celltype'] = 'HSCs'\n",
    "Deseq2_HSC_1_v_0['Celltype'] = 'HSCs'\n",
    "Deseq2_HSC_3_v_0['Celltype'] = 'HSCs'\n",
    "Deseq2_HSC_10_v_0['Celltype'] = 'HSCs'\n",
    "Deseq2_HSC_30_v_0['Celltype'] = 'HSCs'\n",
    "\n",
    "Deseq2_LSEC_0_01_v_0['Celltype'] = 'LSECs'\n",
    "Deseq2_LSEC_0_03_v_0['Celltype'] = 'LSECs'\n",
    "Deseq2_LSEC_0_1_v_0['Celltype'] = 'LSECs'\n",
    "Deseq2_LSEC_0_3_v_0['Celltype'] = 'LSECs'\n",
    "Deseq2_LSEC_1_v_0['Celltype'] = 'LSECs'\n",
    "Deseq2_LSEC_3_v_0['Celltype'] = 'LSECs'\n",
    "Deseq2_LSEC_10_v_0['Celltype'] = 'LSECs'\n",
    "Deseq2_LSEC_30_v_0['Celltype'] = 'LSECs'\n",
    "\n",
    "Deseq2_Macrophage_0_01_v_0['Celltype'] = 'Macrophages'\n",
    "Deseq2_Macrophage_0_03_v_0['Celltype'] = 'Macrophages'\n",
    "Deseq2_Macrophage_0_1_v_0['Celltype'] = 'Macrophages'\n",
    "Deseq2_Macrophage_0_3_v_0['Celltype'] = 'Macrophages'\n",
    "Deseq2_Macrophage_1_v_0['Celltype'] = 'Macrophages'\n",
    "Deseq2_Macrophage_3_v_0['Celltype'] = 'Macrophages'\n",
    "Deseq2_Macrophage_10_v_0['Celltype'] = 'Macrophages'\n",
    "Deseq2_Macrophage_30_v_0['Celltype'] = 'Macrophages'\n",
    "\n",
    "Deseq2_Neutrophil_0_01_v_0['Celltype'] = 'Neutrophils'\n",
    "Deseq2_Neutrophil_0_03_v_0['Celltype'] = 'Neutrophils'\n",
    "Deseq2_Neutrophil_0_1_v_0['Celltype'] = 'Neutrophils'\n",
    "Deseq2_Neutrophil_0_3_v_0['Celltype'] = 'Neutrophils'\n",
    "Deseq2_Neutrophil_1_v_0['Celltype'] = 'Neutrophils'\n",
    "Deseq2_Neutrophil_3_v_0['Celltype'] = 'Neutrophils'\n",
    "Deseq2_Neutrophil_10_v_0['Celltype'] = 'Neutrophils'\n",
    "Deseq2_Neutrophil_30_v_0['Celltype'] = 'Neutrophils'\n",
    "\n",
    "Deseq2_PF_0_01_v_0['Celltype'] = 'PFs'\n",
    "Deseq2_PF_0_03_v_0['Celltype'] = 'PFs'\n",
    "Deseq2_PF_0_1_v_0['Celltype'] = 'PFs'\n",
    "Deseq2_PF_0_3_v_0['Celltype'] = 'PFs'\n",
    "Deseq2_PF_1_v_0['Celltype'] = 'PFs'\n",
    "Deseq2_PF_3_v_0['Celltype'] = 'PFs'\n",
    "Deseq2_PF_10_v_0['Celltype'] = 'PFs'\n",
    "Deseq2_PF_30_v_0['Celltype'] = 'PFs'\n",
    "\n",
    "Deseq2_Periportal_Hepatocyte_0_01_v_0['Celltype'] = 'Periportal Hepatocytes'\n",
    "Deseq2_Periportal_Hepatocyte_0_03_v_0['Celltype'] = 'Periportal Hepatocytes'\n",
    "Deseq2_Periportal_Hepatocyte_0_1_v_0['Celltype'] = 'Periportal Hepatocytes'\n",
    "Deseq2_Periportal_Hepatocyte_0_3_v_0['Celltype'] = 'Periportal Hepatocytes'\n",
    "Deseq2_Periportal_Hepatocyte_1_v_0['Celltype'] = 'Periportal Hepatocytes'\n",
    "Deseq2_Periportal_Hepatocyte_3_v_0['Celltype'] = 'Periportal Hepatocytes'\n",
    "Deseq2_Periportal_Hepatocyte_10_v_0['Celltype'] = 'Periportal Hepatocytes'\n",
    "Deseq2_Periportal_Hepatocyte_30_v_0['Celltype'] = 'Periportal Hepatocytes'\n",
    "\n",
    "Deseq2_T_Cell_0_01_v_0['Celltype'] = 'T Cells'\n",
    "Deseq2_T_Cell_0_03_v_0['Celltype'] = 'T Cells'\n",
    "Deseq2_T_Cell_0_1_v_0['Celltype'] = 'T Cells'\n",
    "Deseq2_T_Cell_0_3_v_0['Celltype'] = 'T Cells'\n",
    "Deseq2_T_Cell_1_v_0['Celltype'] = 'T Cells'\n",
    "Deseq2_T_Cell_3_v_0['Celltype'] = 'T Cells'\n",
    "Deseq2_T_Cell_10_v_0['Celltype'] = 'T Cells'\n",
    "Deseq2_T_Cell_30_v_0['Celltype'] = 'T Cells'\n",
    "\n",
    "Deseq2_pDC_0_01_v_0['Celltype'] = 'pDCs'\n",
    "Deseq2_pDC_0_03_v_0['Celltype'] = 'pDCs'\n",
    "Deseq2_pDC_0_1_v_0['Celltype'] = 'pDCs'\n",
    "Deseq2_pDC_0_3_v_0['Celltype'] = 'pDCs'\n",
    "Deseq2_pDC_1_v_0['Celltype'] = 'pDCs'\n",
    "Deseq2_pDC_3_v_0['Celltype'] = 'pDCs'\n",
    "Deseq2_pDC_10_v_0['Celltype'] = 'pDCs'\n",
    "Deseq2_pDC_30_v_0['Celltype'] = 'pDCs'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign 'Dose' labels to each DESeq2 result DataFrame to indicate the treatment concentration compared to control\n",
    "Deseq2_B_Cell_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_B_Cell_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_B_Cell_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_B_Cell_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_B_Cell_1_v_0['Dose'] = '1'\n",
    "Deseq2_B_Cell_3_v_0['Dose'] = '3'\n",
    "Deseq2_B_Cell_10_v_0['Dose'] = '10'\n",
    "Deseq2_B_Cell_30_v_0['Dose'] = '30'\n",
    "\n",
    "Deseq2_Centrilobular_Hepatocyte_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_Centrilobular_Hepatocyte_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_Centrilobular_Hepatocyte_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_Centrilobular_Hepatocyte_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_Centrilobular_Hepatocyte_1_v_0['Dose'] = '1'\n",
    "Deseq2_Centrilobular_Hepatocyte_3_v_0['Dose'] = '3'\n",
    "Deseq2_Centrilobular_Hepatocyte_10_v_0['Dose'] = '10'\n",
    "Deseq2_Centrilobular_Hepatocyte_30_v_0['Dose'] = '30'\n",
    "\n",
    "Deseq2_Cholangiocyte_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_Cholangiocyte_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_Cholangiocyte_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_Cholangiocyte_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_Cholangiocyte_1_v_0['Dose'] = '1'\n",
    "Deseq2_Cholangiocyte_3_v_0['Dose'] = '3'\n",
    "Deseq2_Cholangiocyte_10_v_0['Dose'] = '10'\n",
    "Deseq2_Cholangiocyte_30_v_0['Dose'] = '30'\n",
    "\n",
    "Deseq2_HSC_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_HSC_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_HSC_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_HSC_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_HSC_1_v_0['Dose'] = '1'\n",
    "Deseq2_HSC_3_v_0['Dose'] = '3'\n",
    "Deseq2_HSC_10_v_0['Dose'] = '10'\n",
    "Deseq2_HSC_30_v_0['Dose'] = '30'\n",
    "\n",
    "Deseq2_LSEC_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_LSEC_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_LSEC_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_LSEC_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_LSEC_1_v_0['Dose'] = '1'\n",
    "Deseq2_LSEC_3_v_0['Dose'] = '3'\n",
    "Deseq2_LSEC_10_v_0['Dose'] = '10'\n",
    "Deseq2_LSEC_30_v_0['Dose'] = '30'\n",
    "\n",
    "Deseq2_Macrophage_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_Macrophage_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_Macrophage_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_Macrophage_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_Macrophage_1_v_0['Dose'] = '1'\n",
    "Deseq2_Macrophage_3_v_0['Dose'] = '3'\n",
    "Deseq2_Macrophage_10_v_0['Dose'] = '10'\n",
    "Deseq2_Macrophage_30_v_0['Dose'] = '30'\n",
    "\n",
    "Deseq2_Neutrophil_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_Neutrophil_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_Neutrophil_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_Neutrophil_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_Neutrophil_1_v_0['Dose'] = '1'\n",
    "Deseq2_Neutrophil_3_v_0['Dose'] = '3'\n",
    "Deseq2_Neutrophil_10_v_0['Dose'] = '10'\n",
    "Deseq2_Neutrophil_30_v_0['Dose'] = '30'\n",
    "\n",
    "Deseq2_PF_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_PF_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_PF_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_PF_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_PF_1_v_0['Dose'] = '1'\n",
    "Deseq2_PF_3_v_0['Dose'] = '3'\n",
    "Deseq2_PF_10_v_0['Dose'] = '10'\n",
    "Deseq2_PF_30_v_0['Dose'] = '30'\n",
    "\n",
    "Deseq2_Periportal_Hepatocyte_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_Periportal_Hepatocyte_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_Periportal_Hepatocyte_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_Periportal_Hepatocyte_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_Periportal_Hepatocyte_1_v_0['Dose'] = '1'\n",
    "Deseq2_Periportal_Hepatocyte_3_v_0['Dose'] = '3'\n",
    "Deseq2_Periportal_Hepatocyte_10_v_0['Dose'] = '10'\n",
    "Deseq2_Periportal_Hepatocyte_30_v_0['Dose'] = '30'\n",
    "\n",
    "Deseq2_T_Cell_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_T_Cell_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_T_Cell_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_T_Cell_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_T_Cell_1_v_0['Dose'] = '1'\n",
    "Deseq2_T_Cell_3_v_0['Dose'] = '3'\n",
    "Deseq2_T_Cell_10_v_0['Dose'] = '10'\n",
    "Deseq2_T_Cell_30_v_0['Dose'] = '30'\n",
    "\n",
    "Deseq2_pDC_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_pDC_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_pDC_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_pDC_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_pDC_1_v_0['Dose'] = '1'\n",
    "Deseq2_pDC_3_v_0['Dose'] = '3'\n",
    "Deseq2_pDC_10_v_0['Dose'] = '10'\n",
    "Deseq2_pDC_30_v_0['Dose'] = '30'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Celltype</th>\n",
       "      <th>Dose</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23217</th>\n",
       "      <td>Xlr4a</td>\n",
       "      <td>5.349154</td>\n",
       "      <td>-2.648294</td>\n",
       "      <td>2.64747</td>\n",
       "      <td>-1.000311</td>\n",
       "      <td>0.31716</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Periportal Hepatocytes</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       index  baseMean  log2FoldChange    lfcSE      stat   pvalue  padj  \\\n",
       "23217  Xlr4a  5.349154       -2.648294  2.64747 -1.000311  0.31716   1.0   \n",
       "\n",
       "                     Celltype  Dose  \n",
       "23217  Periportal Hepatocytes  0.01  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Deseq2_Periportal_Hepatocyte_0_01_v_0[Deseq2_Periportal_Hepatocyte_0_01_v_0['index'] == 'Xlr4a']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 3. Concat All Data Into One Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Concat into one dataframe\n",
    "Deseq2_Master = pd.concat([Deseq2_B_Cell_0_01_v_0, \n",
    "                            Deseq2_B_Cell_0_03_v_0, \n",
    "                            Deseq2_B_Cell_0_1_v_0, \n",
    "                            Deseq2_B_Cell_0_3_v_0, \n",
    "                            Deseq2_B_Cell_1_v_0, \n",
    "                            Deseq2_B_Cell_3_v_0, \n",
    "                            Deseq2_B_Cell_10_v_0, \n",
    "                            Deseq2_B_Cell_30_v_0, \n",
    "                            Deseq2_Centrilobular_Hepatocyte_0_01_v_0, \n",
    "                            Deseq2_Centrilobular_Hepatocyte_0_03_v_0, \n",
    "                            Deseq2_Centrilobular_Hepatocyte_0_1_v_0, \n",
    "                            Deseq2_Centrilobular_Hepatocyte_0_3_v_0, \n",
    "                            Deseq2_Centrilobular_Hepatocyte_1_v_0, \n",
    "                            Deseq2_Centrilobular_Hepatocyte_3_v_0, \n",
    "                            Deseq2_Centrilobular_Hepatocyte_10_v_0, \n",
    "                            Deseq2_Centrilobular_Hepatocyte_30_v_0, \n",
    "                            Deseq2_Cholangiocyte_0_01_v_0, \n",
    "                            Deseq2_Cholangiocyte_0_03_v_0, \n",
    "                            Deseq2_Cholangiocyte_0_1_v_0, \n",
    "                            Deseq2_Cholangiocyte_0_3_v_0, \n",
    "                            Deseq2_Cholangiocyte_1_v_0, \n",
    "                            Deseq2_Cholangiocyte_3_v_0, \n",
    "                            Deseq2_Cholangiocyte_10_v_0, \n",
    "                            Deseq2_Cholangiocyte_30_v_0, \n",
    "                            Deseq2_HSC_0_01_v_0, \n",
    "                            Deseq2_HSC_0_03_v_0, \n",
    "                            Deseq2_HSC_0_1_v_0, \n",
    "                            Deseq2_HSC_0_3_v_0, \n",
    "                            Deseq2_HSC_1_v_0, \n",
    "                            Deseq2_HSC_3_v_0, \n",
    "                            Deseq2_HSC_10_v_0, \n",
    "                            Deseq2_HSC_30_v_0, \n",
    "                            Deseq2_LSEC_0_01_v_0, \n",
    "                            Deseq2_LSEC_0_03_v_0, \n",
    "                            Deseq2_LSEC_0_1_v_0, \n",
    "                            Deseq2_LSEC_0_3_v_0, \n",
    "                            Deseq2_LSEC_1_v_0, \n",
    "                            Deseq2_LSEC_3_v_0, \n",
    "                            Deseq2_LSEC_10_v_0, \n",
    "                            Deseq2_LSEC_30_v_0, \n",
    "                            Deseq2_Macrophage_0_01_v_0, \n",
    "                            Deseq2_Macrophage_0_03_v_0, \n",
    "                            Deseq2_Macrophage_0_1_v_0, \n",
    "                            Deseq2_Macrophage_0_3_v_0, \n",
    "                            Deseq2_Macrophage_1_v_0, \n",
    "                            Deseq2_Macrophage_3_v_0, \n",
    "                            Deseq2_Macrophage_10_v_0, \n",
    "                            Deseq2_Macrophage_30_v_0, \n",
    "                            Deseq2_Neutrophil_0_01_v_0, \n",
    "                            Deseq2_Neutrophil_0_03_v_0, \n",
    "                            Deseq2_Neutrophil_0_1_v_0, \n",
    "                            Deseq2_Neutrophil_0_3_v_0, \n",
    "                            Deseq2_Neutrophil_1_v_0, \n",
    "                            Deseq2_Neutrophil_3_v_0, \n",
    "                            Deseq2_Neutrophil_10_v_0, \n",
    "                            Deseq2_Neutrophil_30_v_0, \n",
    "                            Deseq2_PF_0_01_v_0, \n",
    "                            Deseq2_PF_0_03_v_0, \n",
    "                            Deseq2_PF_0_1_v_0, \n",
    "                            Deseq2_PF_0_3_v_0, \n",
    "                            Deseq2_PF_1_v_0, \n",
    "                            Deseq2_PF_3_v_0, \n",
    "                            Deseq2_PF_10_v_0, \n",
    "                            Deseq2_PF_30_v_0, \n",
    "                            Deseq2_Periportal_Hepatocyte_0_01_v_0, \n",
    "                            Deseq2_Periportal_Hepatocyte_0_03_v_0, \n",
    "                            Deseq2_Periportal_Hepatocyte_0_1_v_0, \n",
    "                            Deseq2_Periportal_Hepatocyte_0_3_v_0, \n",
    "                            Deseq2_Periportal_Hepatocyte_1_v_0, \n",
    "                            Deseq2_Periportal_Hepatocyte_3_v_0, \n",
    "                            Deseq2_Periportal_Hepatocyte_10_v_0, \n",
    "                            Deseq2_Periportal_Hepatocyte_30_v_0, \n",
    "                            Deseq2_T_Cell_0_01_v_0, \n",
    "                            Deseq2_T_Cell_0_03_v_0, \n",
    "                            Deseq2_T_Cell_0_1_v_0, \n",
    "                            Deseq2_T_Cell_0_3_v_0, \n",
    "                            Deseq2_T_Cell_1_v_0, \n",
    "                            Deseq2_T_Cell_3_v_0, \n",
    "                            Deseq2_T_Cell_10_v_0, \n",
    "                            Deseq2_T_Cell_30_v_0, \n",
    "                            Deseq2_pDC_0_01_v_0, \n",
    "                            Deseq2_pDC_0_03_v_0, \n",
    "                            Deseq2_pDC_0_1_v_0, \n",
    "                            Deseq2_pDC_0_3_v_0, \n",
    "                            Deseq2_pDC_1_v_0, \n",
    "                            Deseq2_pDC_3_v_0, \n",
    "                            Deseq2_pDC_10_v_0, \n",
    "                            Deseq2_pDC_30_v_0 ],\n",
    "                          ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Celltype</th>\n",
       "      <th>Dose</th>\n",
       "      <th>Fold-Change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc31752</td>\n",
       "      <td>5632.003911</td>\n",
       "      <td>0.060929</td>\n",
       "      <td>0.161392</td>\n",
       "      <td>0.377519</td>\n",
       "      <td>0.705788</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B Cells</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.043137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc4822</td>\n",
       "      <td>3146.387654</td>\n",
       "      <td>-0.368362</td>\n",
       "      <td>0.385441</td>\n",
       "      <td>-0.955691</td>\n",
       "      <td>0.339228</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B Cells</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.774662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Foxp1</td>\n",
       "      <td>941.129495</td>\n",
       "      <td>-0.079404</td>\n",
       "      <td>0.359311</td>\n",
       "      <td>-0.220991</td>\n",
       "      <td>0.825099</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B Cells</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.946448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ebf1</td>\n",
       "      <td>616.758548</td>\n",
       "      <td>-0.548443</td>\n",
       "      <td>0.564479</td>\n",
       "      <td>-0.971590</td>\n",
       "      <td>0.331254</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B Cells</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.683758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Dpyd</td>\n",
       "      <td>924.659365</td>\n",
       "      <td>0.007061</td>\n",
       "      <td>0.575075</td>\n",
       "      <td>0.012278</td>\n",
       "      <td>0.990204</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>B Cells</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.004906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1124699</th>\n",
       "      <td>lnc1191</td>\n",
       "      <td>1.467205</td>\n",
       "      <td>0.158104</td>\n",
       "      <td>1.308503</td>\n",
       "      <td>0.120828</td>\n",
       "      <td>0.903827</td>\n",
       "      <td>0.955543</td>\n",
       "      <td>pDCs</td>\n",
       "      <td>30</td>\n",
       "      <td>1.115820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1124700</th>\n",
       "      <td>lnc1440</td>\n",
       "      <td>2.409755</td>\n",
       "      <td>-4.176886</td>\n",
       "      <td>1.226191</td>\n",
       "      <td>-3.406392</td>\n",
       "      <td>0.000658</td>\n",
       "      <td>0.004671</td>\n",
       "      <td>pDCs</td>\n",
       "      <td>30</td>\n",
       "      <td>0.055288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1124701</th>\n",
       "      <td>lnc1686</td>\n",
       "      <td>0.496732</td>\n",
       "      <td>3.818752</td>\n",
       "      <td>4.487664</td>\n",
       "      <td>0.850944</td>\n",
       "      <td>0.394800</td>\n",
       "      <td>0.614526</td>\n",
       "      <td>pDCs</td>\n",
       "      <td>30</td>\n",
       "      <td>14.111033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1124702</th>\n",
       "      <td>lnc2311</td>\n",
       "      <td>2.592510</td>\n",
       "      <td>-4.979600</td>\n",
       "      <td>1.133711</td>\n",
       "      <td>-4.392302</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>pDCs</td>\n",
       "      <td>30</td>\n",
       "      <td>0.031695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1124703</th>\n",
       "      <td>lnc5775</td>\n",
       "      <td>2.392779</td>\n",
       "      <td>-2.530937</td>\n",
       "      <td>1.193175</td>\n",
       "      <td>-2.121178</td>\n",
       "      <td>0.033907</td>\n",
       "      <td>0.115058</td>\n",
       "      <td>pDCs</td>\n",
       "      <td>30</td>\n",
       "      <td>0.173026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1124704 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene     baseMean  log2FoldChange     lfcSE      stat    pvalue  \\\n",
       "0        lnc31752  5632.003911        0.060929  0.161392  0.377519  0.705788   \n",
       "1         lnc4822  3146.387654       -0.368362  0.385441 -0.955691  0.339228   \n",
       "2           Foxp1   941.129495       -0.079404  0.359311 -0.220991  0.825099   \n",
       "3            Ebf1   616.758548       -0.548443  0.564479 -0.971590  0.331254   \n",
       "4            Dpyd   924.659365        0.007061  0.575075  0.012278  0.990204   \n",
       "...           ...          ...             ...       ...       ...       ...   \n",
       "1124699   lnc1191     1.467205        0.158104  1.308503  0.120828  0.903827   \n",
       "1124700   lnc1440     2.409755       -4.176886  1.226191 -3.406392  0.000658   \n",
       "1124701   lnc1686     0.496732        3.818752  4.487664  0.850944  0.394800   \n",
       "1124702   lnc2311     2.592510       -4.979600  1.133711 -4.392302  0.000011   \n",
       "1124703   lnc5775     2.392779       -2.530937  1.193175 -2.121178  0.033907   \n",
       "\n",
       "             padj Celltype  Dose  Fold-Change  \n",
       "0        1.000000  B Cells  0.01     1.043137  \n",
       "1        1.000000  B Cells  0.01     0.774662  \n",
       "2        1.000000  B Cells  0.01     0.946448  \n",
       "3        1.000000  B Cells  0.01     0.683758  \n",
       "4        1.000000  B Cells  0.01     1.004906  \n",
       "...           ...      ...   ...          ...  \n",
       "1124699  0.955543     pDCs    30     1.115820  \n",
       "1124700  0.004671     pDCs    30     0.055288  \n",
       "1124701  0.614526     pDCs    30    14.111033  \n",
       "1124702  0.000141     pDCs    30     0.031695  \n",
       "1124703  0.115058     pDCs    30     0.173026  \n",
       "\n",
       "[1124704 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename 'index' to 'Gene' and compute fold change from log2FoldChange\n",
    "Deseq2_Master = Deseq2_Master.rename(columns={'index': 'Gene'})\n",
    "Deseq2_Master['Fold-Change'] = pow(2, Deseq2_Master['log2FoldChange'])\n",
    "\n",
    "\n",
    "Deseq2_Master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>blank</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Genebody</th>\n",
       "      <th>Exon</th>\n",
       "      <th>Frame</th>\n",
       "      <th>mgi_chr</th>\n",
       "      <th>...</th>\n",
       "      <th>mgi_end</th>\n",
       "      <th>mgi_symbol</th>\n",
       "      <th>mgi_score</th>\n",
       "      <th>mgi_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Type</th>\n",
       "      <th>NR_Annotation</th>\n",
       "      <th>NR_annotation2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3224554</td>\n",
       "      <td>3224816</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>lnc_inter_chr1_15560</td>\n",
       "      <td>lnc_inter_chr1_15560</td>\n",
       "      <td>lnc15560</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc15560</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3273138</td>\n",
       "      <td>3274559</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>lnc_inter_chr1_15561</td>\n",
       "      <td>lnc_inter_chr1_15561</td>\n",
       "      <td>lnc15561</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc15561</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3275710</td>\n",
       "      <td>3276991</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>lnc_inter_chr1_15562</td>\n",
       "      <td>lnc_inter_chr1_15562</td>\n",
       "      <td>lnc15562</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc15562</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3284704</td>\n",
       "      <td>3741721</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>GB_RefseqNM19801</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>...</td>\n",
       "      <td>3729127</td>\n",
       "      <td>Gm19938</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>NM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3536809</td>\n",
       "      <td>3583776</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>EnsmblNR4698</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>...</td>\n",
       "      <td>3583776</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84244</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90614769</td>\n",
       "      <td>90617133</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>EnsmblNR4698</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>90617133</td>\n",
       "      <td>Gm28300</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Gm28300</td>\n",
       "      <td>Gm28300</td>\n",
       "      <td>Gm28300</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84245</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90676614</td>\n",
       "      <td>90678894</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>EnsmblNR4698</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>90678894</td>\n",
       "      <td>Gm28301</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Gm28301</td>\n",
       "      <td>Gm28301</td>\n",
       "      <td>Gm28301</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84246</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90762408</td>\n",
       "      <td>90766319</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>GB_RefseqNR2077</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>90766736</td>\n",
       "      <td>G530011O06Riky</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>G530011O06Rik@chrY(-)</td>\n",
       "      <td>G530011O06Rik@chrY(-)</td>\n",
       "      <td>G530011O06Rik@chrY(-)</td>\n",
       "      <td>NR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84247</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90796710</td>\n",
       "      <td>90827734</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>GB_RefseqNM#GB_RefseqNR1159</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>90827734</td>\n",
       "      <td>Erdr1y</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>NM#NR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84248</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90796710</td>\n",
       "      <td>90827734</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>GB_RefseqNM#GB_RefseqNR1159</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>90820875</td>\n",
       "      <td>Gm33272</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>NM#NR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84249 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr     Start       End blank Score Strand  \\\n",
       "0      chr1   3224554   3224816     .  1000      -   \n",
       "1      chr1   3273138   3274559     .  1000      -   \n",
       "2      chr1   3275710   3276991     .  1000      -   \n",
       "3      chr1   3284704   3741721     .  1000      -   \n",
       "4      chr1   3536809   3583776     .  1000      +   \n",
       "...     ...       ...       ...   ...   ...    ...   \n",
       "84244  chrY  90614769  90617133     .  1000      -   \n",
       "84245  chrY  90676614  90678894     .  1000      -   \n",
       "84246  chrY  90762408  90766319     .  1000      -   \n",
       "84247  chrY  90796710  90827734     .  1000      +   \n",
       "84248  chrY  90796710  90827734     .  1000      +   \n",
       "\n",
       "                          Genebody  Exon Frame mgi_chr  ...   mgi_end  \\\n",
       "0           Genebody_mm10_lnc48261  exon     .       .  ...        -1   \n",
       "1           Genebody_mm10_lnc48261  exon     .       .  ...        -1   \n",
       "2           Genebody_mm10_lnc48261  exon     .       .  ...        -1   \n",
       "3                 GB_RefseqNM19801  exon     .    chr1  ...   3729127   \n",
       "4                     EnsmblNR4698  exon     .    chr1  ...   3583776   \n",
       "...                            ...   ...   ...     ...  ...       ...   \n",
       "84244                 EnsmblNR4698  exon     .    chrY  ...  90617133   \n",
       "84245                 EnsmblNR4698  exon     .    chrY  ...  90678894   \n",
       "84246              GB_RefseqNR2077  exon     .    chrY  ...  90766736   \n",
       "84247  GB_RefseqNM#GB_RefseqNR1159  exon     .    chrY  ...  90827734   \n",
       "84248  GB_RefseqNM#GB_RefseqNR1159  exon     .    chrY  ...  90820875   \n",
       "\n",
       "           mgi_symbol mgi_score  mgi_strand                gene_id  \\\n",
       "0                   .        -1           .   lnc_inter_chr1_15560   \n",
       "1                   .        -1           .   lnc_inter_chr1_15561   \n",
       "2                   .        -1           .   lnc_inter_chr1_15562   \n",
       "3             Gm19938      1000           -                   Xkr4   \n",
       "4              Gm1992      1000           +                 Gm1992   \n",
       "...               ...       ...         ...                    ...   \n",
       "84244         Gm28300      1000           -                Gm28300   \n",
       "84245         Gm28301      1000           -                Gm28301   \n",
       "84246  G530011O06Riky      1000           -  G530011O06Rik@chrY(-)   \n",
       "84247          Erdr1y      1000           +                  Erdr1   \n",
       "84248         Gm33272      1000           +                  Erdr1   \n",
       "\n",
       "               transcript_id                   Gene       Type NR_Annotation  \\\n",
       "0       lnc_inter_chr1_15560               lnc15560     lncRNA      lnc15560   \n",
       "1       lnc_inter_chr1_15561               lnc15561     lncRNA      lnc15561   \n",
       "2       lnc_inter_chr1_15562               lnc15562     lncRNA      lnc15562   \n",
       "3                       Xkr4                   Xkr4         NM           NaN   \n",
       "4                     Gm1992                 Gm1992  antisense           NaN   \n",
       "...                      ...                    ...        ...           ...   \n",
       "84244                Gm28300                Gm28300    lincRNA           NaN   \n",
       "84245                Gm28301                Gm28301    lincRNA           NaN   \n",
       "84246  G530011O06Rik@chrY(-)  G530011O06Rik@chrY(-)         NR           NaN   \n",
       "84247                  Erdr1                  Erdr1      NM#NR           NaN   \n",
       "84248                  Erdr1                  Erdr1      NM#NR           NaN   \n",
       "\n",
       "      NR_annotation2  \n",
       "0                NaN  \n",
       "1                NaN  \n",
       "2                NaN  \n",
       "3                NaN  \n",
       "4                NaN  \n",
       "...              ...  \n",
       "84244            NaN  \n",
       "84245            NaN  \n",
       "84246            NaN  \n",
       "84247            NaN  \n",
       "84248            NaN  \n",
       "\n",
       "[84249 rows x 21 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Gene_Type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>MGI_Symbol</th>\n",
       "      <th>Celltype</th>\n",
       "      <th>Dose</th>\n",
       "      <th>Type</th>\n",
       "      <th>pDRE</th>\n",
       "      <th>AHR</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>NR_Annotation</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>Fold-Change</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc31752</td>\n",
       "      <td>Malat1</td>\n",
       "      <td>B Cells</td>\n",
       "      <td>0.01</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>5632.003911</td>\n",
       "      <td>chr19</td>\n",
       "      <td>5842642</td>\n",
       "      <td>5852983</td>\n",
       "      <td>-</td>\n",
       "      <td>lnc31752_Mascrna#Malat1</td>\n",
       "      <td>0.060929</td>\n",
       "      <td>1.043137</td>\n",
       "      <td>0.161392</td>\n",
       "      <td>0.377519</td>\n",
       "      <td>0.705788</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc4822</td>\n",
       "      <td>.</td>\n",
       "      <td>B Cells</td>\n",
       "      <td>0.01</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3146.387654</td>\n",
       "      <td>chr6</td>\n",
       "      <td>3181101</td>\n",
       "      <td>3202501</td>\n",
       "      <td>+</td>\n",
       "      <td>lnc4822</td>\n",
       "      <td>-0.368362</td>\n",
       "      <td>0.774662</td>\n",
       "      <td>0.385441</td>\n",
       "      <td>-0.955691</td>\n",
       "      <td>0.339228</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Foxp1</td>\n",
       "      <td>9130401L11Rik</td>\n",
       "      <td>B Cells</td>\n",
       "      <td>0.01</td>\n",
       "      <td>NM</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>941.129495</td>\n",
       "      <td>chr6</td>\n",
       "      <td>98902302</td>\n",
       "      <td>99161922</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>-0.079404</td>\n",
       "      <td>0.946448</td>\n",
       "      <td>0.359311</td>\n",
       "      <td>-0.220991</td>\n",
       "      <td>0.825099</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Foxp1</td>\n",
       "      <td>6030492E11Rik</td>\n",
       "      <td>B Cells</td>\n",
       "      <td>0.01</td>\n",
       "      <td>NM</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>941.129495</td>\n",
       "      <td>chr6</td>\n",
       "      <td>98902302</td>\n",
       "      <td>99161922</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>-0.079404</td>\n",
       "      <td>0.946448</td>\n",
       "      <td>0.359311</td>\n",
       "      <td>-0.220991</td>\n",
       "      <td>0.825099</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Foxp1</td>\n",
       "      <td>Gm38865</td>\n",
       "      <td>B Cells</td>\n",
       "      <td>0.01</td>\n",
       "      <td>NM</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>941.129495</td>\n",
       "      <td>chr6</td>\n",
       "      <td>98902302</td>\n",
       "      <td>99161922</td>\n",
       "      <td>-</td>\n",
       "      <td>NA</td>\n",
       "      <td>-0.079404</td>\n",
       "      <td>0.946448</td>\n",
       "      <td>0.359311</td>\n",
       "      <td>-0.220991</td>\n",
       "      <td>0.825099</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1443131</th>\n",
       "      <td>lnc1440</td>\n",
       "      <td>.</td>\n",
       "      <td>pDCs</td>\n",
       "      <td>30</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.409755</td>\n",
       "      <td>chr2</td>\n",
       "      <td>69058214</td>\n",
       "      <td>69067863</td>\n",
       "      <td>-</td>\n",
       "      <td>lnc1440</td>\n",
       "      <td>-4.176886</td>\n",
       "      <td>0.055288</td>\n",
       "      <td>1.226191</td>\n",
       "      <td>-3.406392</td>\n",
       "      <td>0.000658</td>\n",
       "      <td>0.004671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1443132</th>\n",
       "      <td>lnc1686</td>\n",
       "      <td>Gm34484</td>\n",
       "      <td>pDCs</td>\n",
       "      <td>30</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.496732</td>\n",
       "      <td>chr2</td>\n",
       "      <td>124519234</td>\n",
       "      <td>124553367</td>\n",
       "      <td>-</td>\n",
       "      <td>lnc1686</td>\n",
       "      <td>3.818752</td>\n",
       "      <td>14.111033</td>\n",
       "      <td>4.487664</td>\n",
       "      <td>0.850944</td>\n",
       "      <td>0.3948</td>\n",
       "      <td>0.614526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1443133</th>\n",
       "      <td>lnc2311</td>\n",
       "      <td>Gm40040</td>\n",
       "      <td>pDCs</td>\n",
       "      <td>30</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2.592510</td>\n",
       "      <td>chr3</td>\n",
       "      <td>41395983</td>\n",
       "      <td>41447665</td>\n",
       "      <td>-</td>\n",
       "      <td>lnc2311_Platr4Gm40040</td>\n",
       "      <td>-4.979600</td>\n",
       "      <td>0.031695</td>\n",
       "      <td>1.133711</td>\n",
       "      <td>-4.392302</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.000141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1443134</th>\n",
       "      <td>lnc2311</td>\n",
       "      <td>Platr4</td>\n",
       "      <td>pDCs</td>\n",
       "      <td>30</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2.592510</td>\n",
       "      <td>chr3</td>\n",
       "      <td>41395983</td>\n",
       "      <td>41447665</td>\n",
       "      <td>-</td>\n",
       "      <td>lnc2311_Platr4Gm40040</td>\n",
       "      <td>-4.979600</td>\n",
       "      <td>0.031695</td>\n",
       "      <td>1.133711</td>\n",
       "      <td>-4.392302</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.000141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1443135</th>\n",
       "      <td>lnc5775</td>\n",
       "      <td>.</td>\n",
       "      <td>pDCs</td>\n",
       "      <td>30</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2.392779</td>\n",
       "      <td>chr6</td>\n",
       "      <td>144320563</td>\n",
       "      <td>144375764</td>\n",
       "      <td>-</td>\n",
       "      <td>lnc5775</td>\n",
       "      <td>-2.530937</td>\n",
       "      <td>0.173026</td>\n",
       "      <td>1.193175</td>\n",
       "      <td>-2.121178</td>\n",
       "      <td>0.033907</td>\n",
       "      <td>0.115058</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1443136 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene     MGI_Symbol Celltype  Dose    Type  pDRE  AHR  \\\n",
       "0        lnc31752         Malat1  B Cells  0.01  lncRNA     0    2   \n",
       "1         lnc4822              .  B Cells  0.01  lncRNA     0    1   \n",
       "2           Foxp1  9130401L11Rik  B Cells  0.01      NM    13   19   \n",
       "3           Foxp1  6030492E11Rik  B Cells  0.01      NM    13   19   \n",
       "4           Foxp1        Gm38865  B Cells  0.01      NM    13   19   \n",
       "...           ...            ...      ...   ...     ...   ...  ...   \n",
       "1443131   lnc1440              .     pDCs    30  lncRNA     0    0   \n",
       "1443132   lnc1686        Gm34484     pDCs    30  lncRNA     0    1   \n",
       "1443133   lnc2311        Gm40040     pDCs    30  lncRNA     3    3   \n",
       "1443134   lnc2311         Platr4     pDCs    30  lncRNA     3    3   \n",
       "1443135   lnc5775              .     pDCs    30  lncRNA     4    0   \n",
       "\n",
       "            baseMean    Chr      Start        End Strand  \\\n",
       "0        5632.003911  chr19    5842642    5852983      -   \n",
       "1        3146.387654   chr6    3181101    3202501      +   \n",
       "2         941.129495   chr6   98902302   99161922      -   \n",
       "3         941.129495   chr6   98902302   99161922      -   \n",
       "4         941.129495   chr6   98902302   99161922      -   \n",
       "...              ...    ...        ...        ...    ...   \n",
       "1443131     2.409755   chr2   69058214   69067863      -   \n",
       "1443132     0.496732   chr2  124519234  124553367      -   \n",
       "1443133     2.592510   chr3   41395983   41447665      -   \n",
       "1443134     2.592510   chr3   41395983   41447665      -   \n",
       "1443135     2.392779   chr6  144320563  144375764      -   \n",
       "\n",
       "                   NR_Annotation  log2FoldChange  Fold-Change     lfcSE  \\\n",
       "0        lnc31752_Mascrna#Malat1        0.060929     1.043137  0.161392   \n",
       "1                        lnc4822       -0.368362     0.774662  0.385441   \n",
       "2                             NA       -0.079404     0.946448  0.359311   \n",
       "3                             NA       -0.079404     0.946448  0.359311   \n",
       "4                             NA       -0.079404     0.946448  0.359311   \n",
       "...                          ...             ...          ...       ...   \n",
       "1443131                  lnc1440       -4.176886     0.055288  1.226191   \n",
       "1443132                  lnc1686        3.818752    14.111033  4.487664   \n",
       "1443133    lnc2311_Platr4Gm40040       -4.979600     0.031695  1.133711   \n",
       "1443134    lnc2311_Platr4Gm40040       -4.979600     0.031695  1.133711   \n",
       "1443135                  lnc5775       -2.530937     0.173026  1.193175   \n",
       "\n",
       "             stat    pvalue      padj  \n",
       "0        0.377519  0.705788       1.0  \n",
       "1       -0.955691  0.339228       1.0  \n",
       "2       -0.220991  0.825099       1.0  \n",
       "3       -0.220991  0.825099       1.0  \n",
       "4       -0.220991  0.825099       1.0  \n",
       "...           ...       ...       ...  \n",
       "1443131 -3.406392  0.000658  0.004671  \n",
       "1443132  0.850944    0.3948  0.614526  \n",
       "1443133 -4.392302  0.000011  0.000141  \n",
       "1443134 -4.392302  0.000011  0.000141  \n",
       "1443135 -2.121178  0.033907  0.115058  \n",
       "\n",
       "[1443136 rows x 19 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge Deseq2_Master with all metadata tables\n",
    "Deseq2_Master2 = pd.merge(Deseq2_Master, Gene_Type, left_on='Gene', right_on='Gene', how='left').fillna(\"NA\")\n",
    "Deseq2_Master3 = pd.merge(Deseq2_Master2, AHR_Master, left_on='Gene', right_on='Gene', how='left').fillna(0)\n",
    "Deseq2_Master4 = pd.merge(Deseq2_Master3, pDREs_Master, left_on='Gene', right_on='Gene', how='left').fillna(0)\n",
    "\n",
    "Deseq2_Master4[['pDRE', 'AHR']] = Deseq2_Master4[['pDRE', 'AHR']].astype(int)\n",
    "\n",
    "Deseq2_Master_Working = Deseq2_Master4[['Gene', 'MGI_Symbol', 'Celltype', 'Dose','Type', 'pDRE', 'AHR',  'baseMean',\n",
    "                                        'Chr','Start','End', 'Strand',  'NR_Annotation',\n",
    "                                      'log2FoldChange', 'Fold-Change', 'lfcSE', 'stat', 'pvalue', 'padj']]\n",
    "Deseq2_Master_Working\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "Deseq2_Master_Working.to_csv('./Results/03d_MASTER_Deseq2_Concatonated/RDDR_Deseq2_Master_Wald_with_MGI.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "SCVI New",
   "language": "python",
   "name": "scvi-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
