{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Concatonate All Deseq2 Data with Karri RefGenome Metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 1. Import Required Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 2. Import Deseq2 Data and All Other Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>pDRE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Spire1</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Wwox</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Apbb2</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sdk1</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Apoc2</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene  pDRE\n",
       "1  Spire1    75\n",
       "2    Wwox    47\n",
       "3   Apbb2    43\n",
       "4    Sdk1    38\n",
       "5   Apoc2    37"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pDREs_Master = pd.read_csv('../../99_mm39_DREs/02_CleanUp_DRE_BED_File/DRE_Counts_by_Gene.txt', \n",
    "                            delimiter = '\\t')\n",
    "\n",
    "pDREs_Master = pDREs_Master.dropna()\n",
    "\n",
    "pDREs_Master = pDREs_Master.rename(columns={'gene_name': 'Gene', 'count': 'pDRE'})\n",
    "\n",
    "pDREs_Master.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>AHR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc10851</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc7308</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hjurp</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc3293</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene   AHR\n",
       "0       NaN  2652\n",
       "1  lnc10851   153\n",
       "2   lnc7308    51\n",
       "3     Hjurp    44\n",
       "4   lnc3293    43"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AHR_Master = pd.read_csv('../../99_mm39_AHR_Enrichment/02_CleanUp_BED_File/AHR_Binding_Counts.txt', \n",
    "                            delimiter = '\\t')\n",
    "\n",
    "AHR_Master = AHR_Master.rename(columns={'gene_name': 'Gene', 'count': 'AHR'})\n",
    "\n",
    "AHR_Master.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/local/51388227/ipykernel_3251170/4085587502.py:1: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  Annotations_Master = pd.read_csv('../../00_Karri_et_al_GTF/MGI_and_Karri_Annotations_Restructured_mm39.txt',\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>blank</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>frame</th>\n",
       "      <th>mgi_chr</th>\n",
       "      <th>...</th>\n",
       "      <th>mgi_end</th>\n",
       "      <th>MGI_Symbol</th>\n",
       "      <th>mgi_score</th>\n",
       "      <th>mgi_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Type</th>\n",
       "      <th>NR_annotation</th>\n",
       "      <th>NR_annotation2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3224554</td>\n",
       "      <td>3224816</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>lnc_inter_chr1_15560</td>\n",
       "      <td>lnc_inter_chr1_15560</td>\n",
       "      <td>lnc15560</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc15560</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3273138</td>\n",
       "      <td>3274559</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>lnc_inter_chr1_15561</td>\n",
       "      <td>lnc_inter_chr1_15561</td>\n",
       "      <td>lnc15561</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc15561</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3275710</td>\n",
       "      <td>3276991</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>lnc_inter_chr1_15562</td>\n",
       "      <td>lnc_inter_chr1_15562</td>\n",
       "      <td>lnc15562</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc15562</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3284704</td>\n",
       "      <td>3741721</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>GB_RefseqNM19801</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>...</td>\n",
       "      <td>3729127</td>\n",
       "      <td>Gm19938</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>NM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3536809</td>\n",
       "      <td>3583776</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>EnsmblNR4698</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>...</td>\n",
       "      <td>3583776</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    start      end blank score strand                  source feature  \\\n",
       "0  chr1  3224554  3224816     .  1000      -  Genebody_mm10_lnc48261    exon   \n",
       "1  chr1  3273138  3274559     .  1000      -  Genebody_mm10_lnc48261    exon   \n",
       "2  chr1  3275710  3276991     .  1000      -  Genebody_mm10_lnc48261    exon   \n",
       "3  chr1  3284704  3741721     .  1000      -        GB_RefseqNM19801    exon   \n",
       "4  chr1  3536809  3583776     .  1000      +            EnsmblNR4698    exon   \n",
       "\n",
       "  frame mgi_chr  ...  mgi_end  MGI_Symbol mgi_score  mgi_strand  \\\n",
       "0     .       .  ...       -1           .        -1           .   \n",
       "1     .       .  ...       -1           .        -1           .   \n",
       "2     .       .  ...       -1           .        -1           .   \n",
       "3     .    chr1  ...  3729127     Gm19938      1000           -   \n",
       "4     .    chr1  ...  3583776      Gm1992      1000           +   \n",
       "\n",
       "                gene_id         transcript_id gene_name       Type  \\\n",
       "0  lnc_inter_chr1_15560  lnc_inter_chr1_15560  lnc15560     lncRNA   \n",
       "1  lnc_inter_chr1_15561  lnc_inter_chr1_15561  lnc15561     lncRNA   \n",
       "2  lnc_inter_chr1_15562  lnc_inter_chr1_15562  lnc15562     lncRNA   \n",
       "3                  Xkr4                  Xkr4      Xkr4         NM   \n",
       "4                Gm1992                Gm1992    Gm1992  antisense   \n",
       "\n",
       "  NR_annotation NR_annotation2  \n",
       "0      lnc15560            NaN  \n",
       "1      lnc15561            NaN  \n",
       "2      lnc15562            NaN  \n",
       "3           NaN            NaN  \n",
       "4           NaN            NaN  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Annotations_Master = pd.read_csv('../../00_Karri_et_al_GTF/MGI_and_Karri_Annotations_Restructured_mm39.txt', \n",
    "                                              delimiter = '\\t')\n",
    "\n",
    "Annotations_Master = Annotations_Master.rename(columns={'mgi_symbol': 'MGI_Symbol',\n",
    "                                                       'gene_biotype': 'Type'})\n",
    "\n",
    "Annotations_Master.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = './Deseq2_Results/'\n",
    "\n",
    "Deseq2_0_03_v_0 = pd.read_csv(directory + '0.03_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_0_1_v_0 = pd.read_csv(directory + '0.1_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_0_3_v_0 = pd.read_csv(directory + '0.3_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_1_v_0 = pd.read_csv(directory + '1_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_3_v_0 = pd.read_csv(directory + '3_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_10_v_0 = pd.read_csv(directory + '10_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_30_v_0 = pd.read_csv(directory + '30_vs_0.txt', delimiter = '\\t').reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "Deseq2_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_1_v_0['Dose'] = '1'\n",
    "Deseq2_3_v_0['Dose'] = '3'\n",
    "Deseq2_10_v_0['Dose'] = '10'\n",
    "Deseq2_30_v_0['Dose'] = '30'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Dose</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc_inter_chr1_100</td>\n",
       "      <td>0.738981</td>\n",
       "      <td>0.246277</td>\n",
       "      <td>2.463530</td>\n",
       "      <td>0.099969</td>\n",
       "      <td>0.920369</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc_inter_chr1_1000</td>\n",
       "      <td>0.821042</td>\n",
       "      <td>-0.262164</td>\n",
       "      <td>2.064551</td>\n",
       "      <td>-0.126984</td>\n",
       "      <td>0.898953</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>0.300701</td>\n",
       "      <td>1.832959</td>\n",
       "      <td>5.629859</td>\n",
       "      <td>0.325578</td>\n",
       "      <td>0.744744</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc_as_chr11_10004</td>\n",
       "      <td>2.027109</td>\n",
       "      <td>3.473974</td>\n",
       "      <td>1.722713</td>\n",
       "      <td>2.016571</td>\n",
       "      <td>0.043740</td>\n",
       "      <td>0.141112</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_inter_chr11_10005</td>\n",
       "      <td>8.027321</td>\n",
       "      <td>-0.202391</td>\n",
       "      <td>0.591410</td>\n",
       "      <td>-0.342218</td>\n",
       "      <td>0.732187</td>\n",
       "      <td>0.868661</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   index  baseMean  log2FoldChange     lfcSE      stat  \\\n",
       "0     lnc_inter_chr1_100  0.738981        0.246277  2.463530  0.099969   \n",
       "1    lnc_inter_chr1_1000  0.821042       -0.262164  2.064551 -0.126984   \n",
       "2     lnc_as_chr11_10003  0.300701        1.832959  5.629859  0.325578   \n",
       "3     lnc_as_chr11_10004  2.027109        3.473974  1.722713  2.016571   \n",
       "4  lnc_inter_chr11_10005  8.027321       -0.202391  0.591410 -0.342218   \n",
       "\n",
       "     pvalue      padj Dose  \n",
       "0  0.920369       NaN   10  \n",
       "1  0.898953       NaN   10  \n",
       "2  0.744744       NaN   10  \n",
       "3  0.043740  0.141112   10  \n",
       "4  0.732187  0.868661   10  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Deseq2_10_v_0.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 3. Concat All Data Into One Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "Deseq2_Master = pd.concat([Deseq2_0_03_v_0,\n",
    "                          Deseq2_0_1_v_0,\n",
    "                          Deseq2_0_3_v_0,\n",
    "                          Deseq2_1_v_0,\n",
    "                          Deseq2_3_v_0,\n",
    "                          Deseq2_10_v_0,\n",
    "                          Deseq2_30_v_0],\n",
    "                          ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Dose</th>\n",
       "      <th>FoldChange</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc_inter_chr1_100</td>\n",
       "      <td>0.738981</td>\n",
       "      <td>0.862954</td>\n",
       "      <td>2.429333</td>\n",
       "      <td>0.355223</td>\n",
       "      <td>0.722423</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.818759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc_inter_chr1_1000</td>\n",
       "      <td>0.821042</td>\n",
       "      <td>0.929693</td>\n",
       "      <td>1.989028</td>\n",
       "      <td>0.467411</td>\n",
       "      <td>0.640206</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.904871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>0.300701</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.656128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc_as_chr11_10004</td>\n",
       "      <td>2.027109</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.904898</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_inter_chr11_10005</td>\n",
       "      <td>8.027321</td>\n",
       "      <td>-0.269934</td>\n",
       "      <td>0.603792</td>\n",
       "      <td>-0.447065</td>\n",
       "      <td>0.654828</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.829357</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene  baseMean  log2FoldChange     lfcSE      stat  \\\n",
       "0     lnc_inter_chr1_100  0.738981        0.862954  2.429333  0.355223   \n",
       "1    lnc_inter_chr1_1000  0.821042        0.929693  1.989028  0.467411   \n",
       "2     lnc_as_chr11_10003  0.300701        0.000000  5.656128  0.000000   \n",
       "3     lnc_as_chr11_10004  2.027109        0.000000  1.904898  0.000000   \n",
       "4  lnc_inter_chr11_10005  8.027321       -0.269934  0.603792 -0.447065   \n",
       "\n",
       "     pvalue  padj  Dose  FoldChange  \n",
       "0  0.722423   NaN  0.03    1.818759  \n",
       "1  0.640206   NaN  0.03    1.904871  \n",
       "2  1.000000   NaN  0.03    1.000000  \n",
       "3  1.000000   1.0  0.03    1.000000  \n",
       "4  0.654828   1.0  0.03    0.829357  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Deseq2_Master = Deseq2_Master.rename(columns={'index': 'Gene'})\n",
    "Deseq2_Master['FoldChange'] = pow(2, Deseq2_Master['log2FoldChange'])\n",
    "\n",
    "Deseq2_Master.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Dose</th>\n",
       "      <th>FoldChange</th>\n",
       "      <th>chrom</th>\n",
       "      <th>...</th>\n",
       "      <th>mgi_end</th>\n",
       "      <th>MGI_Symbol</th>\n",
       "      <th>mgi_score</th>\n",
       "      <th>mgi_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Type</th>\n",
       "      <th>NR_annotation</th>\n",
       "      <th>NR_annotation2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc_inter_chr1_100</td>\n",
       "      <td>0.738981</td>\n",
       "      <td>0.862954</td>\n",
       "      <td>2.429333</td>\n",
       "      <td>0.355223</td>\n",
       "      <td>0.722423</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.818759</td>\n",
       "      <td>chr1</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>lnc_inter_chr1_100</td>\n",
       "      <td>lnc_inter_chr1_100</td>\n",
       "      <td>lnc100</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc100</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc_inter_chr1_1000</td>\n",
       "      <td>0.821042</td>\n",
       "      <td>0.929693</td>\n",
       "      <td>1.989028</td>\n",
       "      <td>0.467411</td>\n",
       "      <td>0.640206</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.904871</td>\n",
       "      <td>chr1</td>\n",
       "      <td>...</td>\n",
       "      <td>191154841</td>\n",
       "      <td>Gm32128</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>lnc_inter_chr1_1000</td>\n",
       "      <td>lnc_inter_chr1_1000</td>\n",
       "      <td>lnc1000</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc1000</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>0.300701</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.656128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr11</td>\n",
       "      <td>...</td>\n",
       "      <td>94900160</td>\n",
       "      <td>Gm51875</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>lnc10003</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc10003</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>0.300701</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.656128</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr11</td>\n",
       "      <td>...</td>\n",
       "      <td>94900743</td>\n",
       "      <td>Gm36385</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>lnc10003</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc10003</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc_as_chr11_10004</td>\n",
       "      <td>2.027109</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.904898</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr11</td>\n",
       "      <td>...</td>\n",
       "      <td>95049873</td>\n",
       "      <td>Dlx4os</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>lnc_as_chr11_10004</td>\n",
       "      <td>lnc_as_chr11_10004</td>\n",
       "      <td>lnc10004</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc10004_Dlx4os</td>\n",
       "      <td>Dlx4os</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Gene  baseMean  log2FoldChange     lfcSE      stat  \\\n",
       "0   lnc_inter_chr1_100  0.738981        0.862954  2.429333  0.355223   \n",
       "1  lnc_inter_chr1_1000  0.821042        0.929693  1.989028  0.467411   \n",
       "2   lnc_as_chr11_10003  0.300701        0.000000  5.656128  0.000000   \n",
       "3   lnc_as_chr11_10003  0.300701        0.000000  5.656128  0.000000   \n",
       "4   lnc_as_chr11_10004  2.027109        0.000000  1.904898  0.000000   \n",
       "\n",
       "     pvalue padj  Dose  FoldChange  chrom  ...    mgi_end  MGI_Symbol  \\\n",
       "0  0.722423   NA  0.03    1.818759   chr1  ...         -1           .   \n",
       "1  0.640206   NA  0.03    1.904871   chr1  ...  191154841     Gm32128   \n",
       "2       1.0   NA  0.03    1.000000  chr11  ...   94900160     Gm51875   \n",
       "3       1.0   NA  0.03    1.000000  chr11  ...   94900743     Gm36385   \n",
       "4       1.0  1.0  0.03    1.000000  chr11  ...   95049873      Dlx4os   \n",
       "\n",
       "  mgi_score mgi_strand              gene_id        transcript_id gene_name  \\\n",
       "0        -1          .   lnc_inter_chr1_100   lnc_inter_chr1_100    lnc100   \n",
       "1      1000          +  lnc_inter_chr1_1000  lnc_inter_chr1_1000   lnc1000   \n",
       "2      1000          -   lnc_as_chr11_10003   lnc_as_chr11_10003  lnc10003   \n",
       "3      1000          -   lnc_as_chr11_10003   lnc_as_chr11_10003  lnc10003   \n",
       "4      1000          +   lnc_as_chr11_10004   lnc_as_chr11_10004  lnc10004   \n",
       "\n",
       "     Type    NR_annotation  NR_annotation2  \n",
       "0  lncRNA           lnc100              NA  \n",
       "1  lncRNA          lnc1000              NA  \n",
       "2  lncRNA         lnc10003              NA  \n",
       "3  lncRNA         lnc10003              NA  \n",
       "4  lncRNA  lnc10004_Dlx4os          Dlx4os  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge Deseq2_Master with all metadata tables\n",
    "Deseq2_Master2 = pd.merge(Deseq2_Master, Annotations_Master, left_on='Gene', right_on='gene_id', how='left').fillna(\"NA\")\n",
    "\n",
    "Deseq2_Master2.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>MGI_Symbol</th>\n",
       "      <th>Dose</th>\n",
       "      <th>pDRE</th>\n",
       "      <th>AHR</th>\n",
       "      <th>Type</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>Log2FC</th>\n",
       "      <th>Fold-Change</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>...</th>\n",
       "      <th>padj</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>frame</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>NR_annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc100</td>\n",
       "      <td>.</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.738981</td>\n",
       "      <td>0.862954</td>\n",
       "      <td>1.818759</td>\n",
       "      <td>2.429333</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>chr1</td>\n",
       "      <td>21246734</td>\n",
       "      <td>21250265</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>lnc_inter_chr1_100</td>\n",
       "      <td>lnc_inter_chr1_100</td>\n",
       "      <td>lnc100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc1000</td>\n",
       "      <td>Gm32128</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.821042</td>\n",
       "      <td>0.929693</td>\n",
       "      <td>1.904871</td>\n",
       "      <td>1.989028</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>chr1</td>\n",
       "      <td>191152395</td>\n",
       "      <td>191156841</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>lnc_inter_chr1_1000</td>\n",
       "      <td>lnc_inter_chr1_1000</td>\n",
       "      <td>lnc1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc10003</td>\n",
       "      <td>Gm51875</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.300701</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.656128</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>chr11</td>\n",
       "      <td>94897782</td>\n",
       "      <td>94903363</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>lnc10003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc10003</td>\n",
       "      <td>Gm36385</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.300701</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.656128</td>\n",
       "      <td>...</td>\n",
       "      <td>NA</td>\n",
       "      <td>chr11</td>\n",
       "      <td>94897782</td>\n",
       "      <td>94903363</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>lnc_as_chr11_10003</td>\n",
       "      <td>lnc10003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc10004</td>\n",
       "      <td>Dlx4os</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>2.027109</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.904898</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>chr11</td>\n",
       "      <td>95035841</td>\n",
       "      <td>95050060</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>lnc_as_chr11_10004</td>\n",
       "      <td>lnc_as_chr11_10004</td>\n",
       "      <td>lnc10004_Dlx4os</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365969</th>\n",
       "      <td>mt-Nd4l</td>\n",
       "      <td>.</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>mitochondrial protein-coding gene</td>\n",
       "      <td>28.894681</td>\n",
       "      <td>-0.353649</td>\n",
       "      <td>0.782602</td>\n",
       "      <td>0.350916</td>\n",
       "      <td>...</td>\n",
       "      <td>0.461948</td>\n",
       "      <td>chrM</td>\n",
       "      <td>9876</td>\n",
       "      <td>10173</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000065947</td>\n",
       "      <td>ENSMUSG00000065947</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365970</th>\n",
       "      <td>mt-Nd4</td>\n",
       "      <td>.</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>mitochondrial protein-coding gene</td>\n",
       "      <td>98368.928940</td>\n",
       "      <td>-0.549831</td>\n",
       "      <td>0.683100</td>\n",
       "      <td>0.082421</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>chrM</td>\n",
       "      <td>10166</td>\n",
       "      <td>11544</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000064363</td>\n",
       "      <td>ENSMUSG00000064363</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365971</th>\n",
       "      <td>mt-Nd5</td>\n",
       "      <td>.</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>mitochondrial protein-coding gene</td>\n",
       "      <td>67723.157378</td>\n",
       "      <td>-0.101532</td>\n",
       "      <td>0.932043</td>\n",
       "      <td>0.100825</td>\n",
       "      <td>...</td>\n",
       "      <td>0.462225</td>\n",
       "      <td>chrM</td>\n",
       "      <td>11741</td>\n",
       "      <td>13565</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000064367</td>\n",
       "      <td>ENSMUSG00000064367</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365972</th>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>.</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>mitochondrial protein-coding gene</td>\n",
       "      <td>19830.183491</td>\n",
       "      <td>0.154663</td>\n",
       "      <td>1.113162</td>\n",
       "      <td>0.185883</td>\n",
       "      <td>...</td>\n",
       "      <td>0.556118</td>\n",
       "      <td>chrM</td>\n",
       "      <td>13551</td>\n",
       "      <td>14070</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064368</td>\n",
       "      <td>ENSMUSG00000064368</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365973</th>\n",
       "      <td>mt-Cytb</td>\n",
       "      <td>.</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>mitochondrial protein-coding gene</td>\n",
       "      <td>169173.377707</td>\n",
       "      <td>-0.665665</td>\n",
       "      <td>0.630398</td>\n",
       "      <td>0.084196</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>chrM</td>\n",
       "      <td>14144</td>\n",
       "      <td>15288</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSMUSG00000064370</td>\n",
       "      <td>ENSMUSG00000064370</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>365974 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene MGI_Symbol  Dose  pDRE  AHR  \\\n",
       "0         lnc100          .  0.03     1    0   \n",
       "1        lnc1000    Gm32128  0.03     0    8   \n",
       "2       lnc10003    Gm51875  0.03     0    0   \n",
       "3       lnc10003    Gm36385  0.03     0    0   \n",
       "4       lnc10004     Dlx4os  0.03     1    0   \n",
       "...          ...        ...   ...   ...  ...   \n",
       "365969   mt-Nd4l          .    30     0    0   \n",
       "365970    mt-Nd4          .    30     0    0   \n",
       "365971    mt-Nd5          .    30     0    0   \n",
       "365972    mt-Nd6          .    30     0    0   \n",
       "365973   mt-Cytb          .    30     0    0   \n",
       "\n",
       "                                     Type       baseMean    Log2FC  \\\n",
       "0                                  lncRNA       0.738981  0.862954   \n",
       "1                                  lncRNA       0.821042  0.929693   \n",
       "2                                  lncRNA       0.300701  0.000000   \n",
       "3                                  lncRNA       0.300701  0.000000   \n",
       "4                                  lncRNA       2.027109  0.000000   \n",
       "...                                   ...            ...       ...   \n",
       "365969  mitochondrial protein-coding gene      28.894681 -0.353649   \n",
       "365970  mitochondrial protein-coding gene   98368.928940 -0.549831   \n",
       "365971  mitochondrial protein-coding gene   67723.157378 -0.101532   \n",
       "365972  mitochondrial protein-coding gene   19830.183491  0.154663   \n",
       "365973  mitochondrial protein-coding gene  169173.377707 -0.665665   \n",
       "\n",
       "        Fold-Change     lfcSE  ...      padj  chrom      start        end  \\\n",
       "0          1.818759  2.429333  ...        NA   chr1   21246734   21250265   \n",
       "1          1.904871  1.989028  ...        NA   chr1  191152395  191156841   \n",
       "2          1.000000  5.656128  ...        NA  chr11   94897782   94903363   \n",
       "3          1.000000  5.656128  ...        NA  chr11   94897782   94903363   \n",
       "4          1.000000  1.904898  ...       1.0  chr11   95035841   95050060   \n",
       "...             ...       ...  ...       ...    ...        ...        ...   \n",
       "365969     0.782602  0.350916  ...  0.461948   chrM       9876      10173   \n",
       "365970     0.683100  0.082421  ...       0.0   chrM      10166      11544   \n",
       "365971     0.932043  0.100825  ...  0.462225   chrM      11741      13565   \n",
       "365972     1.113162  0.185883  ...  0.556118   chrM      13551      14070   \n",
       "365973     0.630398  0.084196  ...       0.0   chrM      14144      15288   \n",
       "\n",
       "        frame  score strand              gene_id        transcript_id  \\\n",
       "0           .   1000      +   lnc_inter_chr1_100   lnc_inter_chr1_100   \n",
       "1           .   1000      +  lnc_inter_chr1_1000  lnc_inter_chr1_1000   \n",
       "2           .   1000      -   lnc_as_chr11_10003   lnc_as_chr11_10003   \n",
       "3           .   1000      -   lnc_as_chr11_10003   lnc_as_chr11_10003   \n",
       "4           .   1000      +   lnc_as_chr11_10004   lnc_as_chr11_10004   \n",
       "...       ...    ...    ...                  ...                  ...   \n",
       "365969      .      .      +   ENSMUSG00000065947   ENSMUSG00000065947   \n",
       "365970      .      .      +   ENSMUSG00000064363   ENSMUSG00000064363   \n",
       "365971      .      .      +   ENSMUSG00000064367   ENSMUSG00000064367   \n",
       "365972      .      .      -   ENSMUSG00000064368   ENSMUSG00000064368   \n",
       "365973      .      .      +   ENSMUSG00000064370   ENSMUSG00000064370   \n",
       "\n",
       "          NR_annotation  \n",
       "0                lnc100  \n",
       "1               lnc1000  \n",
       "2              lnc10003  \n",
       "3              lnc10003  \n",
       "4       lnc10004_Dlx4os  \n",
       "...                 ...  \n",
       "365969               NA  \n",
       "365970               NA  \n",
       "365971               NA  \n",
       "365972               NA  \n",
       "365973               NA  \n",
       "\n",
       "[365974 rows x 22 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop the 'Gene' column\n",
    "Deseq2_Master2 = Deseq2_Master2.drop(columns=['Gene'])\n",
    "\n",
    "Deseq2_Master2 = Deseq2_Master2.rename(columns={'gene_name': 'Gene'})\n",
    "Deseq2_Master2 = Deseq2_Master2.rename(columns={'FoldChange': 'Fold-Change'})\n",
    "Deseq2_Master2 = Deseq2_Master2.rename(columns={'log2FoldChange': 'Log2FC'})\n",
    "\n",
    "Deseq2_Master2 = pd.merge(Deseq2_Master2, pDREs_Master, left_on='Gene', right_on='Gene', how='left').fillna(0)\n",
    "Deseq2_Master2 = pd.merge(Deseq2_Master2, AHR_Master, left_on='Gene', right_on='Gene', how='left').fillna(0)\n",
    "\n",
    "Deseq2_Master2[['pDRE', 'AHR']] = Deseq2_Master2[['pDRE', 'AHR']].astype(int)\n",
    "\n",
    "Deseq2_Master_Working = Deseq2_Master2[['Gene', 'MGI_Symbol', 'Dose', 'pDRE', 'AHR', 'Type',  'baseMean',\n",
    "                                      'Log2FC', 'Fold-Change', 'lfcSE', 'stat', 'pvalue', 'padj',\n",
    "                                       'chrom', 'start', 'end', 'frame', 'score','strand',\n",
    "                                        'gene_id', 'transcript_id', 'NR_annotation']].drop_duplicates(keep='first')\n",
    "\n",
    "Deseq2_Master_Working"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "Deseq2_Master_Working.to_csv('./RDDR_Deseq2_Master_Wald_Bulk_Mouse_with_Mouse_MGI.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python (scvi-new)",
   "language": "python",
   "name": "scvi"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
