{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process GTF file from Karri et al. for mm10 to rn7 Liftover"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 1. Import Required Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 2. Liftover Karri et al. from mm10 to mm39"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/local/52526326/ipykernel_292824/1828373986.py:5: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  RAW_Master = pd.read_csv('./RAW_Data/GTFB_FullGeneBody_MouseLiver_snRNAseq.gtf',\n"
     ]
    }
   ],
   "source": [
    "# Define the column names\n",
    "columns = ['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attributes']\n",
    "\n",
    "# Read the GTF file into a DataFrame\n",
    "RAW_Master = pd.read_csv('./RAW_Data/GTFB_FullGeneBody_MouseLiver_snRNAseq.gtf', \n",
    "                         delimiter='\\t',\n",
    "                         header=None,  # No header in the file\n",
    "                         names=columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ID</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4470732</td>\n",
       "      <td>4471325</td>\n",
       "      <td>0</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4744316</td>\n",
       "      <td>4745480</td>\n",
       "      <td>1</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>21176511</td>\n",
       "      <td>21180041</td>\n",
       "      <td>2</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>191420199</td>\n",
       "      <td>191424644</td>\n",
       "      <td>3</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>94731060</td>\n",
       "      <td>94731724</td>\n",
       "      <td>4</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname      start        end  ID score strand\n",
       "0    chr1    4470732    4471325   0  1000      -\n",
       "1    chr1    4744316    4745480   1  1000      -\n",
       "2    chr1   21176511   21180041   2  1000      +\n",
       "3    chr1  191420199  191424644   3  1000      +\n",
       "4   chr11   94731060   94731724   4  1000      -"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract relevant columns\n",
    "bed_df = RAW_Master.copy()\n",
    "\n",
    "# Add the 'ID' column as the row index + 1 to start from 1\n",
    "bed_df['ID'] = bed_df.index\n",
    "\n",
    "# Add 'chr' to every seqname value\n",
    "bed_df['seqname'] = 'chr' + bed_df['seqname'].astype(str)\n",
    "\n",
    "# Reorder columns to place 'ID' between 'end' and 'score'\n",
    "bed_df_filtered = bed_df[['seqname', 'start', 'end', 'ID', 'score', 'strand']]\n",
    "\n",
    "# Define the list of standard chromosomes to keep\n",
    "standard_chromosomes = {f'chr{i}' for i in range(1, 20)} | {'chrX', 'chrY', 'chrMT'}\n",
    "\n",
    "# Filter the DataFrame\n",
    "bed_df_filtered = bed_df_filtered[bed_df_filtered['seqname'].isin(standard_chromosomes)]\n",
    "bed_df_filtered['seqname'] = bed_df_filtered['seqname'].replace({'chrMT': 'chrM'})\n",
    "\n",
    "\n",
    "bed_df_filtered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export to BED format (tab-delimited)\n",
    "bed_df_filtered.to_csv('./RAW_Data/mm10_Karri_Annotations_Restructured_For_Liftovers.bed', sep='\\t', \n",
    "              header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-03-25 14:20:04--  https://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/liftOver\n",
      "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
      "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 24388592 (23M)\n",
      "Saving to: ‘liftOver’\n",
      "\n",
      "liftOver            100%[===================>]  23.26M  10.1MB/s    in 2.3s    \n",
      "\n",
      "2025-03-25 14:20:07 (10.1 MB/s) - ‘liftOver’ saved [24388592/24388592]\n",
      "\n",
      "--2025-03-25 14:20:07--  https://hgdownload.soe.ucsc.edu/goldenPath/mm10/liftOver/mm10ToRn7.over.chain.gz\n",
      "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
      "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 76330380 (73M) [application/x-gzip]\n",
      "Saving to: ‘mm10ToRn7.over.chain.gz’\n",
      "\n",
      "mm10ToRn7.over.chai 100%[===================>]  72.79M  21.3MB/s    in 3.8s    \n",
      "\n",
      "2025-03-25 14:20:11 (19.2 MB/s) - ‘mm10ToRn7.over.chain.gz’ saved [76330380/76330380]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Download the UCSC LiftOver executable\n",
    "!wget -O liftOver https://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/liftOver\n",
    "\n",
    "# Make the file executable\n",
    "!chmod +x liftOver\n",
    "\n",
    "# Download mm10 to rn6 liftover chain\n",
    "!wget -O mm10ToRn7.over.chain.gz https://hgdownload.soe.ucsc.edu/goldenPath/mm10/liftOver/mm10ToRn7.over.chain.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading liftover chains\n",
      "Mapping coordinates\n"
     ]
    }
   ],
   "source": [
    "# Run UCSC LiftOver to map the coordinates\n",
    "# Define variables\n",
    "min_match = 0.7\n",
    "input_bed = \"./RAW_Data/mm10_Karri_Annotations_Restructured_For_Liftovers.bed\"\n",
    "chain_file = \"mm10ToRn7.over.chain.gz\"\n",
    "output_bed = f\"./RAW_Data/Karri_et_al_rn7_liftover_annotations_ratio_{min_match}.bed\"\n",
    "unmapped_bed = f\"./RAW_Data/unmapped_coordinates_ratio_{min_match}.bed\"\n",
    "\n",
    "# Construct and run the command\n",
    "!./liftOver -minMatch={min_match} {input_bed} {chain_file} {output_bed} {unmapped_bed}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cleanup directory\n",
    "!rm -r liftOver\n",
    "!rm -r mm10ToRn7.over.chain.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read LiftOver output (only successfully mapped entries)\n",
    "columns = ['rat_chr', 'rat_start', 'rat_end', 'mouse_ID', 'rat_score', 'rat_strand']\n",
    "\n",
    "liftover_df = pd.read_csv('./RAW_Data/Karri_et_al_rn7_liftover_annotations_ratio_'+ str(min_match) +'.bed', sep='\\t', header=None,\n",
    "                          names=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/local/52526326/ipykernel_292824/3571990473.py:5: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  Karri_Master = pd.read_csv('../00_Karri_et_al_GTF/RAW_Data/GTFB_FullGeneBody_MouseLiver_snRNAseq.gtf',\n"
     ]
    }
   ],
   "source": [
    "# Define the column names\n",
    "columns = ['mouse_chr', 'mouse_source', 'mouse_feature', 'mouse_start', 'mouse_end', 'mouse_score', 'mouse_strand', 'mouse_frame', 'mouse_attributes']\n",
    "\n",
    "# Read the GTF file into a DataFrame\n",
    "Karri_Master = pd.read_csv('../00_Karri_et_al_GTF/RAW_Data/GTFB_FullGeneBody_MouseLiver_snRNAseq.gtf', \n",
    "                         delimiter='\\t',\n",
    "                         header=None,  # No header in the file\n",
    "                         names=columns)\n",
    "\n",
    "# Add the 'ID' column as the row index + 1 to start from 1\n",
    "Karri_Master['mouse_ID'] = Karri_Master.index \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge Rat_Master and Karri_Master on 'ID' (right) and 'rat_ID' (left)\n",
    "merged_df = pd.merge(liftover_df, Karri_Master, left_on='mouse_ID', right_on='mouse_ID', how='left')\n",
    "merged_df_trimmed = merged_df[['rat_chr', 'rat_start', 'rat_end', 'mouse_ID', 'rat_score','rat_strand','mouse_attributes']]\n",
    "merged_df_trimmed = merged_df_trimmed.drop_duplicates(keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to parse attributes and extract key-value pairs\n",
    "def parse_attributes(attribute_string):\n",
    "    attr_dict = {}\n",
    "    if pd.notnull(attribute_string):  # Check if the string is not NaN\n",
    "        attributes = attribute_string.split(';')\n",
    "        for attribute in attributes:\n",
    "            if attribute.strip():  # Ensure it's not an empty string\n",
    "                key, value = attribute.strip().split(' ', 1)\n",
    "                attr_dict[key] = value.strip('\"')  # Strip quotes from value\n",
    "    return attr_dict\n",
    "\n",
    "# Apply the function to extract the information into a DataFrame\n",
    "attributes_df = merged_df_trimmed['mouse_attributes'].apply(parse_attributes).apply(pd.Series)\n",
    "\n",
    "# Drop the 'gene_id' column\n",
    "attributes_df = attributes_df.drop(columns=['gene_id', 'transcript_id', 'gene_name2', 'NR_annotation'])\n",
    "\n",
    "# Rename 'gene_name' to 'gene_id'\n",
    "attributes_df = attributes_df.rename(columns={'gene_name': 'gene_id'})\n",
    "attributes_df = attributes_df.fillna('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenate the two DataFrames side by side\n",
    "merged_combined = pd.concat([merged_df_trimmed, attributes_df], axis=1)\n",
    "\n",
    "# Define the values to keep\n",
    "biotypes_to_keep = ['lncRNA', 'NR', 'lincRNA', 'antisense', 'lncOfInterest']\n",
    "\n",
    "# Filter the DataFrame\n",
    "merged_combined_trimmed = merged_combined[merged_combined['gene_biotype'].isin(biotypes_to_keep)]\n",
    "#merged_combined_trimmed[['gene_biotype']] = 'lncRNA'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/local/52526326/ipykernel_292824/4117768066.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  merged_combined_trimmed['attribute'] = (\n",
      "/tmp/local/52526326/ipykernel_292824/4117768066.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  merged_combined_trimmed['source'] = 'liftover'\n",
      "/tmp/local/52526326/ipykernel_292824/4117768066.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  merged_combined_trimmed['feature'] = 'exon'\n",
      "/tmp/local/52526326/ipykernel_292824/4117768066.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  merged_combined_trimmed['frame'] = '.'\n"
     ]
    }
   ],
   "source": [
    "#Build GTF-style 'attribute' column with mouse and gene IDs\n",
    "merged_combined_trimmed['attribute'] = (\n",
    "    'Mouse_ID \"' + merged_combined_trimmed['mouse_ID'].astype(str) + \n",
    "    '\"; gene_id \"' + merged_combined_trimmed['gene_id'] + '\";'\n",
    ")\n",
    "\n",
    "\n",
    "#Set GTF-required columns\n",
    "merged_combined_trimmed['source'] = 'liftover'\n",
    "merged_combined_trimmed['feature'] = 'exon'\n",
    "merged_combined_trimmed['frame'] = '.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select and order columns for final LiftOver GTF output\n",
    "liftover_GTF = merged_combined_trimmed[['rat_chr', 'source', 'feature', \n",
    "                          'rat_start', 'rat_end', 'rat_score', \n",
    "                          'rat_strand', 'frame', 'attribute']]\n",
    "\n",
    "# Define the list of standard chromosomes to keep\n",
    "standard_chromosomes = {f'chr{i}' for i in range(1, 20)} | {'chrX', 'chrY', 'chrM'}\n",
    "\n",
    "# Filter the DataFrame\n",
    "liftover_GTF = liftover_GTF[liftover_GTF['rat_chr'].isin(standard_chromosomes)]\n",
    "liftover_GTF[['transcript_type']] = 'lncRNA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rat_chr</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>rat_start</th>\n",
       "      <th>rat_end</th>\n",
       "      <th>rat_score</th>\n",
       "      <th>rat_strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>15038462</td>\n",
       "      <td>15039082</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"0\"; gene_id \"lnc1\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>14766774</td>\n",
       "      <td>14767820</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"1\"; gene_id \"lnc10\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>103001424</td>\n",
       "      <td>103005924</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"3\"; gene_id \"lnc1000\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>79676628</td>\n",
       "      <td>79677257</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"4\"; gene_id \"lnc10000\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>79677716</td>\n",
       "      <td>79678362</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"5\"; gene_id \"lnc10001\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65892</th>\n",
       "      <td>chr3</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>155794311</td>\n",
       "      <td>155794889</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"82486\"; gene_id \"Zfas1\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66085</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>158072676</td>\n",
       "      <td>158072721</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"82711\"; gene_id \"Znf41-ps\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66086</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>157249657</td>\n",
       "      <td>157249708</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"82712\"; gene_id \"Znf41-ps\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66087</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>157912765</td>\n",
       "      <td>157912821</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"82713\"; gene_id \"Znf41-ps\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66088</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>157480498</td>\n",
       "      <td>157480624</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"82714\"; gene_id \"Znf41-ps\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41312 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      rat_chr    source feature  rat_start    rat_end rat_score rat_strand  \\\n",
       "0        chr5  liftover    exon   15038462   15039082      1000          +   \n",
       "1        chr5  liftover    exon   14766774   14767820      1000          +   \n",
       "2       chr13  liftover    exon  103001424  103005924      1000          +   \n",
       "3       chr10  liftover    exon   79676628   79677257      1000          -   \n",
       "4       chr10  liftover    exon   79677716   79678362      1000          -   \n",
       "...       ...       ...     ...        ...        ...       ...        ...   \n",
       "65892    chr3  liftover    exon  155794311  155794889      1000          +   \n",
       "66085    chr5  liftover    exon  158072676  158072721      1000          -   \n",
       "66086    chr5  liftover    exon  157249657  157249708      1000          +   \n",
       "66087    chr5  liftover    exon  157912765  157912821      1000          -   \n",
       "66088    chr5  liftover    exon  157480498  157480624      1000          +   \n",
       "\n",
       "      frame                              attribute transcript_type  \n",
       "0         .          Mouse_ID \"0\"; gene_id \"lnc1\";          lncRNA  \n",
       "1         .         Mouse_ID \"1\"; gene_id \"lnc10\";          lncRNA  \n",
       "2         .       Mouse_ID \"3\"; gene_id \"lnc1000\";          lncRNA  \n",
       "3         .      Mouse_ID \"4\"; gene_id \"lnc10000\";          lncRNA  \n",
       "4         .      Mouse_ID \"5\"; gene_id \"lnc10001\";          lncRNA  \n",
       "...     ...                                    ...             ...  \n",
       "65892     .     Mouse_ID \"82486\"; gene_id \"Zfas1\";          lncRNA  \n",
       "66085     .  Mouse_ID \"82711\"; gene_id \"Znf41-ps\";          lncRNA  \n",
       "66086     .  Mouse_ID \"82712\"; gene_id \"Znf41-ps\";          lncRNA  \n",
       "66087     .  Mouse_ID \"82713\"; gene_id \"Znf41-ps\";          lncRNA  \n",
       "66088     .  Mouse_ID \"82714\"; gene_id \"Znf41-ps\";          lncRNA  \n",
       "\n",
       "[41312 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "liftover_GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Rename columns to match standard GTF field names\n",
    "liftover_GTF = liftover_GTF.rename(columns={\n",
    "    'rat_chr': 'seqname',\n",
    "    'rat_start': 'start',\n",
    "    'rat_end': 'end',\n",
    "    'rat_score': 'score',\n",
    "    'rat_strand': 'strand'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>15038462</td>\n",
       "      <td>15039082</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"0\"; gene_id \"lnc1\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>14766774</td>\n",
       "      <td>14767820</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"1\"; gene_id \"lnc10\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>103001424</td>\n",
       "      <td>103005924</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"3\"; gene_id \"lnc1000\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>79676628</td>\n",
       "      <td>79677257</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"4\"; gene_id \"lnc10000\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>79677716</td>\n",
       "      <td>79678362</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"5\"; gene_id \"lnc10001\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65892</th>\n",
       "      <td>chr3</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>155794311</td>\n",
       "      <td>155794889</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"82486\"; gene_id \"Zfas1\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66085</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>158072676</td>\n",
       "      <td>158072721</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"82711\"; gene_id \"Znf41-ps\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66086</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>157249657</td>\n",
       "      <td>157249708</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"82712\"; gene_id \"Znf41-ps\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66087</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>157912765</td>\n",
       "      <td>157912821</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"82713\"; gene_id \"Znf41-ps\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66088</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>157480498</td>\n",
       "      <td>157480624</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"82714\"; gene_id \"Znf41-ps\";</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41312 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqname    source feature      start        end score strand frame  \\\n",
       "0        chr5  liftover    exon   15038462   15039082  1000      +     .   \n",
       "1        chr5  liftover    exon   14766774   14767820  1000      +     .   \n",
       "2       chr13  liftover    exon  103001424  103005924  1000      +     .   \n",
       "3       chr10  liftover    exon   79676628   79677257  1000      -     .   \n",
       "4       chr10  liftover    exon   79677716   79678362  1000      -     .   \n",
       "...       ...       ...     ...        ...        ...   ...    ...   ...   \n",
       "65892    chr3  liftover    exon  155794311  155794889  1000      +     .   \n",
       "66085    chr5  liftover    exon  158072676  158072721  1000      -     .   \n",
       "66086    chr5  liftover    exon  157249657  157249708  1000      +     .   \n",
       "66087    chr5  liftover    exon  157912765  157912821  1000      -     .   \n",
       "66088    chr5  liftover    exon  157480498  157480624  1000      +     .   \n",
       "\n",
       "                                   attribute transcript_type  \n",
       "0              Mouse_ID \"0\"; gene_id \"lnc1\";          lncRNA  \n",
       "1             Mouse_ID \"1\"; gene_id \"lnc10\";          lncRNA  \n",
       "2           Mouse_ID \"3\"; gene_id \"lnc1000\";          lncRNA  \n",
       "3          Mouse_ID \"4\"; gene_id \"lnc10000\";          lncRNA  \n",
       "4          Mouse_ID \"5\"; gene_id \"lnc10001\";          lncRNA  \n",
       "...                                      ...             ...  \n",
       "65892     Mouse_ID \"82486\"; gene_id \"Zfas1\";          lncRNA  \n",
       "66085  Mouse_ID \"82711\"; gene_id \"Znf41-ps\";          lncRNA  \n",
       "66086  Mouse_ID \"82712\"; gene_id \"Znf41-ps\";          lncRNA  \n",
       "66087  Mouse_ID \"82713\"; gene_id \"Znf41-ps\";          lncRNA  \n",
       "66088  Mouse_ID \"82714\"; gene_id \"Znf41-ps\";          lncRNA  \n",
       "\n",
       "[41312 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "liftover_GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>76909</td>\n",
       "      <td>77114</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Vom2r3\"; transcript_id \"XM_039088200....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>79753</td>\n",
       "      <td>80035</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Vom2r3\"; transcript_id \"XM_039088200....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>80483</td>\n",
       "      <td>81286</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Vom2r3\"; transcript_id \"XM_039088200....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>83590</td>\n",
       "      <td>83713</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Vom2r3\"; transcript_id \"XM_039088200....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>84879</td>\n",
       "      <td>85762</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Vom2r3\"; transcript_id \"XM_039088200....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529504</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>293338</td>\n",
       "      <td>293506</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC103694730\"; transcript_id \"XR_0054...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529505</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>300877</td>\n",
       "      <td>300988</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC103694730\"; transcript_id \"XR_0054...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529507</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>317993</td>\n",
       "      <td>319851</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC120099645\"; transcript_id \"XR_0054...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529508</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>320383</td>\n",
       "      <td>320541</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC120099645\"; transcript_id \"XR_0054...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529509</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>321212</td>\n",
       "      <td>321506</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC120099645\"; transcript_id \"XR_0054...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1142751 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            seqname      source feature   start     end score  \\\n",
       "1                              chr1  ncbiRefSeq    exon   76909   77114     .   \n",
       "3                              chr1  ncbiRefSeq    exon   79753   80035     .   \n",
       "5                              chr1  ncbiRefSeq    exon   80483   81286     .   \n",
       "7                              chr1  ncbiRefSeq    exon   83590   83713     .   \n",
       "9                              chr1  ncbiRefSeq    exon   84879   85762     .   \n",
       "...                             ...         ...     ...     ...     ...   ...   \n",
       "2529504  chrY_NW_023637722v1_random  ncbiRefSeq    exon  293338  293506     .   \n",
       "2529505  chrY_NW_023637722v1_random  ncbiRefSeq    exon  300877  300988     .   \n",
       "2529507  chrY_NW_023637722v1_random  ncbiRefSeq    exon  317993  319851     .   \n",
       "2529508  chrY_NW_023637722v1_random  ncbiRefSeq    exon  320383  320541     .   \n",
       "2529509  chrY_NW_023637722v1_random  ncbiRefSeq    exon  321212  321506     .   \n",
       "\n",
       "        strand frame                                          attribute  \n",
       "1            +     .  gene_id \"Vom2r3\"; transcript_id \"XM_039088200....  \n",
       "3            +     .  gene_id \"Vom2r3\"; transcript_id \"XM_039088200....  \n",
       "5            +     .  gene_id \"Vom2r3\"; transcript_id \"XM_039088200....  \n",
       "7            +     .  gene_id \"Vom2r3\"; transcript_id \"XM_039088200....  \n",
       "9            +     .  gene_id \"Vom2r3\"; transcript_id \"XM_039088200....  \n",
       "...        ...   ...                                                ...  \n",
       "2529504      +     .  gene_id \"LOC103694730\"; transcript_id \"XR_0054...  \n",
       "2529505      +     .  gene_id \"LOC103694730\"; transcript_id \"XR_0054...  \n",
       "2529507      -     .  gene_id \"LOC120099645\"; transcript_id \"XR_0054...  \n",
       "2529508      -     .  gene_id \"LOC120099645\"; transcript_id \"XR_0054...  \n",
       "2529509      -     .  gene_id \"LOC120099645\"; transcript_id \"XR_0054...  \n",
       "\n",
       "[1142751 rows x 9 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# URL to the file\n",
    "url = \"http://hgdownload.soe.ucsc.edu/goldenPath/rn7/bigZips/genes/ncbiRefSeq.gtf.gz\"\n",
    "\n",
    "# Download and read the .gtf.gz file\n",
    "rn7_GTF = pd.read_csv(url, \n",
    "                 compression='gzip', \n",
    "                 sep='\\t', \n",
    "                 comment='#', \n",
    "                 header=None, \n",
    "                 names=['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute'])\n",
    "\n",
    "#Filter transcript entries from GTF\n",
    "rn7_GTF_transcripts = rn7_GTF[rn7_GTF['feature'] == 'transcript']\n",
    "rn7_GTF_transcripts\n",
    "\n",
    "#Filter exon entries from GTF\n",
    "rn7_GTF_exons = rn7_GTF[rn7_GTF['feature'] == 'exon']\n",
    "rn7_GTF_exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>XM_039088200.1.1</td>\n",
       "      <td>Vom2r3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>XM_039088200.1.2</td>\n",
       "      <td>Vom2r3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>3</td>\n",
       "      <td>XM_039088200.1.3</td>\n",
       "      <td>Vom2r3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4</td>\n",
       "      <td>XM_039088200.1.4</td>\n",
       "      <td>Vom2r3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>5</td>\n",
       "      <td>XM_039088200.1.5</td>\n",
       "      <td>Vom2r3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529504</th>\n",
       "      <td>3</td>\n",
       "      <td>XR_005498756.1.3</td>\n",
       "      <td>LOC103694730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529505</th>\n",
       "      <td>4</td>\n",
       "      <td>XR_005498756.1.4</td>\n",
       "      <td>LOC103694730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529507</th>\n",
       "      <td>3</td>\n",
       "      <td>XR_005498757.1.3</td>\n",
       "      <td>LOC120099645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529508</th>\n",
       "      <td>2</td>\n",
       "      <td>XR_005498757.1.2</td>\n",
       "      <td>LOC120099645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529509</th>\n",
       "      <td>1</td>\n",
       "      <td>XR_005498757.1.1</td>\n",
       "      <td>LOC120099645</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1142751 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        exon_number           exon_id       gene_id\n",
       "1                 1  XM_039088200.1.1        Vom2r3\n",
       "3                 2  XM_039088200.1.2        Vom2r3\n",
       "5                 3  XM_039088200.1.3        Vom2r3\n",
       "7                 4  XM_039088200.1.4        Vom2r3\n",
       "9                 5  XM_039088200.1.5        Vom2r3\n",
       "...             ...               ...           ...\n",
       "2529504           3  XR_005498756.1.3  LOC103694730\n",
       "2529505           4  XR_005498756.1.4  LOC103694730\n",
       "2529507           3  XR_005498757.1.3  LOC120099645\n",
       "2529508           2  XR_005498757.1.2  LOC120099645\n",
       "2529509           1  XR_005498757.1.1  LOC120099645\n",
       "\n",
       "[1142751 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define a function to parse attributes and extract key-value pairs\n",
    "def parse_attributes(attribute_string):\n",
    "    attr_dict = {}\n",
    "    if pd.notnull(attribute_string):  # Check if the string is not NaN\n",
    "        attributes = attribute_string.split(';')\n",
    "        for attribute in attributes:\n",
    "            if attribute.strip():  # Ensure it's not an empty string\n",
    "                key, value = attribute.strip().split(' ', 1)\n",
    "                attr_dict[key] = value.strip('\"')  # Strip quotes from value\n",
    "    return attr_dict\n",
    "\n",
    "# Apply the function to extract the information into a DataFrame\n",
    "attributes_df = rn7_GTF_exons['attribute'].apply(parse_attributes).apply(pd.Series)\n",
    "\n",
    "# Drop the 'gene_id' column\n",
    "attributes_df = attributes_df.drop(columns=['gene_id'])\n",
    "attributes_df = attributes_df.drop(columns=['transcript_id'])\n",
    "\n",
    "# Rename 'gene_name' to 'gene_id'\n",
    "attributes_df = attributes_df.rename(columns={'gene_name': 'gene_id'})\n",
    "attributes_df = attributes_df.fillna('')\n",
    "attributes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33294"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attributes_df['gene_id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>76909</td>\n",
       "      <td>77114</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Vom2r3\"; transcript_id \"XM_039088200....</td>\n",
       "      <td>1</td>\n",
       "      <td>XM_039088200.1.1</td>\n",
       "      <td>Vom2r3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>79753</td>\n",
       "      <td>80035</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Vom2r3\"; transcript_id \"XM_039088200....</td>\n",
       "      <td>2</td>\n",
       "      <td>XM_039088200.1.2</td>\n",
       "      <td>Vom2r3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>80483</td>\n",
       "      <td>81286</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Vom2r3\"; transcript_id \"XM_039088200....</td>\n",
       "      <td>3</td>\n",
       "      <td>XM_039088200.1.3</td>\n",
       "      <td>Vom2r3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>83590</td>\n",
       "      <td>83713</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Vom2r3\"; transcript_id \"XM_039088200....</td>\n",
       "      <td>4</td>\n",
       "      <td>XM_039088200.1.4</td>\n",
       "      <td>Vom2r3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>84879</td>\n",
       "      <td>85762</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Vom2r3\"; transcript_id \"XM_039088200....</td>\n",
       "      <td>5</td>\n",
       "      <td>XM_039088200.1.5</td>\n",
       "      <td>Vom2r3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529504</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>293338</td>\n",
       "      <td>293506</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC103694730\"; transcript_id \"XR_0054...</td>\n",
       "      <td>3</td>\n",
       "      <td>XR_005498756.1.3</td>\n",
       "      <td>LOC103694730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529505</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>300877</td>\n",
       "      <td>300988</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC103694730\"; transcript_id \"XR_0054...</td>\n",
       "      <td>4</td>\n",
       "      <td>XR_005498756.1.4</td>\n",
       "      <td>LOC103694730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529507</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>317993</td>\n",
       "      <td>319851</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC120099645\"; transcript_id \"XR_0054...</td>\n",
       "      <td>3</td>\n",
       "      <td>XR_005498757.1.3</td>\n",
       "      <td>LOC120099645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529508</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>320383</td>\n",
       "      <td>320541</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC120099645\"; transcript_id \"XR_0054...</td>\n",
       "      <td>2</td>\n",
       "      <td>XR_005498757.1.2</td>\n",
       "      <td>LOC120099645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529509</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>321212</td>\n",
       "      <td>321506</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"LOC120099645\"; transcript_id \"XR_0054...</td>\n",
       "      <td>1</td>\n",
       "      <td>XR_005498757.1.1</td>\n",
       "      <td>LOC120099645</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1142751 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            seqname      source feature   start     end score  \\\n",
       "1                              chr1  ncbiRefSeq    exon   76909   77114     .   \n",
       "3                              chr1  ncbiRefSeq    exon   79753   80035     .   \n",
       "5                              chr1  ncbiRefSeq    exon   80483   81286     .   \n",
       "7                              chr1  ncbiRefSeq    exon   83590   83713     .   \n",
       "9                              chr1  ncbiRefSeq    exon   84879   85762     .   \n",
       "...                             ...         ...     ...     ...     ...   ...   \n",
       "2529504  chrY_NW_023637722v1_random  ncbiRefSeq    exon  293338  293506     .   \n",
       "2529505  chrY_NW_023637722v1_random  ncbiRefSeq    exon  300877  300988     .   \n",
       "2529507  chrY_NW_023637722v1_random  ncbiRefSeq    exon  317993  319851     .   \n",
       "2529508  chrY_NW_023637722v1_random  ncbiRefSeq    exon  320383  320541     .   \n",
       "2529509  chrY_NW_023637722v1_random  ncbiRefSeq    exon  321212  321506     .   \n",
       "\n",
       "        strand frame                                          attribute  \\\n",
       "1            +     .  gene_id \"Vom2r3\"; transcript_id \"XM_039088200....   \n",
       "3            +     .  gene_id \"Vom2r3\"; transcript_id \"XM_039088200....   \n",
       "5            +     .  gene_id \"Vom2r3\"; transcript_id \"XM_039088200....   \n",
       "7            +     .  gene_id \"Vom2r3\"; transcript_id \"XM_039088200....   \n",
       "9            +     .  gene_id \"Vom2r3\"; transcript_id \"XM_039088200....   \n",
       "...        ...   ...                                                ...   \n",
       "2529504      +     .  gene_id \"LOC103694730\"; transcript_id \"XR_0054...   \n",
       "2529505      +     .  gene_id \"LOC103694730\"; transcript_id \"XR_0054...   \n",
       "2529507      -     .  gene_id \"LOC120099645\"; transcript_id \"XR_0054...   \n",
       "2529508      -     .  gene_id \"LOC120099645\"; transcript_id \"XR_0054...   \n",
       "2529509      -     .  gene_id \"LOC120099645\"; transcript_id \"XR_0054...   \n",
       "\n",
       "        exon_number           exon_id       gene_id  \n",
       "1                 1  XM_039088200.1.1        Vom2r3  \n",
       "3                 2  XM_039088200.1.2        Vom2r3  \n",
       "5                 3  XM_039088200.1.3        Vom2r3  \n",
       "7                 4  XM_039088200.1.4        Vom2r3  \n",
       "9                 5  XM_039088200.1.5        Vom2r3  \n",
       "...             ...               ...           ...  \n",
       "2529504           3  XR_005498756.1.3  LOC103694730  \n",
       "2529505           4  XR_005498756.1.4  LOC103694730  \n",
       "2529507           3  XR_005498757.1.3  LOC120099645  \n",
       "2529508           2  XR_005498757.1.2  LOC120099645  \n",
       "2529509           1  XR_005498757.1.1  LOC120099645  \n",
       "\n",
       "[1142751 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Concatenate the two DataFrames side by side\n",
    "merged_combined = pd.concat([rn7_GTF_exons, attributes_df], axis=1)\n",
    "\n",
    "\n",
    "\n",
    "merged_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2229150</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58096021</td>\n",
       "      <td>58096146</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....</td>\n",
       "      <td>1</td>\n",
       "      <td>XM_006243150.3.1</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229152</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58098636</td>\n",
       "      <td>58099483</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....</td>\n",
       "      <td>2</td>\n",
       "      <td>XM_006243150.3.2</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229155</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58099948</td>\n",
       "      <td>58100074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....</td>\n",
       "      <td>3</td>\n",
       "      <td>XM_006243150.3.3</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229157</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58100159</td>\n",
       "      <td>58100248</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....</td>\n",
       "      <td>4</td>\n",
       "      <td>XM_006243150.3.4</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229159</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58100342</td>\n",
       "      <td>58100465</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....</td>\n",
       "      <td>5</td>\n",
       "      <td>XM_006243150.3.5</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229161</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58100614</td>\n",
       "      <td>58100700</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....</td>\n",
       "      <td>6</td>\n",
       "      <td>XM_006243150.3.6</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229163</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58100847</td>\n",
       "      <td>58102130</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....</td>\n",
       "      <td>7</td>\n",
       "      <td>XM_006243150.3.7</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229169</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58096089</td>\n",
       "      <td>58096146</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...</td>\n",
       "      <td>1</td>\n",
       "      <td>NM_012540.3.1</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229171</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58098633</td>\n",
       "      <td>58099483</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...</td>\n",
       "      <td>2</td>\n",
       "      <td>NM_012540.3.2</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229174</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58099948</td>\n",
       "      <td>58100074</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...</td>\n",
       "      <td>3</td>\n",
       "      <td>NM_012540.3.3</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229176</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58100159</td>\n",
       "      <td>58100248</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...</td>\n",
       "      <td>4</td>\n",
       "      <td>NM_012540.3.4</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229178</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58100342</td>\n",
       "      <td>58100465</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...</td>\n",
       "      <td>5</td>\n",
       "      <td>NM_012540.3.5</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229180</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58100614</td>\n",
       "      <td>58100700</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...</td>\n",
       "      <td>6</td>\n",
       "      <td>NM_012540.3.6</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229182</th>\n",
       "      <td>chr8</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>58100847</td>\n",
       "      <td>58102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...</td>\n",
       "      <td>7</td>\n",
       "      <td>NM_012540.3.7</td>\n",
       "      <td>Cyp1a1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname      source feature     start       end score strand frame  \\\n",
       "2229150    chr8  ncbiRefSeq    exon  58096021  58096146     .      +     .   \n",
       "2229152    chr8  ncbiRefSeq    exon  58098636  58099483     .      +     .   \n",
       "2229155    chr8  ncbiRefSeq    exon  58099948  58100074     .      +     .   \n",
       "2229157    chr8  ncbiRefSeq    exon  58100159  58100248     .      +     .   \n",
       "2229159    chr8  ncbiRefSeq    exon  58100342  58100465     .      +     .   \n",
       "2229161    chr8  ncbiRefSeq    exon  58100614  58100700     .      +     .   \n",
       "2229163    chr8  ncbiRefSeq    exon  58100847  58102130     .      +     .   \n",
       "2229169    chr8  ncbiRefSeq    exon  58096089  58096146     .      +     .   \n",
       "2229171    chr8  ncbiRefSeq    exon  58098633  58099483     .      +     .   \n",
       "2229174    chr8  ncbiRefSeq    exon  58099948  58100074     .      +     .   \n",
       "2229176    chr8  ncbiRefSeq    exon  58100159  58100248     .      +     .   \n",
       "2229178    chr8  ncbiRefSeq    exon  58100342  58100465     .      +     .   \n",
       "2229180    chr8  ncbiRefSeq    exon  58100614  58100700     .      +     .   \n",
       "2229182    chr8  ncbiRefSeq    exon  58100847  58102125     .      +     .   \n",
       "\n",
       "                                                 attribute exon_number  \\\n",
       "2229150  gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....           1   \n",
       "2229152  gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....           2   \n",
       "2229155  gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....           3   \n",
       "2229157  gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....           4   \n",
       "2229159  gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....           5   \n",
       "2229161  gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....           6   \n",
       "2229163  gene_id \"Cyp1a1\"; transcript_id \"XM_006243150....           7   \n",
       "2229169  gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...           1   \n",
       "2229171  gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...           2   \n",
       "2229174  gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...           3   \n",
       "2229176  gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...           4   \n",
       "2229178  gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...           5   \n",
       "2229180  gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...           6   \n",
       "2229182  gene_id \"Cyp1a1\"; transcript_id \"NM_012540.3\";...           7   \n",
       "\n",
       "                  exon_id gene_id  \n",
       "2229150  XM_006243150.3.1  Cyp1a1  \n",
       "2229152  XM_006243150.3.2  Cyp1a1  \n",
       "2229155  XM_006243150.3.3  Cyp1a1  \n",
       "2229157  XM_006243150.3.4  Cyp1a1  \n",
       "2229159  XM_006243150.3.5  Cyp1a1  \n",
       "2229161  XM_006243150.3.6  Cyp1a1  \n",
       "2229163  XM_006243150.3.7  Cyp1a1  \n",
       "2229169     NM_012540.3.1  Cyp1a1  \n",
       "2229171     NM_012540.3.2  Cyp1a1  \n",
       "2229174     NM_012540.3.3  Cyp1a1  \n",
       "2229176     NM_012540.3.4  Cyp1a1  \n",
       "2229178     NM_012540.3.5  Cyp1a1  \n",
       "2229180     NM_012540.3.6  Cyp1a1  \n",
       "2229182     NM_012540.3.7  Cyp1a1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_combined[merged_combined['gene_id'] == \"Cyp1a1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>attribute</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>76909</td>\n",
       "      <td>77114</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>XM_039088200.1.1</td>\n",
       "      <td>Vom2r3</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>79753</td>\n",
       "      <td>80035</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "      <td>XM_039088200.1.2</td>\n",
       "      <td>Vom2r3</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>80483</td>\n",
       "      <td>81286</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>3</td>\n",
       "      <td>XM_039088200.1.3</td>\n",
       "      <td>Vom2r3</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>83590</td>\n",
       "      <td>83713</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>4</td>\n",
       "      <td>XM_039088200.1.4</td>\n",
       "      <td>Vom2r3</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>84879</td>\n",
       "      <td>85762</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>5</td>\n",
       "      <td>XM_039088200.1.5</td>\n",
       "      <td>Vom2r3</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529504</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>293338</td>\n",
       "      <td>293506</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>3</td>\n",
       "      <td>XR_005498756.1.3</td>\n",
       "      <td>LOC103694730</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529505</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>300877</td>\n",
       "      <td>300988</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>4</td>\n",
       "      <td>XR_005498756.1.4</td>\n",
       "      <td>LOC103694730</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529507</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>317993</td>\n",
       "      <td>319851</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>3</td>\n",
       "      <td>XR_005498757.1.3</td>\n",
       "      <td>LOC120099645</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529508</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>320383</td>\n",
       "      <td>320541</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "      <td>XR_005498757.1.2</td>\n",
       "      <td>LOC120099645</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529509</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>321212</td>\n",
       "      <td>321506</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>XR_005498757.1.1</td>\n",
       "      <td>LOC120099645</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1142751 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            seqname      source feature   start     end score  \\\n",
       "1                              chr1  ncbiRefSeq    exon   76909   77114     .   \n",
       "3                              chr1  ncbiRefSeq    exon   79753   80035     .   \n",
       "5                              chr1  ncbiRefSeq    exon   80483   81286     .   \n",
       "7                              chr1  ncbiRefSeq    exon   83590   83713     .   \n",
       "9                              chr1  ncbiRefSeq    exon   84879   85762     .   \n",
       "...                             ...         ...     ...     ...     ...   ...   \n",
       "2529504  chrY_NW_023637722v1_random  ncbiRefSeq    exon  293338  293506     .   \n",
       "2529505  chrY_NW_023637722v1_random  ncbiRefSeq    exon  300877  300988     .   \n",
       "2529507  chrY_NW_023637722v1_random  ncbiRefSeq    exon  317993  319851     .   \n",
       "2529508  chrY_NW_023637722v1_random  ncbiRefSeq    exon  320383  320541     .   \n",
       "2529509  chrY_NW_023637722v1_random  ncbiRefSeq    exon  321212  321506     .   \n",
       "\n",
       "        strand frame exon_number           exon_id     attribute  \\\n",
       "1            +     .           1  XM_039088200.1.1        Vom2r3   \n",
       "3            +     .           2  XM_039088200.1.2        Vom2r3   \n",
       "5            +     .           3  XM_039088200.1.3        Vom2r3   \n",
       "7            +     .           4  XM_039088200.1.4        Vom2r3   \n",
       "9            +     .           5  XM_039088200.1.5        Vom2r3   \n",
       "...        ...   ...         ...               ...           ...   \n",
       "2529504      +     .           3  XR_005498756.1.3  LOC103694730   \n",
       "2529505      +     .           4  XR_005498756.1.4  LOC103694730   \n",
       "2529507      -     .           3  XR_005498757.1.3  LOC120099645   \n",
       "2529508      -     .           2  XR_005498757.1.2  LOC120099645   \n",
       "2529509      -     .           1  XR_005498757.1.1  LOC120099645   \n",
       "\n",
       "          transcript_type  \n",
       "1          predicted_mRNA  \n",
       "3          predicted_mRNA  \n",
       "5          predicted_mRNA  \n",
       "7          predicted_mRNA  \n",
       "9          predicted_mRNA  \n",
       "...                   ...  \n",
       "2529504  predicted_lncRNA  \n",
       "2529505  predicted_lncRNA  \n",
       "2529507  predicted_lncRNA  \n",
       "2529508  predicted_lncRNA  \n",
       "2529509  predicted_lncRNA  \n",
       "\n",
       "[1142751 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a new column 'transcript_type' based on the conditions\n",
    "merged_combined['transcript_type'] = np.where(\n",
    "    merged_combined['attribute'].str.contains('NR_', na=False), 'lncRNA',\n",
    "    np.where(\n",
    "        merged_combined['attribute'].str.contains('XR_', na=False), 'predicted_lncRNA',\n",
    "        np.where(\n",
    "            merged_combined['attribute'].str.contains('XM_', na=False), 'predicted_mRNA',\n",
    "            np.where(\n",
    "                merged_combined['attribute'].str.contains('NM_', na=False), 'mRNA',\n",
    "                'unknown'\n",
    "            )\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "#Drop old 'attribute' column and rename 'gene_id' as the new 'attribute'\n",
    "merged_combined = merged_combined.drop(columns=['attribute'])\n",
    "merged_combined = merged_combined.rename(columns={'gene_id': 'attribute'})\n",
    "\n",
    "merged_combined\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seqname                26\n",
       "source                  1\n",
       "feature                 1\n",
       "start              162274\n",
       "end                162283\n",
       "score                   1\n",
       "strand                  2\n",
       "frame                   1\n",
       "exon_number           118\n",
       "exon_id            172276\n",
       "attribute           16765\n",
       "transcript_type         1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Count unique values per column for rows where 'transcript_type' is 'mRNA'\n",
    "merged_combined[merged_combined['transcript_type'] == 'mRNA'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/local/52896496/ipykernel_1855931/2179824070.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  rn7_GTF_transcripts_filtered['Rat_ID'] = range(len(rn7_GTF_transcripts_filtered))\n"
     ]
    }
   ],
   "source": [
    "#Calculate transcript length\n",
    "merged_combined['length'] = merged_combined['end'] - merged_combined['start']\n",
    "\n",
    "#Filter out rows where transcript_type is 'lncRNA' and length is less than 200\n",
    "rn7_GTF_transcripts_filtered = merged_combined[\n",
    "    ~((merged_combined['transcript_type'] == 'lncRNA') & (merged_combined['length'] < 200))\n",
    "]\n",
    "\n",
    "rn7_GTF_transcripts_filtered = rn7_GTF_transcripts_filtered#.drop_duplicates(subset=['attribute'], keep='first')\n",
    "\n",
    "#Assign unique Rat_IDs to each remaining transcript\n",
    "rn7_GTF_transcripts_filtered['Rat_ID'] = range(len(rn7_GTF_transcripts_filtered))\n",
    "\n",
    "#Create a copy for further processing\n",
    "rn7_GTF_transcripts_filtered2 = rn7_GTF_transcripts_filtered.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/local/52896496/ipykernel_1855931/4061719296.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  rn7_GTF_transcripts_filtered['attribute'] = (\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>attribute</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>76909</td>\n",
       "      <td>77114</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>XM_039088200.1.1</td>\n",
       "      <td>Rat_ID \"0\"; gene_id \"Vom2r3\";</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>79753</td>\n",
       "      <td>80035</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "      <td>XM_039088200.1.2</td>\n",
       "      <td>Rat_ID \"1\"; gene_id \"Vom2r3\";</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>80483</td>\n",
       "      <td>81286</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>3</td>\n",
       "      <td>XM_039088200.1.3</td>\n",
       "      <td>Rat_ID \"2\"; gene_id \"Vom2r3\";</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>83590</td>\n",
       "      <td>83713</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>4</td>\n",
       "      <td>XM_039088200.1.4</td>\n",
       "      <td>Rat_ID \"3\"; gene_id \"Vom2r3\";</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>84879</td>\n",
       "      <td>85762</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>5</td>\n",
       "      <td>XM_039088200.1.5</td>\n",
       "      <td>Rat_ID \"4\"; gene_id \"Vom2r3\";</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529504</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>293338</td>\n",
       "      <td>293506</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>3</td>\n",
       "      <td>XR_005498756.1.3</td>\n",
       "      <td>Rat_ID \"1141644\"; gene_id \"LOC103694730\";</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529505</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>300877</td>\n",
       "      <td>300988</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>4</td>\n",
       "      <td>XR_005498756.1.4</td>\n",
       "      <td>Rat_ID \"1141645\"; gene_id \"LOC103694730\";</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529507</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>317993</td>\n",
       "      <td>319851</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>3</td>\n",
       "      <td>XR_005498757.1.3</td>\n",
       "      <td>Rat_ID \"1141646\"; gene_id \"LOC120099645\";</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529508</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>320383</td>\n",
       "      <td>320541</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "      <td>XR_005498757.1.2</td>\n",
       "      <td>Rat_ID \"1141647\"; gene_id \"LOC120099645\";</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529509</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>321212</td>\n",
       "      <td>321506</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>XR_005498757.1.1</td>\n",
       "      <td>Rat_ID \"1141648\"; gene_id \"LOC120099645\";</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1141649 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            seqname      source feature   start     end score  \\\n",
       "1                              chr1  ncbiRefSeq    exon   76909   77114     .   \n",
       "3                              chr1  ncbiRefSeq    exon   79753   80035     .   \n",
       "5                              chr1  ncbiRefSeq    exon   80483   81286     .   \n",
       "7                              chr1  ncbiRefSeq    exon   83590   83713     .   \n",
       "9                              chr1  ncbiRefSeq    exon   84879   85762     .   \n",
       "...                             ...         ...     ...     ...     ...   ...   \n",
       "2529504  chrY_NW_023637722v1_random  ncbiRefSeq    exon  293338  293506     .   \n",
       "2529505  chrY_NW_023637722v1_random  ncbiRefSeq    exon  300877  300988     .   \n",
       "2529507  chrY_NW_023637722v1_random  ncbiRefSeq    exon  317993  319851     .   \n",
       "2529508  chrY_NW_023637722v1_random  ncbiRefSeq    exon  320383  320541     .   \n",
       "2529509  chrY_NW_023637722v1_random  ncbiRefSeq    exon  321212  321506     .   \n",
       "\n",
       "        strand frame exon_number           exon_id  \\\n",
       "1            +     .           1  XM_039088200.1.1   \n",
       "3            +     .           2  XM_039088200.1.2   \n",
       "5            +     .           3  XM_039088200.1.3   \n",
       "7            +     .           4  XM_039088200.1.4   \n",
       "9            +     .           5  XM_039088200.1.5   \n",
       "...        ...   ...         ...               ...   \n",
       "2529504      +     .           3  XR_005498756.1.3   \n",
       "2529505      +     .           4  XR_005498756.1.4   \n",
       "2529507      -     .           3  XR_005498757.1.3   \n",
       "2529508      -     .           2  XR_005498757.1.2   \n",
       "2529509      -     .           1  XR_005498757.1.1   \n",
       "\n",
       "                                         attribute   transcript_type  \n",
       "1                    Rat_ID \"0\"; gene_id \"Vom2r3\";    predicted_mRNA  \n",
       "3                    Rat_ID \"1\"; gene_id \"Vom2r3\";    predicted_mRNA  \n",
       "5                    Rat_ID \"2\"; gene_id \"Vom2r3\";    predicted_mRNA  \n",
       "7                    Rat_ID \"3\"; gene_id \"Vom2r3\";    predicted_mRNA  \n",
       "9                    Rat_ID \"4\"; gene_id \"Vom2r3\";    predicted_mRNA  \n",
       "...                                            ...               ...  \n",
       "2529504  Rat_ID \"1141644\"; gene_id \"LOC103694730\";  predicted_lncRNA  \n",
       "2529505  Rat_ID \"1141645\"; gene_id \"LOC103694730\";  predicted_lncRNA  \n",
       "2529507  Rat_ID \"1141646\"; gene_id \"LOC120099645\";  predicted_lncRNA  \n",
       "2529508  Rat_ID \"1141647\"; gene_id \"LOC120099645\";  predicted_lncRNA  \n",
       "2529509  Rat_ID \"1141648\"; gene_id \"LOC120099645\";  predicted_lncRNA  \n",
       "\n",
       "[1141649 rows x 12 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a new DataFrame without the 'length' column\n",
    "rn7_GTF_transcripts_filtered['attribute'] = (\n",
    "    'Rat_ID \"' + rn7_GTF_transcripts_filtered['Rat_ID'].astype(str) + \n",
    "    '\"; gene_id \"' + rn7_GTF_transcripts_filtered['attribute'] + '\";'\n",
    ")\n",
    "\n",
    "rn7_GTF_transcripts_filtered = rn7_GTF_transcripts_filtered.drop(columns=['length', 'Rat_ID'])\n",
    "\n",
    "\n",
    "# Display the new DataFrame\n",
    "rn7_GTF_transcripts_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'liftover_GTF' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Concatenate the DataFrames row-wise\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m stacked_df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([\u001b[43mliftover_GTF\u001b[49m, rn7_GTF_transcripts_filtered], axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m, ignore_index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      4\u001b[0m stacked_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnew\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m stacked_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mattribute\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m transcript_type \u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m stacked_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtranscript_type\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m# Drop the old 'attribute' and 'type' columns\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'liftover_GTF' is not defined"
     ]
    }
   ],
   "source": [
    "# Concatenate the DataFrames row-wise\n",
    "stacked_df = pd.concat([liftover_GTF, rn7_GTF_transcripts_filtered], axis=0, ignore_index=True)\n",
    "\n",
    "stacked_df['new'] = stacked_df['attribute'] + ' transcript_type \"' + stacked_df['transcript_type'] + '\"'\n",
    "\n",
    "# Drop the old 'attribute' and 'type' columns\n",
    "stacked_df = stacked_df.drop(columns=['attribute', 'transcript_type'])\n",
    "\n",
    "# Rename the 'new' column to 'attribute'\n",
    "stacked_df = stacked_df.rename(columns={'new': 'attribute'})\n",
    "\n",
    "stacked_df['score'] = 1000\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "stacked_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rn7_GTF_transcripts_filtered2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "liftover_GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to parse attributes and extract key-value pairs\n",
    "def parse_attributes(attribute_string):\n",
    "    attr_dict = {}\n",
    "    if pd.notnull(attribute_string):  # Check if the string is not NaN\n",
    "        attributes = attribute_string.split(';')\n",
    "        for attribute in attributes:\n",
    "            if attribute.strip():  # Ensure it's not an empty string\n",
    "                key, value = attribute.strip().split(' ', 1)\n",
    "                attr_dict[key] = value.strip('\"')  # Strip quotes from value\n",
    "    return attr_dict\n",
    "\n",
    "# Apply the function to extract the information into a DataFrame\n",
    "attributes_df = liftover_GTF['attribute'].apply(parse_attributes).apply(pd.Series)\n",
    "\n",
    "liftover_GTF_temp = pd.concat([liftover_GTF, attributes_df], axis=1)\n",
    "liftover_GTF_temp = liftover_GTF_temp.rename(columns={'gene_id': 'Gene'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Extract unique gene and transcript type pairs from LiftOver GTF\n",
    "liftover_genes = liftover_GTF_temp[['Gene', 'transcript_type']].drop_duplicates(keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Vom2r3</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>Vom2r4</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>Vom2r1</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>Vom2r5</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>Vom2r6</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2528626</th>\n",
       "      <td>Ddx3</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2528665</th>\n",
       "      <td>Usp9y</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529213</th>\n",
       "      <td>Dkc1</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529349</th>\n",
       "      <td>Tspy1</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529418</th>\n",
       "      <td>Sry</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16935 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene transcript_type\n",
       "29       Vom2r3            mRNA\n",
       "175      Vom2r4            mRNA\n",
       "356      Vom2r1            mRNA\n",
       "616      Vom2r5            mRNA\n",
       "704      Vom2r6            mRNA\n",
       "...         ...             ...\n",
       "2528626    Ddx3            mRNA\n",
       "2528665   Usp9y            mRNA\n",
       "2529213    Dkc1            mRNA\n",
       "2529349   Tspy1            mRNA\n",
       "2529418     Sry            mRNA\n",
       "\n",
       "[16935 rows x 2 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Rename 'attribute' to 'Gene' and filter for lncRNA or mRNA transcript types\n",
    "rn7_genes = rn7_GTF_transcripts_filtered2[['attribute','transcript_type']].rename(columns={'attribute': 'Gene'})\n",
    "rn7_genes = rn7_genes[(rn7_genes['transcript_type'] == 'lncRNA') |\n",
    "                      (rn7_genes['transcript_type'] == 'mRNA')].drop_duplicates(keep='first')\n",
    "rn7_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14762</th>\n",
       "      <td>LOC100910237</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29268</th>\n",
       "      <td>Oprm1</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34373</th>\n",
       "      <td>Cahm</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43580</th>\n",
       "      <td>Rps9l1</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45919</th>\n",
       "      <td>Apeg3</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2469326</th>\n",
       "      <td>Kantr</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2482025</th>\n",
       "      <td>RGD1564534</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2488552</th>\n",
       "      <td>LOC100911498</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2488601</th>\n",
       "      <td>Ftx</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2515147</th>\n",
       "      <td>LOC102550394</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>170 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene transcript_type\n",
       "14762    LOC100910237          lncRNA\n",
       "29268           Oprm1          lncRNA\n",
       "34373            Cahm          lncRNA\n",
       "43580          Rps9l1          lncRNA\n",
       "45919           Apeg3          lncRNA\n",
       "...               ...             ...\n",
       "2469326         Kantr          lncRNA\n",
       "2482025    RGD1564534          lncRNA\n",
       "2488552  LOC100911498          lncRNA\n",
       "2488601           Ftx          lncRNA\n",
       "2515147  LOC102550394          lncRNA\n",
       "\n",
       "[170 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Get unique lncRNA gene entries\n",
    "rn7_genes[(rn7_genes['transcript_type'] == 'lncRNA')].drop_duplicates(keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Vom2r3</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>Vom2r4</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>Vom2r1</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>Vom2r5</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>704</th>\n",
       "      <td>Vom2r6</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2528626</th>\n",
       "      <td>Ddx3</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2528665</th>\n",
       "      <td>Usp9y</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529213</th>\n",
       "      <td>Dkc1</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529349</th>\n",
       "      <td>Tspy1</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529418</th>\n",
       "      <td>Sry</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16765 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene transcript_type\n",
       "29       Vom2r3            mRNA\n",
       "175      Vom2r4            mRNA\n",
       "356      Vom2r1            mRNA\n",
       "616      Vom2r5            mRNA\n",
       "704      Vom2r6            mRNA\n",
       "...         ...             ...\n",
       "2528626    Ddx3            mRNA\n",
       "2528665   Usp9y            mRNA\n",
       "2529213    Dkc1            mRNA\n",
       "2529349   Tspy1            mRNA\n",
       "2529418     Sry            mRNA\n",
       "\n",
       "[16765 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get unique mRNA gene entries\n",
    "rn7_genes[(rn7_genes['transcript_type'] == 'mRNA')].drop_duplicates(keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc1</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc10</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc1000</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc10000</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc10001</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2528626</th>\n",
       "      <td>Ddx3</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2528665</th>\n",
       "      <td>Usp9y</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529213</th>\n",
       "      <td>Dkc1</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529349</th>\n",
       "      <td>Tspy1</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2529418</th>\n",
       "      <td>Sry</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>57284 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene transcript_type\n",
       "0            lnc1          lncRNA\n",
       "1           lnc10          lncRNA\n",
       "2         lnc1000          lncRNA\n",
       "3        lnc10000          lncRNA\n",
       "4        lnc10001          lncRNA\n",
       "...           ...             ...\n",
       "2528626      Ddx3            mRNA\n",
       "2528665     Usp9y            mRNA\n",
       "2529213      Dkc1            mRNA\n",
       "2529349     Tspy1            mRNA\n",
       "2529418       Sry            mRNA\n",
       "\n",
       "[57284 rows x 2 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Merge LiftOver and rn7 gene lists, removing duplicates\n",
    "gene_lists_merged = pd.concat([liftover_genes, rn7_genes], axis=0).drop_duplicates(keep='first')\n",
    "gene_lists_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save merged gene list to a tab-separated text file\n",
    "gene_lists_merged.to_csv('RN_and_Karri_Annotations_Restructured_rn7.txt', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Export stacked_df as GTF for gene expression analysis without headers or quotes\n",
    "import csv\n",
    "stacked_df.to_csv('./Karri_Annotations_Restructured_rn7_for_Gene_Expression.gtf', sep='\\t', \n",
    "                  index=False, header=False, quoting=csv.QUOTE_NONE, escapechar='')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "SCVI New",
   "language": "python",
   "name": "scvi-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
