{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge GTF file from Karri et al. with MGI Annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 1. Import Required Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 2. Import and Process Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: If your Wiggle data is a significant portion of available system memory, use the --max-mem and --sort-tmpdir options, or use --do-not-sort to disable post-conversion sorting. See --help for more information.\n"
     ]
    }
   ],
   "source": [
    "# Convert GTF to BED format\n",
    "!gff2bed < ./RAW_Data/Karri_Annotations_Restructured_mm39.gtf > ./RAW_Data/Karri_Annotations_Restructured_mm39.bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Sort BED files by chromosome and start coordinate\n",
    "!sort -k1,1 -k2,2n ./RAW_Data/Karri_Annotations_Restructured_mm39.bed > ./RAW_Data/sorted_Karri_Annotations_Restructured_mm39.bed\n",
    "!sort -k1,1 -k2,2n ./RAW_Data/MGI_lncRNAs_mm39_Processed.txt > ./RAW_Data/sorted_MGI_lncRNAs_mm39_Processed.bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Perform strand-specific intersection using bedtools\n",
    "!bedtools intersect -sorted -loj -a ./RAW_Data/sorted_Karri_Annotations_Restructured_mm39.bed -b ./RAW_Data/sorted_MGI_lncRNAs_mm39_Processed.bed -s > ./RAW_Data/Karri_vs_MGI_lncRNA_Intersection.bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Create a temporary file with headers\n",
    "!echo -e \"chrom\\tstart\\tend\\tblank\\tscore\\tstrand\\tsource\\tfeature\\tframe\\tattributes\\tmgi_chr\\tmgi_start\\tmgi_end\\tmgi_symbol\\tmgi_score\\tmgi_strand\" > ./RAW_Data/Karri_vs_MGI_lncRNA_Intersection_temp.bed\n",
    "\n",
    "# Append the original intersection data to the temporary file\n",
    "!tail -n +2 ./RAW_Data/Karri_vs_MGI_lncRNA_Intersection.bed >> ./RAW_Data/Karri_vs_MGI_lncRNA_Intersection_temp.bed\n",
    "\n",
    "# Replace the original file with the updated one\n",
    "!mv ./RAW_Data/Karri_vs_MGI_lncRNA_Intersection_temp.bed ./RAW_Data/Karri_vs_MGI_lncRNA_Intersection.txt\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/local/51093422/ipykernel_1608946/4097613543.py:2: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  RAW_Master = pd.read_csv('./RAW_Data/Karri_vs_MGI_lncRNA_Intersection.txt',\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>blank</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>mgi_chr</th>\n",
       "      <th>mgi_start</th>\n",
       "      <th>mgi_end</th>\n",
       "      <th>mgi_symbol</th>\n",
       "      <th>mgi_score</th>\n",
       "      <th>mgi_strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3224554</td>\n",
       "      <td>3224816</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"lnc_inter_chr1_15560\"; transcript_id ...</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3273138</td>\n",
       "      <td>3274559</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"lnc_inter_chr1_15561\"; transcript_id ...</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3275710</td>\n",
       "      <td>3276991</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"lnc_inter_chr1_15562\"; transcript_id ...</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3284704</td>\n",
       "      <td>3741721</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>GB_RefseqNM19801</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Xkr4\"; transcript_id \"Xkr4\"; gene_nam...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3717532</td>\n",
       "      <td>3729127</td>\n",
       "      <td>Gm19938</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3536809</td>\n",
       "      <td>3583776</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>EnsmblNR4698</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Gm1992\"; transcript_id \"Gm1992\"; gene...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3536810</td>\n",
       "      <td>3583776</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84244</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90614769</td>\n",
       "      <td>90617133</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>EnsmblNR4698</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Gm28300\"; transcript_id \"Gm28300\"; ge...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>90614770</td>\n",
       "      <td>90617133</td>\n",
       "      <td>Gm28300</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84245</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90676614</td>\n",
       "      <td>90678894</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>EnsmblNR4698</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Gm28301\"; transcript_id \"Gm28301\"; ge...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>90676615</td>\n",
       "      <td>90678894</td>\n",
       "      <td>Gm28301</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84246</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90762408</td>\n",
       "      <td>90766319</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>GB_RefseqNR2077</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"G530011O06Rik@chrY(-)\"; transcript_id...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>90763696</td>\n",
       "      <td>90766736</td>\n",
       "      <td>G530011O06Riky</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84247</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90796710</td>\n",
       "      <td>90827734</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>GB_RefseqNM#GB_RefseqNR1159</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Erdr1\"; transcript_id \"Erdr1\"; gene_n...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>90796007</td>\n",
       "      <td>90827734</td>\n",
       "      <td>Erdr1y</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84248</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90796710</td>\n",
       "      <td>90827734</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>GB_RefseqNM#GB_RefseqNR1159</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"Erdr1\"; transcript_id \"Erdr1\"; gene_n...</td>\n",
       "      <td>chrY</td>\n",
       "      <td>90804956</td>\n",
       "      <td>90820875</td>\n",
       "      <td>Gm33272</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84249 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end blank score strand  \\\n",
       "0      chr1   3224554   3224816     .  1000      -   \n",
       "1      chr1   3273138   3274559     .  1000      -   \n",
       "2      chr1   3275710   3276991     .  1000      -   \n",
       "3      chr1   3284704   3741721     .  1000      -   \n",
       "4      chr1   3536809   3583776     .  1000      +   \n",
       "...     ...       ...       ...   ...   ...    ...   \n",
       "84244  chrY  90614769  90617133     .  1000      -   \n",
       "84245  chrY  90676614  90678894     .  1000      -   \n",
       "84246  chrY  90762408  90766319     .  1000      -   \n",
       "84247  chrY  90796710  90827734     .  1000      +   \n",
       "84248  chrY  90796710  90827734     .  1000      +   \n",
       "\n",
       "                            source feature frame  \\\n",
       "0           Genebody_mm10_lnc48261    exon     .   \n",
       "1           Genebody_mm10_lnc48261    exon     .   \n",
       "2           Genebody_mm10_lnc48261    exon     .   \n",
       "3                 GB_RefseqNM19801    exon     .   \n",
       "4                     EnsmblNR4698    exon     .   \n",
       "...                            ...     ...   ...   \n",
       "84244                 EnsmblNR4698    exon     .   \n",
       "84245                 EnsmblNR4698    exon     .   \n",
       "84246              GB_RefseqNR2077    exon     .   \n",
       "84247  GB_RefseqNM#GB_RefseqNR1159    exon     .   \n",
       "84248  GB_RefseqNM#GB_RefseqNR1159    exon     .   \n",
       "\n",
       "                                              attributes mgi_chr  mgi_start  \\\n",
       "0      gene_id \"lnc_inter_chr1_15560\"; transcript_id ...       .         -1   \n",
       "1      gene_id \"lnc_inter_chr1_15561\"; transcript_id ...       .         -1   \n",
       "2      gene_id \"lnc_inter_chr1_15562\"; transcript_id ...       .         -1   \n",
       "3      gene_id \"Xkr4\"; transcript_id \"Xkr4\"; gene_nam...    chr1    3717532   \n",
       "4      gene_id \"Gm1992\"; transcript_id \"Gm1992\"; gene...    chr1    3536810   \n",
       "...                                                  ...     ...        ...   \n",
       "84244  gene_id \"Gm28300\"; transcript_id \"Gm28300\"; ge...    chrY   90614770   \n",
       "84245  gene_id \"Gm28301\"; transcript_id \"Gm28301\"; ge...    chrY   90676615   \n",
       "84246  gene_id \"G530011O06Rik@chrY(-)\"; transcript_id...    chrY   90763696   \n",
       "84247  gene_id \"Erdr1\"; transcript_id \"Erdr1\"; gene_n...    chrY   90796007   \n",
       "84248  gene_id \"Erdr1\"; transcript_id \"Erdr1\"; gene_n...    chrY   90804956   \n",
       "\n",
       "        mgi_end      mgi_symbol  mgi_score mgi_strand  \n",
       "0            -1               .         -1          .  \n",
       "1            -1               .         -1          .  \n",
       "2            -1               .         -1          .  \n",
       "3       3729127         Gm19938       1000          -  \n",
       "4       3583776          Gm1992       1000          +  \n",
       "...         ...             ...        ...        ...  \n",
       "84244  90617133         Gm28300       1000          -  \n",
       "84245  90678894         Gm28301       1000          -  \n",
       "84246  90766736  G530011O06Riky       1000          -  \n",
       "84247  90827734          Erdr1y       1000          +  \n",
       "84248  90820875         Gm33272       1000          +  \n",
       "\n",
       "[84249 rows x 16 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the GTF file into a DataFrame\n",
    "RAW_Master = pd.read_csv('./RAW_Data/Karri_vs_MGI_lncRNA_Intersection.txt', \n",
    "                         delimiter='\\t', header=0)\n",
    "\n",
    "\n",
    "RAW_Master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>blank</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>frame</th>\n",
       "      <th>mgi_chr</th>\n",
       "      <th>...</th>\n",
       "      <th>mgi_end</th>\n",
       "      <th>mgi_symbol</th>\n",
       "      <th>mgi_score</th>\n",
       "      <th>mgi_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>NR_annotation</th>\n",
       "      <th>NR_annotation2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3224554</td>\n",
       "      <td>3224816</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>lnc_inter_chr1_15560</td>\n",
       "      <td>lnc_inter_chr1_15560</td>\n",
       "      <td>lnc15560</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc15560</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3273138</td>\n",
       "      <td>3274559</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>lnc_inter_chr1_15561</td>\n",
       "      <td>lnc_inter_chr1_15561</td>\n",
       "      <td>lnc15561</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc15561</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3275710</td>\n",
       "      <td>3276991</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Genebody_mm10_lnc48261</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>lnc_inter_chr1_15562</td>\n",
       "      <td>lnc_inter_chr1_15562</td>\n",
       "      <td>lnc15562</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>lnc15562</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3284704</td>\n",
       "      <td>3741721</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>GB_RefseqNM19801</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>...</td>\n",
       "      <td>3729127</td>\n",
       "      <td>Gm19938</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>NM</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3536809</td>\n",
       "      <td>3583776</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>EnsmblNR4698</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1</td>\n",
       "      <td>...</td>\n",
       "      <td>3583776</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>Gm1992</td>\n",
       "      <td>antisense</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84244</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90614769</td>\n",
       "      <td>90617133</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>EnsmblNR4698</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>90617133</td>\n",
       "      <td>Gm28300</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Gm28300</td>\n",
       "      <td>Gm28300</td>\n",
       "      <td>Gm28300</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84245</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90676614</td>\n",
       "      <td>90678894</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>EnsmblNR4698</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>90678894</td>\n",
       "      <td>Gm28301</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>Gm28301</td>\n",
       "      <td>Gm28301</td>\n",
       "      <td>Gm28301</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84246</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90762408</td>\n",
       "      <td>90766319</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>GB_RefseqNR2077</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>90766736</td>\n",
       "      <td>G530011O06Riky</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>G530011O06Rik@chrY(-)</td>\n",
       "      <td>G530011O06Rik@chrY(-)</td>\n",
       "      <td>G530011O06Rik@chrY(-)</td>\n",
       "      <td>NR</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84247</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90796710</td>\n",
       "      <td>90827734</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>GB_RefseqNM#GB_RefseqNR1159</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>90827734</td>\n",
       "      <td>Erdr1y</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>NM#NR</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84248</th>\n",
       "      <td>chrY</td>\n",
       "      <td>90796710</td>\n",
       "      <td>90827734</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>GB_RefseqNM#GB_RefseqNR1159</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>chrY</td>\n",
       "      <td>...</td>\n",
       "      <td>90820875</td>\n",
       "      <td>Gm33272</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>Erdr1</td>\n",
       "      <td>NM#NR</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84249 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end blank score strand  \\\n",
       "0      chr1   3224554   3224816     .  1000      -   \n",
       "1      chr1   3273138   3274559     .  1000      -   \n",
       "2      chr1   3275710   3276991     .  1000      -   \n",
       "3      chr1   3284704   3741721     .  1000      -   \n",
       "4      chr1   3536809   3583776     .  1000      +   \n",
       "...     ...       ...       ...   ...   ...    ...   \n",
       "84244  chrY  90614769  90617133     .  1000      -   \n",
       "84245  chrY  90676614  90678894     .  1000      -   \n",
       "84246  chrY  90762408  90766319     .  1000      -   \n",
       "84247  chrY  90796710  90827734     .  1000      +   \n",
       "84248  chrY  90796710  90827734     .  1000      +   \n",
       "\n",
       "                            source feature frame mgi_chr  ...   mgi_end  \\\n",
       "0           Genebody_mm10_lnc48261    exon     .       .  ...        -1   \n",
       "1           Genebody_mm10_lnc48261    exon     .       .  ...        -1   \n",
       "2           Genebody_mm10_lnc48261    exon     .       .  ...        -1   \n",
       "3                 GB_RefseqNM19801    exon     .    chr1  ...   3729127   \n",
       "4                     EnsmblNR4698    exon     .    chr1  ...   3583776   \n",
       "...                            ...     ...   ...     ...  ...       ...   \n",
       "84244                 EnsmblNR4698    exon     .    chrY  ...  90617133   \n",
       "84245                 EnsmblNR4698    exon     .    chrY  ...  90678894   \n",
       "84246              GB_RefseqNR2077    exon     .    chrY  ...  90766736   \n",
       "84247  GB_RefseqNM#GB_RefseqNR1159    exon     .    chrY  ...  90827734   \n",
       "84248  GB_RefseqNM#GB_RefseqNR1159    exon     .    chrY  ...  90820875   \n",
       "\n",
       "           mgi_symbol mgi_score  mgi_strand                gene_id  \\\n",
       "0                   .        -1           .   lnc_inter_chr1_15560   \n",
       "1                   .        -1           .   lnc_inter_chr1_15561   \n",
       "2                   .        -1           .   lnc_inter_chr1_15562   \n",
       "3             Gm19938      1000           -                   Xkr4   \n",
       "4              Gm1992      1000           +                 Gm1992   \n",
       "...               ...       ...         ...                    ...   \n",
       "84244         Gm28300      1000           -                Gm28300   \n",
       "84245         Gm28301      1000           -                Gm28301   \n",
       "84246  G530011O06Riky      1000           -  G530011O06Rik@chrY(-)   \n",
       "84247          Erdr1y      1000           +                  Erdr1   \n",
       "84248         Gm33272      1000           +                  Erdr1   \n",
       "\n",
       "               transcript_id              gene_name gene_biotype  \\\n",
       "0       lnc_inter_chr1_15560               lnc15560       lncRNA   \n",
       "1       lnc_inter_chr1_15561               lnc15561       lncRNA   \n",
       "2       lnc_inter_chr1_15562               lnc15562       lncRNA   \n",
       "3                       Xkr4                   Xkr4           NM   \n",
       "4                     Gm1992                 Gm1992    antisense   \n",
       "...                      ...                    ...          ...   \n",
       "84244                Gm28300                Gm28300      lincRNA   \n",
       "84245                Gm28301                Gm28301      lincRNA   \n",
       "84246  G530011O06Rik@chrY(-)  G530011O06Rik@chrY(-)           NR   \n",
       "84247                  Erdr1                  Erdr1        NM#NR   \n",
       "84248                  Erdr1                  Erdr1        NM#NR   \n",
       "\n",
       "      NR_annotation NR_annotation2  \n",
       "0          lnc15560            NaN  \n",
       "1          lnc15561            NaN  \n",
       "2          lnc15562            NaN  \n",
       "3              None           None  \n",
       "4              None           None  \n",
       "...             ...            ...  \n",
       "84244          None           None  \n",
       "84245          None           None  \n",
       "84246          None           None  \n",
       "84247          None           None  \n",
       "84248          None           None  \n",
       "\n",
       "[84249 rows x 21 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract information from the 'attributes' column\n",
    "attributes = RAW_Master['attributes'].str.split(';', expand=True)\n",
    "\n",
    "attributes\n",
    "\n",
    "\n",
    "# Extract specific attributes you're interested in\n",
    "RAW_Master['gene_id'] = attributes[0].str.strip().str.split(' ', expand=True)[1].str.strip('\"')\n",
    "RAW_Master['transcript_id'] = attributes[1].str.strip().str.split(' ', expand=True)[1].str.strip('\"')\n",
    "RAW_Master['gene_name'] = attributes[2].str.strip().str.split(' ', expand=True)[1].str.strip('\"')\n",
    "RAW_Master['gene_biotype'] = attributes[4].str.strip().str.split(' ', expand=True)[1].str.strip('\"')\n",
    "RAW_Master['NR_annotation'] = attributes[5].str.strip().str.split(' ', expand=True)[1].str.strip('\"')\n",
    "RAW_Master['NR_annotation2'] = RAW_Master['NR_annotation'].str.split('_').str[1]\n",
    "\n",
    "# Drop the original 'attributes' column\n",
    "RAW_Master.drop(columns=['attributes'], inplace=True)\n",
    "\n",
    "RAW_Master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['lncRNA', 'NM', 'antisense', 'lincRNA', 'NM#NR', 'NR',\n",
       "       'lncOfInterest', None], dtype=object)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check unique gene biotypes in RAW_Master\n",
    "RAW_Master['gene_biotype'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>blank</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>frame</th>\n",
       "      <th>mgi_chr</th>\n",
       "      <th>...</th>\n",
       "      <th>mgi_end</th>\n",
       "      <th>mgi_symbol</th>\n",
       "      <th>mgi_score</th>\n",
       "      <th>mgi_strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>NR_annotation</th>\n",
       "      <th>NR_annotation2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>80642</th>\n",
       "      <td>chrM</td>\n",
       "      <td>2750</td>\n",
       "      <td>3707</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064341</td>\n",
       "      <td>ENSMUSG00000064341</td>\n",
       "      <td>mt-Nd1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80643</th>\n",
       "      <td>chrM</td>\n",
       "      <td>3913</td>\n",
       "      <td>4951</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064345</td>\n",
       "      <td>ENSMUSG00000064345</td>\n",
       "      <td>mt-Nd2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80644</th>\n",
       "      <td>chrM</td>\n",
       "      <td>5327</td>\n",
       "      <td>6872</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064351</td>\n",
       "      <td>ENSMUSG00000064351</td>\n",
       "      <td>mt-Co1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80645</th>\n",
       "      <td>chrM</td>\n",
       "      <td>7012</td>\n",
       "      <td>7696</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064354</td>\n",
       "      <td>ENSMUSG00000064354</td>\n",
       "      <td>mt-Co2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80646</th>\n",
       "      <td>chrM</td>\n",
       "      <td>7765</td>\n",
       "      <td>7969</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064356</td>\n",
       "      <td>ENSMUSG00000064356</td>\n",
       "      <td>mt-Atp8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80647</th>\n",
       "      <td>chrM</td>\n",
       "      <td>7926</td>\n",
       "      <td>8607</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064357</td>\n",
       "      <td>ENSMUSG00000064357</td>\n",
       "      <td>mt-Atp6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80648</th>\n",
       "      <td>chrM</td>\n",
       "      <td>8606</td>\n",
       "      <td>9390</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064358</td>\n",
       "      <td>ENSMUSG00000064358</td>\n",
       "      <td>mt-Co3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80649</th>\n",
       "      <td>chrM</td>\n",
       "      <td>9458</td>\n",
       "      <td>9806</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064360</td>\n",
       "      <td>ENSMUSG00000064360</td>\n",
       "      <td>mt-Nd3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80650</th>\n",
       "      <td>chrM</td>\n",
       "      <td>9876</td>\n",
       "      <td>10173</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000065947</td>\n",
       "      <td>ENSMUSG00000065947</td>\n",
       "      <td>mt-Nd4l</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80651</th>\n",
       "      <td>chrM</td>\n",
       "      <td>10166</td>\n",
       "      <td>11544</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064363</td>\n",
       "      <td>ENSMUSG00000064363</td>\n",
       "      <td>mt-Nd4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80652</th>\n",
       "      <td>chrM</td>\n",
       "      <td>11741</td>\n",
       "      <td>13565</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064367</td>\n",
       "      <td>ENSMUSG00000064367</td>\n",
       "      <td>mt-Nd5</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80653</th>\n",
       "      <td>chrM</td>\n",
       "      <td>13551</td>\n",
       "      <td>14070</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064368</td>\n",
       "      <td>ENSMUSG00000064368</td>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80654</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14144</td>\n",
       "      <td>15288</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>-1</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064370</td>\n",
       "      <td>ENSMUSG00000064370</td>\n",
       "      <td>mt-Cytb</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  start    end blank score strand source feature frame mgi_chr  \\\n",
       "80642  chrM   2750   3707     .     .      +  insdc    exon     .       .   \n",
       "80643  chrM   3913   4951     .     .      +  insdc    exon     .       .   \n",
       "80644  chrM   5327   6872     .     .      +  insdc    exon     .       .   \n",
       "80645  chrM   7012   7696     .     .      +  insdc    exon     .       .   \n",
       "80646  chrM   7765   7969     .     .      +  insdc    exon     .       .   \n",
       "80647  chrM   7926   8607     .     .      +  insdc    exon     .       .   \n",
       "80648  chrM   8606   9390     .     .      +  insdc    exon     .       .   \n",
       "80649  chrM   9458   9806     .     .      +  insdc    exon     .       .   \n",
       "80650  chrM   9876  10173     .     .      +  insdc    exon     .       .   \n",
       "80651  chrM  10166  11544     .     .      +  insdc    exon     .       .   \n",
       "80652  chrM  11741  13565     .     .      +  insdc    exon     .       .   \n",
       "80653  chrM  13551  14070     .     .      -  insdc    exon     .       .   \n",
       "80654  chrM  14144  15288     .     .      +  insdc    exon     .       .   \n",
       "\n",
       "       ...  mgi_end  mgi_symbol mgi_score  mgi_strand             gene_id  \\\n",
       "80642  ...       -1           .        -1           .  ENSMUSG00000064341   \n",
       "80643  ...       -1           .        -1           .  ENSMUSG00000064345   \n",
       "80644  ...       -1           .        -1           .  ENSMUSG00000064351   \n",
       "80645  ...       -1           .        -1           .  ENSMUSG00000064354   \n",
       "80646  ...       -1           .        -1           .  ENSMUSG00000064356   \n",
       "80647  ...       -1           .        -1           .  ENSMUSG00000064357   \n",
       "80648  ...       -1           .        -1           .  ENSMUSG00000064358   \n",
       "80649  ...       -1           .        -1           .  ENSMUSG00000064360   \n",
       "80650  ...       -1           .        -1           .  ENSMUSG00000065947   \n",
       "80651  ...       -1           .        -1           .  ENSMUSG00000064363   \n",
       "80652  ...       -1           .        -1           .  ENSMUSG00000064367   \n",
       "80653  ...       -1           .        -1           .  ENSMUSG00000064368   \n",
       "80654  ...       -1           .        -1           .  ENSMUSG00000064370   \n",
       "\n",
       "            transcript_id gene_name gene_biotype NR_annotation NR_annotation2  \n",
       "80642  ENSMUSG00000064341    mt-Nd1         None          None           None  \n",
       "80643  ENSMUSG00000064345    mt-Nd2         None          None           None  \n",
       "80644  ENSMUSG00000064351    mt-Co1         None          None           None  \n",
       "80645  ENSMUSG00000064354    mt-Co2         None          None           None  \n",
       "80646  ENSMUSG00000064356   mt-Atp8         None          None           None  \n",
       "80647  ENSMUSG00000064357   mt-Atp6         None          None           None  \n",
       "80648  ENSMUSG00000064358    mt-Co3         None          None           None  \n",
       "80649  ENSMUSG00000064360    mt-Nd3         None          None           None  \n",
       "80650  ENSMUSG00000065947   mt-Nd4l         None          None           None  \n",
       "80651  ENSMUSG00000064363    mt-Nd4         None          None           None  \n",
       "80652  ENSMUSG00000064367    mt-Nd5         None          None           None  \n",
       "80653  ENSMUSG00000064368    mt-Nd6         None          None           None  \n",
       "80654  ENSMUSG00000064370   mt-Cytb         None          None           None  \n",
       "\n",
       "[13 rows x 21 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identify rows where gene_biotype is the string 'None'\n",
    "RAW_Master[RAW_Master['gene_biotype'].astype(str) == 'None']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['lncRNA', 'NM', 'antisense', 'lincRNA', 'NM#NR', 'NR',\n",
       "       'lncOfInterest', 'mitochondrial protein-coding gene'], dtype=object)"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Replace NaNs in gene_biotype with 'mitochondrial protein-coding gene' and check updated unique values\n",
    "RAW_Master.loc[RAW_Master['gene_biotype'].isna(), 'gene_biotype'] = 'mitochondrial protein-coding gene'\n",
    "RAW_Master['gene_biotype'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 2. Export Processed Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "RAW_Master.to_csv('MGI_and_Karri_Annotations_Restructured_mm39.txt', sep='\\t', index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "SCVI New",
   "language": "python",
   "name": "scvi-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
