---
title: "R Notebook"
output: html_notebook
---

## Load the R packages that will be needed for the analyis of the mRNA sequencing data.
use `install.packages` or see [bioconductor](https://www.bioconductor.org/) for more 
information on package installation. Version info is listed at the bottom.
```{r}
# R v4.1.2
suppressPackageStartupMessages({
  library(CellChat) #1.5.0
  library(patchwork) #1.1.2
  library(Matrix) #1.5.1
  library(NMF) #0.24.0
  library(ggalluvial) #0.12.3
  library(cowplot) #1.1.1
  library(pheatmap) #1.0.12
})
```

### Custom function to draw the score heatmap - modified from CellChat source code
```{r}
netAnalysis_CustomContributions <- function(object, signaling = NULL, pattern = c("outgoing", "incoming","all"), slot.name = "netP"){
  pattern <- match.arg(pattern)
  if (length(slot(object, slot.name)$centr) == 0) {
    stop("Please run `netAnalysis_computeCentrality` to compute the network centrality scores! ")
  }
  centr <- slot(object, slot.name)$centr
  outgoing <- matrix(0, nrow = nlevels(object@idents), ncol = length(centr))
  incoming <- matrix(0, nrow = nlevels(object@idents), ncol = length(centr))
  dimnames(outgoing) <- list(levels(object@idents), names(centr))
  dimnames(incoming) <- dimnames(outgoing)
  for (i in 1:length(centr)) {
    outgoing[,i] <- centr[[i]]$outdeg
    incoming[,i] <- centr[[i]]$indeg
  }
  if (pattern == "outgoing") {
    mat <- t(outgoing)
    legend.name <- "Outgoing"
  } else if (pattern == "incoming") {
    mat <- t(incoming)
    legend.name <- "Incoming"
  } else if (pattern == "all") {
    mat <- t(outgoing + incoming)
    legend.name <- "Overall"
  }
  
  mat.ori <- mat
  mat <- sweep(mat, 1L, apply(mat, 1, max), '/', check.margin = FALSE)
  mat[mat == 0] <- NA
  
  pSum <- rowSums(mat.ori)
  pSum.original <- pSum
  pSum <- -1/log(pSum)
  pSum[is.na(pSum)] <- 0
  idx1 <- which(is.infinite(pSum) | pSum < 0)
  if (length(idx1) > 0) {
    values.assign <- seq(max(pSum)*1.1, max(pSum)*1.5, length.out = length(idx1))
    position <- sort(pSum.original[idx1], index.return = TRUE)$ix
    pSum[idx1] <- values.assign[match(1:length(idx1), position)]
  }
  
  return(pSum)
}

range01 <- function(x){(x-min(x))/(max(x)-min(x))}
```


# Import data which can be download from the [Broad Single Cell Portal (SCP1871)](https://singlecell.broadinstitute.org/single_cell/study/SCP1871/dose-response-single-nuclei-rna-sequencing-of-livers-from-male-mice-treated-with-2-3-7-8-tetrachlorodibenzo-p-dioxin-tcdd#study-download)

```{r}
expMat <- readMM(gzfile("./expression_matrix.mtx.gz"))
barcodes <- read.csv(gzfile("./barcodes.tsv.gz"), sep = '\t', header = FALSE)
genes <- read.csv(gzfile("./genes.tsv.gz"), sep = '\t', header = FALSE)
metadata <- read.csv("./DR-metadata.tsv", sep = '\t', header = TRUE)
metadata <- metadata[-1,]
rownames(expMat) <- genes$V1
rownames(metadata) <- metadata$NAME
colnames(expMat) <- barcodes$V1
metadata$labels <- paste0('label_', metadata$cell_type__ontology_label)
```

# For each dose group run CellChat. 
Code for CellChat analysis was modified from vignettes [Inference and analysis of cell-cell communication using CellChat](https://htmlpreview.github.io/?https://github.com/sqjin/CellChat/blob/master/tutorial/CellChat-vignette.html) and [Comparison analysis of multiple datasets using CellChat](https://htmlpreview.github.io/?https://github.com/sqjin/CellChat/blob/master/tutorial/CellChat-vignette.html)

Analysis may take several hours
```{r}
object.list <- list()

for (d in unique(metadata$dose)){
  print(d)
  cells.use = rownames(metadata)[metadata$dose == d]
  # Prepare input data for CellChat
  data.input = expMat[,cells.use]
  meta <- metadata[cells.use,]

  cellchat <- createCellChat(object = data.input, meta = meta, group.by = "labels")

  cellchat <- addMeta(cellchat, meta = meta)
  cellchat <- setIdent(cellchat, ident.use = "labels") # set "labels" as default cell identity
  levels(cellchat@idents) # show factor levels of the cell labels
  groupSize <- as.numeric(table(cellchat@idents)) # number of cells in each cell group

  
  CellChatDB <- CellChatDB.mouse # use CellChatDB.mouse if running on mouse data
  showDatabaseCategory(CellChatDB)

  # use a subset of CellChatDB for cell-cell communication analysis
  CellChatDB.use <- subsetDB(CellChatDB, search = "Secreted Signaling") # use Secreted Signaling
 
  # set the used database in the object
  cellchat@DB <- CellChatDB.use

  print('subsetting')
  # subset the expression data of signaling genes for saving computation cost
  cellchat <- subsetData(cellchat) # This step is necessary even if using the whole database
  future::plan("multiprocess", workers = 4) # do parallel

  cellchat <- identifyOverExpressedGenes(cellchat)
  cellchat <- identifyOverExpressedInteractions(cellchat)

  print('prob comp')
  cellchat <- computeCommunProb(cellchat, population.size = TRUE)
  # Filter out the cell-cell communication if there are only few number of cells in certain cell groups
  cellchat <- filterCommunication(cellchat, min.cells = 10)

  cellchat <- computeCommunProbPathway(cellchat)

  cellchat <- aggregateNet(cellchat)
  
  cellchat <- netAnalysis_computeCentrality(cellchat)
  
  print('saving')
  object.list[[paste0('Dose_',d)]] <- cellchat
  saveRDS(cellchat, file = paste0('cellchat_', d, '.RData'))
  print('moving on')
}
```

## Use this to load pre-computed CellChat Objects (see above)
```{r eval=FALSE}
object.list <- list()
dg <- c(0, 0.01, 0.03, 0.1, 0.3, 1, 3, 10, 30)
for (f in dg){
  fname = paste0('cellchat_', f, '.RData')
  print(fname)
  cc <- readRDS(fname)
  object.list[[paste0('D', f)]] <- cc
}
```


# Compare the total number of interactions and interaction strength
To answer on question on whether the cell-cell communication is enhanced or not, CellChat compares the the total number of interactions and interaction strength of the inferred cell-cell communication networks from different biological conditions.

```{r}
cellchat <- mergeCellChat(object.list, add.names = names(object.list))
cellchat1 <- mergeCellChat(object.list[c(1,2)], add.names = names(object.list[c(1,2)])) # 0.01 vs veh
cellchat2 <- mergeCellChat(object.list[c(1,3)], add.names = names(object.list[c(1,3)])) # 0.03 vs veh
cellchat3 <- mergeCellChat(object.list[c(1,4)], add.names = names(object.list[c(1,4)])) # 0.1 vs veh
cellchat4 <- mergeCellChat(object.list[c(1,5)], add.names = names(object.list[c(1,5)])) # 0.3 vs veh
cellchat5 <- mergeCellChat(object.list[c(1,6)], add.names = names(object.list[c(1,6)])) # 1 vs veh
cellchat6 <- mergeCellChat(object.list[c(1,7)], add.names = names(object.list[c(1,7)])) # 3 vs veh
cellchat7 <- mergeCellChat(object.list[c(1,8)], add.names = names(object.list[c(1,8)])) # 10 vs veh
cellchat8 <- mergeCellChat(object.list[c(1,9)], add.names = names(object.list[c(1,9)])) # 30 vs veh
```

## Figure 1A
```{r fig.width = 8, fig.height = 3}
NumInteractions <- compareInteractions(cellchat, show.legend = F, group = c(1:9))
StrengthInteractions <- compareInteractions(cellchat, show.legend = F, group = c(1:9), measure = "weight")

NumInteractions + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
StrengthInteractions + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```

## Circle plots of the interaction numbers between each cell type (Figure 1B).
```{r fig.height = 12, fig.width = 12}
weight.max <- getMaxWeight(object.list, attribute = c("idents","count"))
par(mfrow = c(3,3), xpd=TRUE)

j = 0
for (i in 1:length(object.list)) {
  j = j + 1
  netVisual_circle(object.list[[j]]@net$count, weight.scale = T, 
                   label.edge= F, edge.weight.max = weight.max[2], edge.width.max = 12, 
                   title.name = paste0("Number of interactions - ", names(object.list)[j]))
}
```


## Figure S1
```{r fig.width = 12, fig.height = 5}
ht1 <- netVisual_heatmap(object.list[[1]], color.heatmap = "Reds")
ht2 <- netVisual_heatmap(object.list[[2]], color.heatmap = "Reds")
ht3 <- netVisual_heatmap(object.list[[3]], color.heatmap = "Reds")
ht4 <- netVisual_heatmap(object.list[[4]], color.heatmap = "Reds")
ht5 <- netVisual_heatmap(object.list[[5]], color.heatmap = "Reds")
ht6 <- netVisual_heatmap(object.list[[6]], color.heatmap = "Reds")
ht7 <- netVisual_heatmap(object.list[[7]], color.heatmap = "Reds")
ht8 <- netVisual_heatmap(object.list[[8]], color.heatmap = "Reds")
ht9 <- netVisual_heatmap(object.list[[9]], color.heatmap = "Reds")

ht1 + ht2 + ht3
ht4 + ht5 + ht6
ht7 + ht8 + ht9
```

## Figure 2
```{r fig.width = 12, fig.height = 5}
incoming.list <- list()
outgoing.list <- list()
for (ds in names(object.list)){
  ht1 <- CellChat::netAnalysis_signalingRole_heatmap(object.list[[ds]], pattern = "outgoing")
  ht2 <- CellChat::netAnalysis_signalingRole_heatmap(object.list[[ds]], pattern = "incoming")
  outgoing.list[[ds]] <- netAnalysis_CustomContributions(object.list[[ds]], pattern = 'outgoing')
  incoming.list[[ds]] <- netAnalysis_CustomContributions(object.list[[ds]], pattern = 'all')
    print(ht1 + ht2)
}

mat_in <- data.frame(Reduce(bind_rows, incoming.list))
mat_in[is.na(mat_in)] <- 0
rownames(mat_in) <- c('0','0.01', '0.03', '0.1', '0.3', '1', '3', '10', '30')

pheatmap::pheatmap(range01(scale(t(mat_in))), cluster_cols = FALSE, cluster_rows = TRUE, color = colorRampPalette(colors = c("white", "#014f16"))(100),
         show_colnames = TRUE, border_color = 'black')

```

## Figure S2
### Compare the major sources and targets in 2D space
Comparing the outgoing and incoming interaction strength in 2D space allows ready identification of the cell populations with significant changes in sending or receiving signals between different datasets.

```{r fig.height = 10, fig.width = 16}
num.link <- sapply(object.list, function(x) {rowSums(x@net$count) + colSums(x@net$count)-diag(x@net$count)})
weight.MinMax <- c(min(num.link), max(num.link)) # control the dot size in the different datasets
gg <- list()
for (i in 1:length(object.list)) {
  gg[[i]] <- netAnalysis_signalingRole_scatter(object.list[[i]], title = names(object.list)[i], weight.MinMax = weight.MinMax) + xlim(0,0.15) + ylim(0, 0.15)
}

suppressMessages({
  patchwork::wrap_plots(plots = gg)
})
```

## Figure S3
Furthermore, we can identify the specific signaling changes of individual cell types between treatment groups.
```{r fig.width = 10, fig.height = 10}
i = 0
pl.list <- list()
#for(ds in c(cellchat1, cellchat2, cellchat3, cellchat4, cellchat5, cellchat6, cellchat7, cellchat8)){
for (ct in c('label_centrilobular region hepatocyte', 'label_periportal region hepatocyte', 'label_endothelial cell of hepatic sinusoid', 'label_macrophage')){
  for(ds in c(cellchat4, cellchat6, cellchat8)){ # doses 0.3, 3, and 30
    i = i + 1
    pl.list[[paste0(i,ct)]] <- netAnalysis_signalingChanges_scatter(ds, idents.use = ct) + ylim(-0.02, 0.02) + xlim(-0.02, 0.02) + theme(legend.position="none")
  }
}

(patchwork::wrap_plots(plots = pl.list, ncol = 3))

```

#Figure S4
```{r fig.height = 8}
# Compute the network centrality scores
pathways.show <- c('PARs')
pl.list <- list()
par(mfrow = c(3,3))

for (ds in names(object.list)){
  pl.list[[ds]] <- netAnalysis_contribution(object.list[[ds]], signaling = pathways.show)
  netVisual_chord_gene(object.list[[1]], sources.use = c(1:16), targets.use = c(1:16), signaling = pathways.show, legend.pos.x = 15)
}

wrap_plots(pl.list)

plotGeneExpression(cellchat8, signaling = "PARs", split.by = "datasets", colors.ggplot = T)
```

#Figure 3
```{r fig.height = 8}
# Compute the network centrality scores
pathways.show <- c('NRG')
pl.list <- list()
par(mfrow = c(3,3))

for (ds in names(object.list)){
  pl.list[[ds]] <- netAnalysis_contribution(object.list[[ds]], signaling = pathways.show)
  netVisual_chord_gene(object.list[[1]], sources.use = c(1:16), targets.use = c(1:16), signaling = pathways.show, legend.pos.x = 15)
}

wrap_plots(pl.list)

plotGeneExpression(cellchat8, signaling = "PARs", split.by = "datasets", colors.ggplot = T)
```

```{r}
sessionInfo()
```

