{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Concatonate All Deseq2 Data with Karri RefGenome Metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 1. Import Required Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 2. Import Deseq2 Data and All Other Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>pDRE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc10851</td>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tenm3</td>\n",
       "      <td>249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Auts2l1</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Camta1</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Cdh13</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene  pDRE\n",
       "1  lnc10851   273\n",
       "2     Tenm3   249\n",
       "3   Auts2l1   193\n",
       "4    Camta1   179\n",
       "5     Cdh13   141"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pDREs_Master = pd.read_csv('../../99_rn7_DREs/02_CleanUp_DRE_BED_File/DRE_Counts_by_Gene.txt', \n",
    "                            delimiter = '\\t')\n",
    "\n",
    "pDREs_Master = pDREs_Master.dropna()\n",
    "\n",
    "pDREs_Master = pDREs_Master.rename(columns={'gene_id': 'Gene', 'count': 'pDRE'})\n",
    "\n",
    "pDREs_Master.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>AHR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc10851</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc16454</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc4726</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Prkce</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene   AHR\n",
       "0       NaN  3089\n",
       "1  lnc10851   144\n",
       "2  lnc16454    41\n",
       "3   lnc4726    23\n",
       "4     Prkce    19"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AHR_Master = pd.read_csv('../../99_rn7_AHR_Enrichment/02_CleanUp_BED_File/rn7_AHR_Binding_Counts.txt', \n",
    "                            delimiter = '\\t')\n",
    "\n",
    "AHR_Master = AHR_Master.rename(columns={'gene_id': 'Gene', 'count': 'AHR'})\n",
    "\n",
    "AHR_Master.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>15038462</td>\n",
       "      <td>15039082</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"0\"; gene_id \"lnc1\"; transcript_type ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>14766774</td>\n",
       "      <td>14767820</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"1\"; gene_id \"lnc10\"; transcript_type...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>103001424</td>\n",
       "      <td>103005924</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"3\"; gene_id \"lnc1000\"; transcript_ty...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>79676628</td>\n",
       "      <td>79677257</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"4\"; gene_id \"lnc10000\"; transcript_t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>79677716</td>\n",
       "      <td>79678362</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"5\"; gene_id \"lnc10001\"; transcript_t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182956</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>293338</td>\n",
       "      <td>293506</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Rat_ID \"1141644\"; gene_id \"LOC103694730\"; tran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182957</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>300877</td>\n",
       "      <td>300988</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Rat_ID \"1141645\"; gene_id \"LOC103694730\"; tran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182958</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>317993</td>\n",
       "      <td>319851</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Rat_ID \"1141646\"; gene_id \"LOC120099645\"; tran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182959</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>320383</td>\n",
       "      <td>320541</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Rat_ID \"1141647\"; gene_id \"LOC120099645\"; tran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182960</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>321212</td>\n",
       "      <td>321506</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Rat_ID \"1141648\"; gene_id \"LOC120099645\"; tran...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1182961 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              chrom      source feature      start        end  \\\n",
       "0                              chr5    liftover    exon   15038462   15039082   \n",
       "1                              chr5    liftover    exon   14766774   14767820   \n",
       "2                             chr13    liftover    exon  103001424  103005924   \n",
       "3                             chr10    liftover    exon   79676628   79677257   \n",
       "4                             chr10    liftover    exon   79677716   79678362   \n",
       "...                             ...         ...     ...        ...        ...   \n",
       "1182956  chrY_NW_023637722v1_random  ncbiRefSeq    exon     293338     293506   \n",
       "1182957  chrY_NW_023637722v1_random  ncbiRefSeq    exon     300877     300988   \n",
       "1182958  chrY_NW_023637722v1_random  ncbiRefSeq    exon     317993     319851   \n",
       "1182959  chrY_NW_023637722v1_random  ncbiRefSeq    exon     320383     320541   \n",
       "1182960  chrY_NW_023637722v1_random  ncbiRefSeq    exon     321212     321506   \n",
       "\n",
       "         score strand frame                                          attribute  \n",
       "0         1000      +     .  Mouse_ID \"0\"; gene_id \"lnc1\"; transcript_type ...  \n",
       "1         1000      +     .  Mouse_ID \"1\"; gene_id \"lnc10\"; transcript_type...  \n",
       "2         1000      +     .  Mouse_ID \"3\"; gene_id \"lnc1000\"; transcript_ty...  \n",
       "3         1000      -     .  Mouse_ID \"4\"; gene_id \"lnc10000\"; transcript_t...  \n",
       "4         1000      -     .  Mouse_ID \"5\"; gene_id \"lnc10001\"; transcript_t...  \n",
       "...        ...    ...   ...                                                ...  \n",
       "1182956   1000      +     .  Rat_ID \"1141644\"; gene_id \"LOC103694730\"; tran...  \n",
       "1182957   1000      +     .  Rat_ID \"1141645\"; gene_id \"LOC103694730\"; tran...  \n",
       "1182958   1000      -     .  Rat_ID \"1141646\"; gene_id \"LOC120099645\"; tran...  \n",
       "1182959   1000      -     .  Rat_ID \"1141647\"; gene_id \"LOC120099645\"; tran...  \n",
       "1182960   1000      -     .  Rat_ID \"1141648\"; gene_id \"LOC120099645\"; tran...  \n",
       "\n",
       "[1182961 rows x 9 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define column names based on GTF format\n",
    "gtf_columns = [\n",
    "    \"chrom\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"space1\", \"space2\", \"attribute\"\n",
    "]\n",
    "\n",
    "Annotations_Master = pd.read_csv('../../00_Karri_et_al_GTF/Karri_Annotations_Restructured_rn7_for_Gene_Expression.gtf', \n",
    "                                delimiter = '\\t',\n",
    "                                names=gtf_columns)  # Assign column names)\n",
    "\n",
    "# Drop unwanted columns\n",
    "Annotations_Master = Annotations_Master.drop(columns=[\"space1\", \"space2\"])\n",
    "\n",
    "Annotations_Master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>Rat_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>lnc1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>lnc10</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>lnc1000</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>lnc10000</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>lnc10001</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182956</th>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC103694730</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>1141644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182957</th>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC103694730</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>1141645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182958</th>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC120099645</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>1141646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182959</th>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC120099645</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>1141647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182960</th>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC120099645</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>1141648</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1182961 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Mouse_ID       gene_id   transcript_type   Rat_ID\n",
       "0              0          lnc1            lncRNA      NaN\n",
       "1              1         lnc10            lncRNA      NaN\n",
       "2              3       lnc1000            lncRNA      NaN\n",
       "3              4      lnc10000            lncRNA      NaN\n",
       "4              5      lnc10001            lncRNA      NaN\n",
       "...          ...           ...               ...      ...\n",
       "1182956      NaN  LOC103694730  predicted_lncRNA  1141644\n",
       "1182957      NaN  LOC103694730  predicted_lncRNA  1141645\n",
       "1182958      NaN  LOC120099645  predicted_lncRNA  1141646\n",
       "1182959      NaN  LOC120099645  predicted_lncRNA  1141647\n",
       "1182960      NaN  LOC120099645  predicted_lncRNA  1141648\n",
       "\n",
       "[1182961 rows x 4 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define a function to parse attributes and extract key-value pairs\n",
    "def parse_attributes(attribute_string):\n",
    "    attr_dict = {}\n",
    "    if pd.notnull(attribute_string):  # Check if the string is not NaN\n",
    "        attributes = attribute_string.split(';')\n",
    "        for attribute in attributes:\n",
    "            if attribute.strip():  # Ensure it's not an empty string\n",
    "                key, value = attribute.strip().split(' ', 1)\n",
    "                attr_dict[key] = value.strip('\"')  # Strip quotes from value\n",
    "    return attr_dict\n",
    "\n",
    "# Apply the function to extract the information into a DataFrame\n",
    "attributes_df = Annotations_Master['attribute'].apply(parse_attributes).apply(pd.Series)\n",
    "\n",
    "attributes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>Rat_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>15038462</td>\n",
       "      <td>15039082</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"0\"; gene_id \"lnc1\"; transcript_type ...</td>\n",
       "      <td>0</td>\n",
       "      <td>lnc1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>14766774</td>\n",
       "      <td>14767820</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"1\"; gene_id \"lnc10\"; transcript_type...</td>\n",
       "      <td>1</td>\n",
       "      <td>lnc10</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr13</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>103001424</td>\n",
       "      <td>103005924</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"3\"; gene_id \"lnc1000\"; transcript_ty...</td>\n",
       "      <td>3</td>\n",
       "      <td>lnc1000</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>79676628</td>\n",
       "      <td>79677257</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"4\"; gene_id \"lnc10000\"; transcript_t...</td>\n",
       "      <td>4</td>\n",
       "      <td>lnc10000</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>liftover</td>\n",
       "      <td>exon</td>\n",
       "      <td>79677716</td>\n",
       "      <td>79678362</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Mouse_ID \"5\"; gene_id \"lnc10001\"; transcript_t...</td>\n",
       "      <td>5</td>\n",
       "      <td>lnc10001</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182956</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>293338</td>\n",
       "      <td>293506</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Rat_ID \"1141644\"; gene_id \"LOC103694730\"; tran...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC103694730</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>1141644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182957</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>300877</td>\n",
       "      <td>300988</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Rat_ID \"1141645\"; gene_id \"LOC103694730\"; tran...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC103694730</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>1141645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182958</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>317993</td>\n",
       "      <td>319851</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Rat_ID \"1141646\"; gene_id \"LOC120099645\"; tran...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC120099645</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>1141646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182959</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>320383</td>\n",
       "      <td>320541</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Rat_ID \"1141647\"; gene_id \"LOC120099645\"; tran...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC120099645</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>1141647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1182960</th>\n",
       "      <td>chrY_NW_023637722v1_random</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>321212</td>\n",
       "      <td>321506</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Rat_ID \"1141648\"; gene_id \"LOC120099645\"; tran...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC120099645</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>1141648</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1182961 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              chrom      source feature      start        end  \\\n",
       "0                              chr5    liftover    exon   15038462   15039082   \n",
       "1                              chr5    liftover    exon   14766774   14767820   \n",
       "2                             chr13    liftover    exon  103001424  103005924   \n",
       "3                             chr10    liftover    exon   79676628   79677257   \n",
       "4                             chr10    liftover    exon   79677716   79678362   \n",
       "...                             ...         ...     ...        ...        ...   \n",
       "1182956  chrY_NW_023637722v1_random  ncbiRefSeq    exon     293338     293506   \n",
       "1182957  chrY_NW_023637722v1_random  ncbiRefSeq    exon     300877     300988   \n",
       "1182958  chrY_NW_023637722v1_random  ncbiRefSeq    exon     317993     319851   \n",
       "1182959  chrY_NW_023637722v1_random  ncbiRefSeq    exon     320383     320541   \n",
       "1182960  chrY_NW_023637722v1_random  ncbiRefSeq    exon     321212     321506   \n",
       "\n",
       "         score strand frame  \\\n",
       "0         1000      +     .   \n",
       "1         1000      +     .   \n",
       "2         1000      +     .   \n",
       "3         1000      -     .   \n",
       "4         1000      -     .   \n",
       "...        ...    ...   ...   \n",
       "1182956   1000      +     .   \n",
       "1182957   1000      +     .   \n",
       "1182958   1000      -     .   \n",
       "1182959   1000      -     .   \n",
       "1182960   1000      -     .   \n",
       "\n",
       "                                                 attribute Mouse_ID  \\\n",
       "0        Mouse_ID \"0\"; gene_id \"lnc1\"; transcript_type ...        0   \n",
       "1        Mouse_ID \"1\"; gene_id \"lnc10\"; transcript_type...        1   \n",
       "2        Mouse_ID \"3\"; gene_id \"lnc1000\"; transcript_ty...        3   \n",
       "3        Mouse_ID \"4\"; gene_id \"lnc10000\"; transcript_t...        4   \n",
       "4        Mouse_ID \"5\"; gene_id \"lnc10001\"; transcript_t...        5   \n",
       "...                                                    ...      ...   \n",
       "1182956  Rat_ID \"1141644\"; gene_id \"LOC103694730\"; tran...      NaN   \n",
       "1182957  Rat_ID \"1141645\"; gene_id \"LOC103694730\"; tran...      NaN   \n",
       "1182958  Rat_ID \"1141646\"; gene_id \"LOC120099645\"; tran...      NaN   \n",
       "1182959  Rat_ID \"1141647\"; gene_id \"LOC120099645\"; tran...      NaN   \n",
       "1182960  Rat_ID \"1141648\"; gene_id \"LOC120099645\"; tran...      NaN   \n",
       "\n",
       "              gene_id   transcript_type   Rat_ID  \n",
       "0                lnc1            lncRNA      NaN  \n",
       "1               lnc10            lncRNA      NaN  \n",
       "2             lnc1000            lncRNA      NaN  \n",
       "3            lnc10000            lncRNA      NaN  \n",
       "4            lnc10001            lncRNA      NaN  \n",
       "...               ...               ...      ...  \n",
       "1182956  LOC103694730  predicted_lncRNA  1141644  \n",
       "1182957  LOC103694730  predicted_lncRNA  1141645  \n",
       "1182958  LOC120099645  predicted_lncRNA  1141646  \n",
       "1182959  LOC120099645  predicted_lncRNA  1141647  \n",
       "1182960  LOC120099645  predicted_lncRNA  1141648  \n",
       "\n",
       "[1182961 rows x 13 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Concatenate the two DataFrames side by side\n",
    "merged_combined = pd.concat([Annotations_Master, attributes_df], axis=1)\n",
    "merged_combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chrom', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
       "       'frame', 'attribute', 'Mouse_ID', 'gene_id', 'transcript_type',\n",
       "       'Rat_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_combined.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_combined2 = merged_combined[[]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = './Deseq2_Results/'\n",
    "\n",
    "Deseq2_0_01_v_0 = pd.read_csv(directory + '0.01_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_0_03_v_0 = pd.read_csv(directory + '0.03_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_0_1_v_0 = pd.read_csv(directory + '0.1_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_0_3_v_0 = pd.read_csv(directory + '0.3_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_1_v_0 = pd.read_csv(directory + '1_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_3_v_0 = pd.read_csv(directory + '3_vs_0.txt', delimiter = '\\t').reset_index()\n",
    "Deseq2_10_v_0 = pd.read_csv(directory + '10_vs_0.txt', delimiter = '\\t').reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "Deseq2_0_01_v_0['Dose'] = '0.01'\n",
    "Deseq2_0_03_v_0['Dose'] = '0.03'\n",
    "Deseq2_0_1_v_0['Dose'] = '0.1'\n",
    "Deseq2_0_3_v_0['Dose'] = '0.3'\n",
    "Deseq2_1_v_0['Dose'] = '1'\n",
    "Deseq2_3_v_0['Dose'] = '3'\n",
    "Deseq2_10_v_0['Dose'] = '10'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Dose</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc1000</td>\n",
       "      <td>1.021823</td>\n",
       "      <td>1.224408</td>\n",
       "      <td>1.866052</td>\n",
       "      <td>0.656149</td>\n",
       "      <td>0.511728</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc10000</td>\n",
       "      <td>0.046372</td>\n",
       "      <td>-1.398125</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>-0.232795</td>\n",
       "      <td>0.815921</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc10001</td>\n",
       "      <td>0.146825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc10003</td>\n",
       "      <td>0.177874</td>\n",
       "      <td>-1.398177</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>-0.232804</td>\n",
       "      <td>0.815914</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc10004</td>\n",
       "      <td>0.740087</td>\n",
       "      <td>3.883753</td>\n",
       "      <td>3.422047</td>\n",
       "      <td>1.134921</td>\n",
       "      <td>0.256409</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      index  baseMean  log2FoldChange     lfcSE      stat    pvalue  padj Dose\n",
       "0   lnc1000  1.021823        1.224408  1.866052  0.656149  0.511728   NaN   10\n",
       "1  lnc10000  0.046372       -1.398125  6.005823 -0.232795  0.815921   NaN   10\n",
       "2  lnc10001  0.146825        0.000000  6.005823  0.000000  1.000000   NaN   10\n",
       "3  lnc10003  0.177874       -1.398177  6.005823 -0.232804  0.815914   NaN   10\n",
       "4  lnc10004  0.740087        3.883753  3.422047  1.134921  0.256409   NaN   10"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Deseq2_10_v_0.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <br> 3. Concat All Data Into One Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "Deseq2_Master = pd.concat([Deseq2_0_01_v_0,\n",
    "                          Deseq2_0_03_v_0,\n",
    "                          Deseq2_0_1_v_0,\n",
    "                          Deseq2_0_3_v_0,\n",
    "                          Deseq2_1_v_0,\n",
    "                          Deseq2_3_v_0,\n",
    "                          Deseq2_10_v_0],\n",
    "                          ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Dose</th>\n",
       "      <th>FoldChange</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc1000</td>\n",
       "      <td>1.021823</td>\n",
       "      <td>-0.020692</td>\n",
       "      <td>1.991121</td>\n",
       "      <td>-0.010392</td>\n",
       "      <td>0.991708</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.985760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc10000</td>\n",
       "      <td>0.046372</td>\n",
       "      <td>-0.929325</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>-0.154737</td>\n",
       "      <td>0.877028</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.525104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc10001</td>\n",
       "      <td>0.146825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc10003</td>\n",
       "      <td>0.177874</td>\n",
       "      <td>0.032402</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.005395</td>\n",
       "      <td>0.995695</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.022714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc10004</td>\n",
       "      <td>0.740087</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.592880</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Gene  baseMean  log2FoldChange     lfcSE      stat    pvalue  padj  \\\n",
       "0   lnc1000  1.021823       -0.020692  1.991121 -0.010392  0.991708   NaN   \n",
       "1  lnc10000  0.046372       -0.929325  6.005823 -0.154737  0.877028   NaN   \n",
       "2  lnc10001  0.146825        0.000000  6.005823  0.000000  1.000000   NaN   \n",
       "3  lnc10003  0.177874        0.032402  6.005823  0.005395  0.995695   NaN   \n",
       "4  lnc10004  0.740087        0.000000  3.592880  0.000000  1.000000   NaN   \n",
       "\n",
       "   Dose  FoldChange  \n",
       "0  0.01    0.985760  \n",
       "1  0.01    0.525104  \n",
       "2  0.01    1.000000  \n",
       "3  0.01    1.022714  \n",
       "4  0.01    1.000000  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Deseq2_Master = Deseq2_Master.rename(columns={'index': 'Gene'})\n",
    "Deseq2_Master['FoldChange'] = pow(2, Deseq2_Master['log2FoldChange'])\n",
    "\n",
    "Deseq2_Master.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chrom', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
       "       'frame', 'attribute', 'Mouse_ID', 'gene_id', 'transcript_type',\n",
       "       'Rat_ID'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_combined.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_combined2 = merged_combined[['chrom', 'start', 'end','frame', 'score','strand', 'source', 'gene_id', 'transcript_type', 'Mouse_ID', 'Rat_ID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>Dose</th>\n",
       "      <th>FoldChange</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>frame</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>Rat_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc1000</td>\n",
       "      <td>1.021823</td>\n",
       "      <td>-0.020692</td>\n",
       "      <td>1.991121</td>\n",
       "      <td>-0.010392</td>\n",
       "      <td>0.991708</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.985760</td>\n",
       "      <td>chr13</td>\n",
       "      <td>103001424</td>\n",
       "      <td>103005924</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>liftover</td>\n",
       "      <td>lnc1000</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc10000</td>\n",
       "      <td>0.046372</td>\n",
       "      <td>-0.929325</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>-0.154737</td>\n",
       "      <td>0.877028</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.525104</td>\n",
       "      <td>chr10</td>\n",
       "      <td>79676628</td>\n",
       "      <td>79677257</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>liftover</td>\n",
       "      <td>lnc10000</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>4</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc10001</td>\n",
       "      <td>0.146825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr10</td>\n",
       "      <td>79677716</td>\n",
       "      <td>79678362</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>liftover</td>\n",
       "      <td>lnc10001</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>5</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc10003</td>\n",
       "      <td>0.177874</td>\n",
       "      <td>0.032402</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.005395</td>\n",
       "      <td>0.995695</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.022714</td>\n",
       "      <td>chr10</td>\n",
       "      <td>79954144</td>\n",
       "      <td>79959715</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>liftover</td>\n",
       "      <td>lnc10003</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>7</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc10004</td>\n",
       "      <td>0.740087</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.592880</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr10</td>\n",
       "      <td>80089972</td>\n",
       "      <td>80105010</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>liftover</td>\n",
       "      <td>lnc10004</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>8</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722262</th>\n",
       "      <td>LOC103694537</td>\n",
       "      <td>17.353741</td>\n",
       "      <td>0.172073</td>\n",
       "      <td>0.466195</td>\n",
       "      <td>0.369101</td>\n",
       "      <td>0.712052</td>\n",
       "      <td>0.850128</td>\n",
       "      <td>10</td>\n",
       "      <td>1.126676</td>\n",
       "      <td>chrY_NW_023637718v1_random</td>\n",
       "      <td>142349</td>\n",
       "      <td>142592</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>LOC103694537</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1141569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722263</th>\n",
       "      <td>LOC103694537</td>\n",
       "      <td>17.353741</td>\n",
       "      <td>0.172073</td>\n",
       "      <td>0.466195</td>\n",
       "      <td>0.369101</td>\n",
       "      <td>0.712052</td>\n",
       "      <td>0.850128</td>\n",
       "      <td>10</td>\n",
       "      <td>1.126676</td>\n",
       "      <td>chrY_NW_023637718v1_random</td>\n",
       "      <td>142792</td>\n",
       "      <td>143078</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>LOC103694537</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1141570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722264</th>\n",
       "      <td>LOC120099632</td>\n",
       "      <td>0.048942</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chrY_NW_023637718v1_random</td>\n",
       "      <td>187188</td>\n",
       "      <td>191437</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>LOC120099632</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1141571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722265</th>\n",
       "      <td>LOC120099632</td>\n",
       "      <td>0.048942</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chrY_NW_023637718v1_random</td>\n",
       "      <td>191439</td>\n",
       "      <td>192518</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>LOC120099632</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1141572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722266</th>\n",
       "      <td>Etaa1</td>\n",
       "      <td>12.976070</td>\n",
       "      <td>0.108092</td>\n",
       "      <td>0.443788</td>\n",
       "      <td>0.243566</td>\n",
       "      <td>0.807567</td>\n",
       "      <td>0.907902</td>\n",
       "      <td>10</td>\n",
       "      <td>1.077802</td>\n",
       "      <td>chrY_NW_023637721v1_random</td>\n",
       "      <td>82207</td>\n",
       "      <td>85543</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>Etaa1</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>NA</td>\n",
       "      <td>1141628</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7722267 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene   baseMean  log2FoldChange     lfcSE      stat  \\\n",
       "0             lnc1000   1.021823       -0.020692  1.991121 -0.010392   \n",
       "1            lnc10000   0.046372       -0.929325  6.005823 -0.154737   \n",
       "2            lnc10001   0.146825        0.000000  6.005823  0.000000   \n",
       "3            lnc10003   0.177874        0.032402  6.005823  0.005395   \n",
       "4            lnc10004   0.740087        0.000000  3.592880  0.000000   \n",
       "...               ...        ...             ...       ...       ...   \n",
       "7722262  LOC103694537  17.353741        0.172073  0.466195  0.369101   \n",
       "7722263  LOC103694537  17.353741        0.172073  0.466195  0.369101   \n",
       "7722264  LOC120099632   0.048942        0.000000  6.005823  0.000000   \n",
       "7722265  LOC120099632   0.048942        0.000000  6.005823  0.000000   \n",
       "7722266         Etaa1  12.976070        0.108092  0.443788  0.243566   \n",
       "\n",
       "           pvalue      padj  Dose  FoldChange                       chrom  \\\n",
       "0        0.991708        NA  0.01    0.985760                       chr13   \n",
       "1        0.877028        NA  0.01    0.525104                       chr10   \n",
       "2             1.0        NA  0.01    1.000000                       chr10   \n",
       "3        0.995695        NA  0.01    1.022714                       chr10   \n",
       "4             1.0        NA  0.01    1.000000                       chr10   \n",
       "...           ...       ...   ...         ...                         ...   \n",
       "7722262  0.712052  0.850128    10    1.126676  chrY_NW_023637718v1_random   \n",
       "7722263  0.712052  0.850128    10    1.126676  chrY_NW_023637718v1_random   \n",
       "7722264       1.0        NA    10    1.000000  chrY_NW_023637718v1_random   \n",
       "7722265       1.0        NA    10    1.000000  chrY_NW_023637718v1_random   \n",
       "7722266  0.807567  0.907902    10    1.077802  chrY_NW_023637721v1_random   \n",
       "\n",
       "             start        end frame  score strand      source       gene_id  \\\n",
       "0        103001424  103005924     .   1000      +    liftover       lnc1000   \n",
       "1         79676628   79677257     .   1000      -    liftover      lnc10000   \n",
       "2         79677716   79678362     .   1000      -    liftover      lnc10001   \n",
       "3         79954144   79959715     .   1000      -    liftover      lnc10003   \n",
       "4         80089972   80105010     .   1000      +    liftover      lnc10004   \n",
       "...            ...        ...   ...    ...    ...         ...           ...   \n",
       "7722262     142349     142592     .   1000      -  ncbiRefSeq  LOC103694537   \n",
       "7722263     142792     143078     .   1000      -  ncbiRefSeq  LOC103694537   \n",
       "7722264     187188     191437     .   1000      -  ncbiRefSeq  LOC120099632   \n",
       "7722265     191439     192518     .   1000      -  ncbiRefSeq  LOC120099632   \n",
       "7722266      82207      85543     .   1000      -  ncbiRefSeq         Etaa1   \n",
       "\n",
       "        transcript_type Mouse_ID   Rat_ID  \n",
       "0                lncRNA        3       NA  \n",
       "1                lncRNA        4       NA  \n",
       "2                lncRNA        5       NA  \n",
       "3                lncRNA        7       NA  \n",
       "4                lncRNA        8       NA  \n",
       "...                 ...      ...      ...  \n",
       "7722262  predicted_mRNA       NA  1141569  \n",
       "7722263  predicted_mRNA       NA  1141570  \n",
       "7722264  predicted_mRNA       NA  1141571  \n",
       "7722265  predicted_mRNA       NA  1141572  \n",
       "7722266  predicted_mRNA       NA  1141628  \n",
       "\n",
       "[7722267 rows x 20 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge Deseq2_Master with all metadata tables\n",
    "Deseq2_Master2 = pd.merge(Deseq2_Master, merged_combined2, left_on='Gene', right_on='gene_id', how='left').fillna(\"NA\")\n",
    "\n",
    "Deseq2_Master2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Dose</th>\n",
       "      <th>pDRE</th>\n",
       "      <th>AHR</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>Log2FC</th>\n",
       "      <th>Fold-Change</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>...</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>frame</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>Rat_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc1000</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1.021823</td>\n",
       "      <td>-0.020692</td>\n",
       "      <td>0.985760</td>\n",
       "      <td>1.991121</td>\n",
       "      <td>-0.010392</td>\n",
       "      <td>...</td>\n",
       "      <td>chr13</td>\n",
       "      <td>103001424</td>\n",
       "      <td>103005924</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>liftover</td>\n",
       "      <td>lnc1000</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc10000</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.046372</td>\n",
       "      <td>-0.929325</td>\n",
       "      <td>0.525104</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>-0.154737</td>\n",
       "      <td>...</td>\n",
       "      <td>chr10</td>\n",
       "      <td>79676628</td>\n",
       "      <td>79677257</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>liftover</td>\n",
       "      <td>lnc10000</td>\n",
       "      <td>4</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc10001</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.146825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>chr10</td>\n",
       "      <td>79677716</td>\n",
       "      <td>79678362</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>liftover</td>\n",
       "      <td>lnc10001</td>\n",
       "      <td>5</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc10003</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.177874</td>\n",
       "      <td>0.032402</td>\n",
       "      <td>1.022714</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.005395</td>\n",
       "      <td>...</td>\n",
       "      <td>chr10</td>\n",
       "      <td>79954144</td>\n",
       "      <td>79959715</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>liftover</td>\n",
       "      <td>lnc10003</td>\n",
       "      <td>7</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc10004</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.740087</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.592880</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>chr10</td>\n",
       "      <td>80089972</td>\n",
       "      <td>80105010</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>+</td>\n",
       "      <td>liftover</td>\n",
       "      <td>lnc10004</td>\n",
       "      <td>8</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722262</th>\n",
       "      <td>LOC103694537</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>17.353741</td>\n",
       "      <td>0.172073</td>\n",
       "      <td>1.126676</td>\n",
       "      <td>0.466195</td>\n",
       "      <td>0.369101</td>\n",
       "      <td>...</td>\n",
       "      <td>chrY_NW_023637718v1_random</td>\n",
       "      <td>142349</td>\n",
       "      <td>142592</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>LOC103694537</td>\n",
       "      <td>NA</td>\n",
       "      <td>1141569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722263</th>\n",
       "      <td>LOC103694537</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>17.353741</td>\n",
       "      <td>0.172073</td>\n",
       "      <td>1.126676</td>\n",
       "      <td>0.466195</td>\n",
       "      <td>0.369101</td>\n",
       "      <td>...</td>\n",
       "      <td>chrY_NW_023637718v1_random</td>\n",
       "      <td>142792</td>\n",
       "      <td>143078</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>LOC103694537</td>\n",
       "      <td>NA</td>\n",
       "      <td>1141570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722264</th>\n",
       "      <td>LOC120099632</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>0.048942</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>chrY_NW_023637718v1_random</td>\n",
       "      <td>187188</td>\n",
       "      <td>191437</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>LOC120099632</td>\n",
       "      <td>NA</td>\n",
       "      <td>1141571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722265</th>\n",
       "      <td>LOC120099632</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>0.048942</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>chrY_NW_023637718v1_random</td>\n",
       "      <td>191439</td>\n",
       "      <td>192518</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>LOC120099632</td>\n",
       "      <td>NA</td>\n",
       "      <td>1141572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722266</th>\n",
       "      <td>Etaa1</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>12.976070</td>\n",
       "      <td>0.108092</td>\n",
       "      <td>1.077802</td>\n",
       "      <td>0.443788</td>\n",
       "      <td>0.243566</td>\n",
       "      <td>...</td>\n",
       "      <td>chrY_NW_023637721v1_random</td>\n",
       "      <td>82207</td>\n",
       "      <td>85543</td>\n",
       "      <td>.</td>\n",
       "      <td>1000</td>\n",
       "      <td>-</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "      <td>Etaa1</td>\n",
       "      <td>NA</td>\n",
       "      <td>1141628</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7722267 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene  Dose  pDRE  AHR transcript_type   baseMean    Log2FC  \\\n",
       "0             lnc1000  0.01     0    1          lncRNA   1.021823 -0.020692   \n",
       "1            lnc10000  0.01     1    0          lncRNA   0.046372 -0.929325   \n",
       "2            lnc10001  0.01     0    0          lncRNA   0.146825  0.000000   \n",
       "3            lnc10003  0.01     2    0          lncRNA   0.177874  0.032402   \n",
       "4            lnc10004  0.01     7    1          lncRNA   0.740087  0.000000   \n",
       "...               ...   ...   ...  ...             ...        ...       ...   \n",
       "7722262  LOC103694537    10     0    0  predicted_mRNA  17.353741  0.172073   \n",
       "7722263  LOC103694537    10     0    0  predicted_mRNA  17.353741  0.172073   \n",
       "7722264  LOC120099632    10     0    0  predicted_mRNA   0.048942  0.000000   \n",
       "7722265  LOC120099632    10     0    0  predicted_mRNA   0.048942  0.000000   \n",
       "7722266         Etaa1    10     0    0  predicted_mRNA  12.976070  0.108092   \n",
       "\n",
       "         Fold-Change     lfcSE      stat  ...                       chrom  \\\n",
       "0           0.985760  1.991121 -0.010392  ...                       chr13   \n",
       "1           0.525104  6.005823 -0.154737  ...                       chr10   \n",
       "2           1.000000  6.005823  0.000000  ...                       chr10   \n",
       "3           1.022714  6.005823  0.005395  ...                       chr10   \n",
       "4           1.000000  3.592880  0.000000  ...                       chr10   \n",
       "...              ...       ...       ...  ...                         ...   \n",
       "7722262     1.126676  0.466195  0.369101  ...  chrY_NW_023637718v1_random   \n",
       "7722263     1.126676  0.466195  0.369101  ...  chrY_NW_023637718v1_random   \n",
       "7722264     1.000000  6.005823  0.000000  ...  chrY_NW_023637718v1_random   \n",
       "7722265     1.000000  6.005823  0.000000  ...  chrY_NW_023637718v1_random   \n",
       "7722266     1.077802  0.443788  0.243566  ...  chrY_NW_023637721v1_random   \n",
       "\n",
       "             start        end  frame  score strand      source       gene_id  \\\n",
       "0        103001424  103005924      .   1000      +    liftover       lnc1000   \n",
       "1         79676628   79677257      .   1000      -    liftover      lnc10000   \n",
       "2         79677716   79678362      .   1000      -    liftover      lnc10001   \n",
       "3         79954144   79959715      .   1000      -    liftover      lnc10003   \n",
       "4         80089972   80105010      .   1000      +    liftover      lnc10004   \n",
       "...            ...        ...    ...    ...    ...         ...           ...   \n",
       "7722262     142349     142592      .   1000      -  ncbiRefSeq  LOC103694537   \n",
       "7722263     142792     143078      .   1000      -  ncbiRefSeq  LOC103694537   \n",
       "7722264     187188     191437      .   1000      -  ncbiRefSeq  LOC120099632   \n",
       "7722265     191439     192518      .   1000      -  ncbiRefSeq  LOC120099632   \n",
       "7722266      82207      85543      .   1000      -  ncbiRefSeq         Etaa1   \n",
       "\n",
       "        Mouse_ID   Rat_ID  \n",
       "0              3       NA  \n",
       "1              4       NA  \n",
       "2              5       NA  \n",
       "3              7       NA  \n",
       "4              8       NA  \n",
       "...          ...      ...  \n",
       "7722262       NA  1141569  \n",
       "7722263       NA  1141570  \n",
       "7722264       NA  1141571  \n",
       "7722265       NA  1141572  \n",
       "7722266       NA  1141628  \n",
       "\n",
       "[7722267 rows x 22 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop the 'Gene' column\n",
    "#Deseq2_Master2 = Deseq2_Master2.drop(columns=['Gene'])\n",
    "\n",
    "#Deseq2_Master2 = Deseq2_Master2.rename(columns={'gene_name': 'Gene'})\n",
    "Deseq2_Master2 = Deseq2_Master2.rename(columns={'FoldChange': 'Fold-Change'})\n",
    "Deseq2_Master2 = Deseq2_Master2.rename(columns={'log2FoldChange': 'Log2FC'})\n",
    "\n",
    "Deseq2_Master2 = pd.merge(Deseq2_Master2, pDREs_Master, left_on='Gene', right_on='Gene', how='left').fillna(0)\n",
    "Deseq2_Master2 = pd.merge(Deseq2_Master2, AHR_Master, left_on='Gene', right_on='Gene', how='left').fillna(0)\n",
    "\n",
    "Deseq2_Master2[['pDRE', 'AHR']] = Deseq2_Master2[['pDRE', 'AHR']].astype(int)\n",
    "\n",
    "Deseq2_Master_Working = Deseq2_Master2[['Gene', 'Dose', 'pDRE', 'AHR', 'transcript_type',  'baseMean',\n",
    "                                      'Log2FC', 'Fold-Change', 'lfcSE', 'stat', 'pvalue', 'padj',\n",
    "                                       'chrom', 'start', 'end', 'frame', 'score','strand', 'source',\n",
    "                                        'gene_id', 'Mouse_ID', 'Rat_ID']].drop_duplicates(keep='first')\n",
    "\n",
    "Deseq2_Master_Working"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Dose</th>\n",
       "      <th>pDRE</th>\n",
       "      <th>AHR</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>Log2FC</th>\n",
       "      <th>Fold-Change</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>lnc1000</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1.021823</td>\n",
       "      <td>-0.020692</td>\n",
       "      <td>0.985760</td>\n",
       "      <td>1.991121</td>\n",
       "      <td>-0.010392</td>\n",
       "      <td>0.991708</td>\n",
       "      <td>NA</td>\n",
       "      <td>liftover</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lnc10000</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.046372</td>\n",
       "      <td>-0.929325</td>\n",
       "      <td>0.525104</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>-0.154737</td>\n",
       "      <td>0.877028</td>\n",
       "      <td>NA</td>\n",
       "      <td>liftover</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lnc10001</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.146825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>liftover</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lnc10003</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.177874</td>\n",
       "      <td>0.032402</td>\n",
       "      <td>1.022714</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.005395</td>\n",
       "      <td>0.995695</td>\n",
       "      <td>NA</td>\n",
       "      <td>liftover</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>lnc10004</td>\n",
       "      <td>0.01</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>0.740087</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.592880</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>liftover</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722234</th>\n",
       "      <td>LOC120099597</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>predicted_lncRNA</td>\n",
       "      <td>0.092333</td>\n",
       "      <td>-1.398142</td>\n",
       "      <td>0.379417</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>-0.232798</td>\n",
       "      <td>0.815918</td>\n",
       "      <td>NA</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722237</th>\n",
       "      <td>Dkc1</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>277.395024</td>\n",
       "      <td>1.901728</td>\n",
       "      <td>3.736606</td>\n",
       "      <td>0.248099</td>\n",
       "      <td>7.665213</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722252</th>\n",
       "      <td>LOC103694537</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>17.353741</td>\n",
       "      <td>0.172073</td>\n",
       "      <td>1.126676</td>\n",
       "      <td>0.466195</td>\n",
       "      <td>0.369101</td>\n",
       "      <td>0.712052</td>\n",
       "      <td>0.850128</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722264</th>\n",
       "      <td>LOC120099632</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>0.048942</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>6.005823</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NA</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7722266</th>\n",
       "      <td>Etaa1</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>predicted_mRNA</td>\n",
       "      <td>12.976070</td>\n",
       "      <td>0.108092</td>\n",
       "      <td>1.077802</td>\n",
       "      <td>0.443788</td>\n",
       "      <td>0.243566</td>\n",
       "      <td>0.807567</td>\n",
       "      <td>0.907902</td>\n",
       "      <td>ncbiRefSeq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>402199 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene  Dose  pDRE  AHR   transcript_type    baseMean  \\\n",
       "0             lnc1000  0.01     0    1            lncRNA    1.021823   \n",
       "1            lnc10000  0.01     1    0            lncRNA    0.046372   \n",
       "2            lnc10001  0.01     0    0            lncRNA    0.146825   \n",
       "3            lnc10003  0.01     2    0            lncRNA    0.177874   \n",
       "4            lnc10004  0.01     7    1            lncRNA    0.740087   \n",
       "...               ...   ...   ...  ...               ...         ...   \n",
       "7722234  LOC120099597    10     1    0  predicted_lncRNA    0.092333   \n",
       "7722237          Dkc1    10     0    2              mRNA  277.395024   \n",
       "7722252  LOC103694537    10     0    0    predicted_mRNA   17.353741   \n",
       "7722264  LOC120099632    10     0    0    predicted_mRNA    0.048942   \n",
       "7722266         Etaa1    10     0    0    predicted_mRNA   12.976070   \n",
       "\n",
       "           Log2FC  Fold-Change     lfcSE      stat    pvalue      padj  \\\n",
       "0       -0.020692     0.985760  1.991121 -0.010392  0.991708        NA   \n",
       "1       -0.929325     0.525104  6.005823 -0.154737  0.877028        NA   \n",
       "2        0.000000     1.000000  6.005823  0.000000       1.0        NA   \n",
       "3        0.032402     1.022714  6.005823  0.005395  0.995695        NA   \n",
       "4        0.000000     1.000000  3.592880  0.000000       1.0        NA   \n",
       "...           ...          ...       ...       ...       ...       ...   \n",
       "7722234 -1.398142     0.379417  6.005823 -0.232798  0.815918        NA   \n",
       "7722237  1.901728     3.736606  0.248099  7.665213       0.0       0.0   \n",
       "7722252  0.172073     1.126676  0.466195  0.369101  0.712052  0.850128   \n",
       "7722264  0.000000     1.000000  6.005823  0.000000       1.0        NA   \n",
       "7722266  0.108092     1.077802  0.443788  0.243566  0.807567  0.907902   \n",
       "\n",
       "             source  \n",
       "0          liftover  \n",
       "1          liftover  \n",
       "2          liftover  \n",
       "3          liftover  \n",
       "4          liftover  \n",
       "...             ...  \n",
       "7722234  ncbiRefSeq  \n",
       "7722237  ncbiRefSeq  \n",
       "7722252  ncbiRefSeq  \n",
       "7722264  ncbiRefSeq  \n",
       "7722266  ncbiRefSeq  \n",
       "\n",
       "[402199 rows x 13 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Deseq2_Master_Working2 = Deseq2_Master_Working[['Gene', 'Dose', 'pDRE', 'AHR', 'transcript_type',  'baseMean',\n",
    "                                      'Log2FC', 'Fold-Change', 'lfcSE', 'stat', 'pvalue', 'padj','source' ]].drop_duplicates(keep='first')\n",
    "Deseq2_Master_Working2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "Deseq2_Master_Working2.to_csv('./RDDR_Deseq2_Master_Wald_Rat_with_Mouse_MGI.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python (scvi-new)",
   "language": "python",
   "name": "scvi"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
